// ng-zorro-antd-mobile/esm2015/tabs/tab-pane.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, TemplateRef } from '@angular/core';
export class TabPaneComponent {
    constructor() {
        this.isTitleString = true;
    }
    /**
     * @return {?}
     */
    get title() {
        return this._title;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set title(value) {
        this.isTitleString = !(value instanceof TemplateRef);
        this._title = value;
    }
}
TabPaneComponent.decorators = [
    { type: Component, args: [{
                selector: 'TabPane, nzm-tab-pane',
                template: "<ng-template #content>\n  <ng-content></ng-content>\n</ng-template>"
            }] }
];
/** @nocollapse */
TabPaneComponent.ctorParameters = () => [];
TabPaneComponent.propDecorators = {
    content: [{ type: ViewChild, args: ['content',] }],
    title: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TabPaneComponent.prototype.isTitleString;
    /**
     * @type {?}
     * @private
     */
    TabPaneComponent.prototype._title;
    /** @type {?} */
    TabPaneComponent.prototype.content;
}
