// ng-zorro-antd-mobile/esm2015/tabs/tab-pane-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, TemplateRef, ViewEncapsulation } from '@angular/core';
export class TabPaneBodyComponent {
    constructor() {
        this._prerender = false;
        this.active = false;
        this.loaded = false;
        this.paneWrap = true;
    }
    /**
     * @return {?}
     */
    get prerender() {
        return this._prerender;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set prerender(value) {
        this._prerender = value;
        if (value) {
            this.loaded = true;
        }
    }
    /**
     * @return {?}
     */
    get wrapActive() {
        return this.active;
    }
    /**
     * @return {?}
     */
    get wrapInactive() {
        return !this.active;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
TabPaneBodyComponent.decorators = [
    { type: Component, args: [{
                selector: '[tab-pane-body]',
                template: "<ng-container *ngIf=\"loaded || prerender\">\n  <ng-template [ngTemplateOutlet]=\"content\"></ng-template>\n</ng-container>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
TabPaneBodyComponent.ctorParameters = () => [];
TabPaneBodyComponent.propDecorators = {
    active: [{ type: Input }],
    loaded: [{ type: Input }],
    content: [{ type: Input }],
    prerender: [{ type: Input }],
    paneWrap: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap',] }],
    wrapActive: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap-active',] }],
    wrapInactive: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap-inactive',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    TabPaneBodyComponent.prototype._prerender;
    /** @type {?} */
    TabPaneBodyComponent.prototype.active;
    /** @type {?} */
    TabPaneBodyComponent.prototype.loaded;
    /** @type {?} */
    TabPaneBodyComponent.prototype.content;
    /** @type {?} */
    TabPaneBodyComponent.prototype.paneWrap;
}
