// ng-zorro-antd-mobile/esm5/modal/modal.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ModalComponent, ModalServiceComponent } from './modal.component';
import { ModalService } from './modal.service';
import { ListModule } from '../list/list.module';
import { WingBlankModule } from '../wing-blank/wing-blank.module';
import { WhiteSpaceModule } from '../white-space/white-space.module';
import { ButtonModule } from '../button/button.module';
import { InputItemModule } from '../input-item/input-item.module';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AlertOptions } from './modal-options.provider';
import { OverlayModule } from '@angular/cdk/overlay';
import { PopupService } from '../core/services/popup.service';
var ModalModule = /** @class */ (function () {
    function ModalModule() {
    }
    ModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ListModule,
                        WingBlankModule,
                        WhiteSpaceModule,
                        ButtonModule,
                        InputItemModule,
                        FormsModule,
                        ReactiveFormsModule,
                        OverlayModule
                    ],
                    exports: [ModalComponent, ModalServiceComponent],
                    declarations: [ModalComponent, ModalServiceComponent],
                    entryComponents: [ModalServiceComponent],
                    providers: [AlertOptions, ModalService, PopupService]
                },] }
    ];
    return ModalModule;
}());
export { ModalModule };
