// ng-zorro-antd-mobile/esm5/modal/modal-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var ModalBaseOptions = /** @class */ (function () {
    function ModalBaseOptions() {
        this.visible = false;
        this.focus = true;
        this.prefixCls = 'am-modal';
        this.animated = true;
        this.closable = false;
        this.maskClosable = true;
        this.transparent = false;
        this.popup = false;
        this.animationType = 'slide-down';
        this.footer = [];
        this.platform = 'ios';
        this.defaultValue = [];
        this.placeholders = [];
        this.transitionName = 'am-zoom';
        this.maskTransitionName = 'am-fade';
    }
    return ModalBaseOptions;
}());
export { ModalBaseOptions };
if (false) {
    /** @type {?} */
    ModalBaseOptions.prototype.visible;
    /** @type {?} */
    ModalBaseOptions.prototype.focus;
    /** @type {?} */
    ModalBaseOptions.prototype.prefixCls;
    /** @type {?} */
    ModalBaseOptions.prototype.animated;
    /** @type {?} */
    ModalBaseOptions.prototype.closable;
    /** @type {?} */
    ModalBaseOptions.prototype.maskClosable;
    /** @type {?} */
    ModalBaseOptions.prototype.onClose;
    /** @type {?} */
    ModalBaseOptions.prototype.transparent;
    /** @type {?} */
    ModalBaseOptions.prototype.popup;
    /** @type {?} */
    ModalBaseOptions.prototype.animationType;
    /** @type {?} */
    ModalBaseOptions.prototype.title;
    /** @type {?} */
    ModalBaseOptions.prototype.footer;
    /** @type {?} */
    ModalBaseOptions.prototype.platform;
    /** @type {?} */
    ModalBaseOptions.prototype.className;
    /** @type {?} */
    ModalBaseOptions.prototype.wrapClassName;
    /** @type {?} */
    ModalBaseOptions.prototype.message;
    /** @type {?} */
    ModalBaseOptions.prototype.actions;
    /** @type {?} */
    ModalBaseOptions.prototype.callbackOrActions;
    /** @type {?} */
    ModalBaseOptions.prototype.type;
    /** @type {?} */
    ModalBaseOptions.prototype.defaultValue;
    /** @type {?} */
    ModalBaseOptions.prototype.placeholders;
    /** @type {?} */
    ModalBaseOptions.prototype.operation;
    /** @type {?} */
    ModalBaseOptions.prototype.transitionName;
    /** @type {?} */
    ModalBaseOptions.prototype.maskTransitionName;
    /** @type {?} */
    ModalBaseOptions.prototype.close;
}
var ModalOptions = /** @class */ (function (_super) {
    tslib_1.__extends(ModalOptions, _super);
    function ModalOptions() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.transitionName = 'am-fade';
        _this.maskTransitionName = 'am-fade';
        return _this;
    }
    ModalOptions.decorators = [
        { type: Injectable }
    ];
    return ModalOptions;
}(ModalBaseOptions));
export { ModalOptions };
if (false) {
    /** @type {?} */
    ModalOptions.prototype.transitionName;
    /** @type {?} */
    ModalOptions.prototype.maskTransitionName;
}
var AlertOptions = /** @class */ (function (_super) {
    tslib_1.__extends(AlertOptions, _super);
    function AlertOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AlertOptions.decorators = [
        { type: Injectable }
    ];
    return AlertOptions;
}(ModalBaseOptions));
export { AlertOptions };
if (false) {
    /** @type {?} */
    AlertOptions.prototype.message;
    /** @type {?} */
    AlertOptions.prototype.actions;
}
var Action = /** @class */ (function () {
    function Action() {
    }
    Action.decorators = [
        { type: Injectable }
    ];
    return Action;
}());
export { Action };
if (false) {
    /** @type {?} */
    Action.prototype.text;
    /** @type {?} */
    Action.prototype.onPress;
    /** @type {?} */
    Action.prototype.style;
}
