// ng-zorro-antd-mobile/esm5/modal/modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostListener, Input, Output, TemplateRef, EventEmitter, ViewEncapsulation, forwardRef } from '@angular/core';
import { ModalOptions } from './modal-options.provider';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var ModalComponent = /** @class */ (function () {
    function ModalComponent(option) {
        this.option = option;
        this.autoFocus = { focus: true, date: new Date() };
        this.transitionName = '';
        this.maskTransitionName = '';
        this.wrapCls = {};
        this.cls = {};
        this.btnGroupClass = {};
        this.data = {
            text: '',
            password: ''
        };
        this.onClose = new EventEmitter();
    }
    Object.defineProperty(ModalComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this.option.title;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "closable", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.closable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "maskClosable", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.maskClosable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "popup", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.popup = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "animationType", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.animationType = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "transparent", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.transparent = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "footer", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.footer = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "platform", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.platform = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "className", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.className = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "wrapClassName", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.wrapClassName = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "actions", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.footer = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "defaultValue", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.defaultValue = value !== undefined ? value : ['', ''];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "type", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.type = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "placeholders", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.placeholders = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ModalComponent.prototype, "operation", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.option.operation = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    ModalComponent.prototype.panend = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.option.closable || this.option.maskClosable || this.option.popup) {
            if ((event && event.target && event.target.getAttribute('role') === 'dialog') ||
                event.target.getAttribute('role') === 'close') {
                if (this.option.close) {
                    this.option.close();
                }
                else {
                    this.onClose.emit();
                    this.leaveAnimation();
                }
            }
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ModalComponent.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof TemplateRef;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ModalComponent.prototype.isNoTitle = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value === '' || value === null || value === undefined;
    };
    /**
     * @param {?} visible
     * @return {?}
     */
    ModalComponent.prototype.setTransitionName = /**
     * @param {?} visible
     * @return {?}
     */
    function (visible) {
        if (!visible) {
            this.leaveAnimation();
        }
        else {
            if (this.option.animated) {
                if (this.option.transparent) {
                    if (this.setActiveName(this.option.transitionName)) {
                        this.transitionName = this.setActiveName(this.option.transitionName);
                        this.maskTransitionName = this.setActiveName(this.option.maskTransitionName);
                    }
                    else {
                        this.transitionName = this.maskTransitionName = this.setActiveName('am-fade');
                    }
                }
                else {
                    if (this.setActiveName(this.option.transitionName)) {
                        this.transitionName = this.setActiveName(this.option.transitionName);
                        this.maskTransitionName = this.setActiveName(this.option.maskTransitionName);
                    }
                    else {
                        this.transitionName = this.maskTransitionName = this.setActiveName('am-slide-up');
                    }
                }
                if (this.option.popup) {
                    this.transitionName =
                        this.option.animationType === 'slide-up'
                            ? this.setActiveName('am-slide-up')
                            : this.setActiveName('am-slide-down');
                    this.maskTransitionName = this.setActiveName('am-fade');
                }
            }
            this.setClassMap();
        }
    };
    /**
     * @param {?} name
     * @return {?}
     */
    ModalComponent.prototype.setActiveName = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return name.length > 0 ? name + "-enter " + name + "-enter-active" : null;
    };
    /**
     * @param {?} name
     * @return {?}
     */
    ModalComponent.prototype.setLeaveActiveName = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return name.length > 0 ? name + "-leave " + name + "-leave-active" : null;
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.setClassMap = /**
     * @return {?}
     */
    function () {
        var _a, _b, _c;
        this.wrapCls = (_a = {},
            _a[this.option.wrapClassName] = true,
            _a[this.option.prefixCls + "-wrap-popup"] = this.option.popup,
            _a);
        this.cls = (_b = {},
            _b[this.option.className] = true,
            _b[this.option.prefixCls + "-transparent"] = this.option.transparent,
            _b[this.option.prefixCls + "-popup"] = this.option.popup,
            _b[this.option.prefixCls + "-popup-" + this.option.animationType] = this.option.popup && this.option.animationType,
            _b[this.option.prefixCls + "-android"] = this.option.platform === 'android',
            _b);
        this.btnGroupClass = (_c = {},
            _c[this.option.prefixCls + "-button-group-" + (this.option.footer.length === 2 && !this.option.operation ? 'h' : 'v')] = true,
            _c[this.option.prefixCls + "-button-group-" + (this.option.operation ? 'operation' : 'normal')] = true,
            _c);
    };
    /**
     * @param {?} type
     * @param {?} value
     * @return {?}
     */
    ModalComponent.prototype.inputChange = /**
     * @param {?} type
     * @param {?} value
     * @return {?}
     */
    function (type, value) {
        this.data[type] = value;
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.leaveAnimation = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.option.animated) {
            if (this.option.transparent) {
                if (this.setLeaveActiveName(this.option.transitionName)) {
                    this.transitionName = this.setLeaveActiveName(this.option.transitionName);
                    this.maskTransitionName = this.setLeaveActiveName(this.option.maskTransitionName);
                }
                else {
                    this.transitionName = this.maskTransitionName = this.setLeaveActiveName('am-fade');
                }
            }
            else {
                if (this.setLeaveActiveName(this.option.transitionName)) {
                    this.transitionName = this.setLeaveActiveName(this.option.transitionName);
                    this.maskTransitionName = this.setLeaveActiveName(this.option.maskTransitionName);
                }
                else {
                    this.transitionName = this.maskTransitionName = this.setLeaveActiveName('am-slide-up');
                }
            }
            if (this.option.popup) {
                this.transitionName =
                    this.option.animationType === 'slide-up'
                        ? this.setLeaveActiveName('am-slide-up')
                        : this.setLeaveActiveName('am-slide-down');
                this.maskTransitionName = this.setLeaveActiveName('am-fade');
            }
        }
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.option.visible = false;
            _this.onChanged(_this.option.visible);
        }), 200);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ModalComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            this.option.visible = value;
        }
        this.setTransitionName(value);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ModalComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChanged = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ModalComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    ModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Modal',
                    template: "<div *ngIf=\"option.visible\">\n  <div class=\"{{option.prefixCls}}-mask {{maskTransitionName}}\"></div>\n  <div role=\"dialog\"\n       class=\"{{option.prefixCls}}-wrap {{transitionName}}\"\n       [ngClass]=\"wrapCls\"\n  >\n    <div role=\"document\"\n         class=\"{{option.prefixCls}}\"\n         [ngClass]=\"cls\"\n    >\n      <div class=\"{{option.prefixCls}}-content\">\n        <div  *ngIf=\"option.closable\" style=\"width: 100%; height: 21px;\">\n          <div role=\"close\"\n               class=\"{{option.prefixCls}}-close\"\n        >\n            <span role=\"close\" class=\"{{option.prefixCls}}-close-x\"></span>\n          </div>\n        </div>\n        <div *ngIf=\"!isNoTitle(option.title)\" class=\"{{option.prefixCls}}-header\">\n          <div *ngIf=\"!isTemplateRef(option.title)\" class=\"{{option.prefixCls}}-title\">{{option.title}}</div>\n          <ng-template *ngIf=\"isTemplateRef(option.title)\" [ngTemplateOutlet]=\"option.title\"></ng-template>\n        </div>\n        <div class=\"{{option.prefixCls}}-body\">\n          <ng-content></ng-content>\n          <div *ngIf=\"!isTemplateRef(option.message)\" class=\"{{option.prefixCls}}-alert-content\">{{option.message}}</div>\n          <ng-template *ngIf=\"isTemplateRef(option.message)\" [ngTemplateOutlet]=\"option.message\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'default'\" [ngTemplateOutlet]=\"promptDefault\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'secure-text'\" [ngTemplateOutlet]=\"promptSecure\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'login-password'\" [ngTemplateOutlet]=\"promptPassword\"></ng-template>\n        </div>\n        <div class=\"{{option.prefixCls}}-footer\">\n          <div [ngClass]=\"btnGroupClass\" role=\"group\">\n            <div Button\n                 role=\"button\"\n                 *ngFor=\"let button of option.footer\"\n                 [className]=\"'am-modal-button'\"\n                 [ngStyle]=\"button.style\"\n                 (onClick)=\"button.onPress()\"\n            >\n              {{button.text}}\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n\n<ng-template #promptPassword>\n  <div class=\"{{option.prefixCls}}-input-container\">\n    <div class=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          autofocus\n          [type]=\"'text'\"\n          [placeholder]=\"option.placeholders[0] || ''\"\n          [(ngModel)]=\"option.defaultValue[0]\"\n          (ngModelChange)=\"inputChange('text', $event)\"\n        >\n    </div>\n    <div className=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          [type]=\"'password'\"\n          [placeholder]=\"option.placeholders[1] || ''\"\n          [(ngModel)]=\"option.defaultValue[1]\"\n          (ngModelChange)=\"inputChange('password', $event)\"\n        >\n    </div>\n  </div>\n</ng-template>\n<ng-template #promptSecure>\n  <div className=\"{{option.prefixCls}}-input-container\">\n    <div className=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          autofocus\n          [type]=\"'password'\"\n          [placeholder]=\"option.placeholders[0] || ''\"\n          [(ngModel)]=\"option.defaultValue[0]\"\n          (ngModelChange)=\"inputChange('password', $event)\"\n        >\n    </div>\n  </div>\n</ng-template>\n<ng-template #promptDefault>\n  <div className=\"{{option.prefixCls}}-input-container\">\n    <div className=\"{{option.prefixCls}}-input\">\n      <input\n        #inputElement\n        autofocus\n        [type]=\"'text'\"\n        [placeholder]=\"option.placeholders[0] || ''\"\n        [(ngModel)]=\"option.defaultValue[0]\"\n        (ngModelChange)=\"inputChange('text', $event)\"\n      >\n    </div>\n  </div>\n</ng-template>",
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        ModalOptions,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ModalComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    ModalComponent.ctorParameters = function () { return [
        { type: ModalOptions }
    ]; };
    ModalComponent.propDecorators = {
        title: [{ type: Input }],
        closable: [{ type: Input }],
        maskClosable: [{ type: Input }],
        popup: [{ type: Input }],
        animationType: [{ type: Input }],
        transparent: [{ type: Input }],
        footer: [{ type: Input }],
        platform: [{ type: Input }],
        className: [{ type: Input }],
        wrapClassName: [{ type: Input }],
        actions: [{ type: Input }],
        defaultValue: [{ type: Input }],
        type: [{ type: Input }],
        placeholders: [{ type: Input }],
        operation: [{ type: Input }],
        onClose: [{ type: Output }],
        panend: [{ type: HostListener, args: ['mouseup', ['$event'],] }, { type: HostListener, args: ['touchend', ['$event'],] }]
    };
    return ModalComponent;
}());
export { ModalComponent };
if (false) {
    /** @type {?} */
    ModalComponent.prototype.autoFocus;
    /** @type {?} */
    ModalComponent.prototype.transitionName;
    /** @type {?} */
    ModalComponent.prototype.maskTransitionName;
    /** @type {?} */
    ModalComponent.prototype.wrapCls;
    /** @type {?} */
    ModalComponent.prototype.cls;
    /** @type {?} */
    ModalComponent.prototype.btnGroupClass;
    /** @type {?} */
    ModalComponent.prototype.data;
    /** @type {?} */
    ModalComponent.prototype.onChanged;
    /** @type {?} */
    ModalComponent.prototype.onTouched;
    /** @type {?} */
    ModalComponent.prototype.onClose;
    /** @type {?} */
    ModalComponent.prototype.option;
}
var ModalServiceComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ModalServiceComponent, _super);
    function ModalServiceComponent(option) {
        var _this = _super.call(this, option) || this;
        _this.option = option;
        _this.setTransitionName(_this.option.visible);
        return _this;
    }
    ModalServiceComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ModalService',
                    template: "<div *ngIf=\"option.visible\">\n  <div class=\"{{option.prefixCls}}-mask {{maskTransitionName}}\"></div>\n  <div role=\"dialog\"\n       class=\"{{option.prefixCls}}-wrap {{transitionName}}\"\n       [ngClass]=\"wrapCls\"\n  >\n    <div role=\"document\"\n         class=\"{{option.prefixCls}}\"\n         [ngClass]=\"cls\"\n    >\n      <div class=\"{{option.prefixCls}}-content\">\n        <div  *ngIf=\"option.closable\" style=\"width: 100%; height: 21px;\">\n          <div role=\"close\"\n               class=\"{{option.prefixCls}}-close\"\n        >\n            <span role=\"close\" class=\"{{option.prefixCls}}-close-x\"></span>\n          </div>\n        </div>\n        <div *ngIf=\"!isNoTitle(option.title)\" class=\"{{option.prefixCls}}-header\">\n          <div *ngIf=\"!isTemplateRef(option.title)\" class=\"{{option.prefixCls}}-title\">{{option.title}}</div>\n          <ng-template *ngIf=\"isTemplateRef(option.title)\" [ngTemplateOutlet]=\"option.title\"></ng-template>\n        </div>\n        <div class=\"{{option.prefixCls}}-body\">\n          <ng-content></ng-content>\n          <div *ngIf=\"!isTemplateRef(option.message)\" class=\"{{option.prefixCls}}-alert-content\">{{option.message}}</div>\n          <ng-template *ngIf=\"isTemplateRef(option.message)\" [ngTemplateOutlet]=\"option.message\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'default'\" [ngTemplateOutlet]=\"promptDefault\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'secure-text'\" [ngTemplateOutlet]=\"promptSecure\"></ng-template>\n          <ng-template *ngIf=\"option.type === 'login-password'\" [ngTemplateOutlet]=\"promptPassword\"></ng-template>\n        </div>\n        <div class=\"{{option.prefixCls}}-footer\">\n          <div [ngClass]=\"btnGroupClass\" role=\"group\">\n            <div Button\n                 role=\"button\"\n                 *ngFor=\"let button of option.footer\"\n                 [className]=\"'am-modal-button'\"\n                 [ngStyle]=\"button.style\"\n                 (onClick)=\"button.onPress()\"\n            >\n              {{button.text}}\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n\n<ng-template #promptPassword>\n  <div class=\"{{option.prefixCls}}-input-container\">\n    <div class=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          autofocus\n          [type]=\"'text'\"\n          [placeholder]=\"option.placeholders[0] || ''\"\n          [(ngModel)]=\"option.defaultValue[0]\"\n          (ngModelChange)=\"inputChange('text', $event)\"\n        >\n    </div>\n    <div className=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          [type]=\"'password'\"\n          [placeholder]=\"option.placeholders[1] || ''\"\n          [(ngModel)]=\"option.defaultValue[1]\"\n          (ngModelChange)=\"inputChange('password', $event)\"\n        >\n    </div>\n  </div>\n</ng-template>\n<ng-template #promptSecure>\n  <div className=\"{{option.prefixCls}}-input-container\">\n    <div className=\"{{option.prefixCls}}-input\">\n        <input\n          #inputElement\n          autofocus\n          [type]=\"'password'\"\n          [placeholder]=\"option.placeholders[0] || ''\"\n          [(ngModel)]=\"option.defaultValue[0]\"\n          (ngModelChange)=\"inputChange('password', $event)\"\n        >\n    </div>\n  </div>\n</ng-template>\n<ng-template #promptDefault>\n  <div className=\"{{option.prefixCls}}-input-container\">\n    <div className=\"{{option.prefixCls}}-input\">\n      <input\n        #inputElement\n        autofocus\n        [type]=\"'text'\"\n        [placeholder]=\"option.placeholders[0] || ''\"\n        [(ngModel)]=\"option.defaultValue[0]\"\n        (ngModelChange)=\"inputChange('text', $event)\"\n      >\n    </div>\n  </div>\n</ng-template>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    ModalServiceComponent.ctorParameters = function () { return [
        { type: ModalOptions }
    ]; };
    return ModalServiceComponent;
}(ModalComponent));
export { ModalServiceComponent };
if (false) {
    /** @type {?} */
    ModalServiceComponent.prototype.option;
}
