// ng-zorro-antd-mobile/esm5/slider/slider-handle/slider-handle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, Output, EventEmitter, HostListener, ViewEncapsulation } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var SliderHandleComponent = /** @class */ (function () {
    function SliderHandleComponent(_elf, _sanitizer) {
        var _this = this;
        this._elf = _elf;
        this._sanitizer = _sanitizer;
        this._disabled = false;
        this._marks = {};
        this._isDraging = false;
        this.onChange = new EventEmitter();
        this.onAfterChange = new EventEmitter();
        this.mouseDown = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (!_this._disabled && _this.isMouseTarget(event)) {
                _this._startX = event.clientX;
                _this._handleStatus = 'start';
                _this._isDraging = true;
                document.addEventListener('mousemove', _this.mouseMove, false);
                document.addEventListener('mouseup', _this.mouseUp, false);
                _this.pauseEvent(event);
            }
        });
        this.mouseMove = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (!_this._disabled && _this._isDraging) {
                _this.pauseEvent(event);
                /** @type {?} */
                var pos = event.clientX;
                _this._value = Math.round(_this.calcValueByPos(pos));
                _this.left = _this.calcOffset(_this._value);
                if (_this._oldValue !== _this._value) {
                    _this._oldValue = _this._value;
                    _this.onChange.emit(_this._value);
                }
            }
        });
        this.mouseUp = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (!_this._disabled && _this._isDraging) {
                _this._handleStatus = 'end';
                _this._isDraging = false;
                /** @type {?} */
                var pos = event.clientX;
                _this._value = Math.round(_this.calcValueByPos(pos));
                _this.left = _this.calcOffset(_this._value);
                _this.onAfterChange.emit(_this._value);
            }
        });
    }
    Object.defineProperty(SliderHandleComponent.prototype, "min", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._min = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "max", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._max = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "minBound", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._minBound = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "maxBound", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._maxBound = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "step", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._step = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "value", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._value = value;
            if (this._value) {
                this.left = this.calcOffset(this._value);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "disabled", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "sliderLength", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._sliderLength = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "sliderStart", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._sliderStart = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderHandleComponent.prototype, "handleStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._handleStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._handleStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    /* 手势操作 */
    /* 手势操作 */
    /**
     * @param {?} event
     * @return {?}
     */
    SliderHandleComponent.prototype.panstart = /* 手势操作 */
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // event.preventDefault();
        if (!this._disabled) {
            this._startX = event && event.changedTouches && event.changedTouches[0] && event.changedTouches[0].clientX;
            this._handleStatus = 'start';
            this._isDraging = true;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    SliderHandleComponent.prototype.panmove = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (!this._disabled && this._isDraging) {
            /** @type {?} */
            var pos = event.changedTouches[0].clientX;
            this._value = Math.round(this.calcValueByPos(pos));
            this.left = this.calcOffset(this._value);
            if (this._oldValue !== this._value) {
                this._oldValue = this._value;
                this.onChange.emit(this._value);
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    SliderHandleComponent.prototype.panend = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (!this._disabled && this._isDraging) {
            this._handleStatus = 'end';
            this._isDraging = false;
            /** @type {?} */
            var pos = event.changedTouches[0].clientX;
            this._value = Math.round(this.calcValueByPos(pos));
            this.left = this.calcOffset(this._value);
            this.onAfterChange.emit(this._value);
        }
    };
    /**
     * @param {?} pos
     * @return {?}
     */
    SliderHandleComponent.prototype.calcValueByPos = /**
     * @param {?} pos
     * @return {?}
     */
    function (pos) {
        /** @type {?} */
        var offset = pos - this._sliderStart;
        /** @type {?} */
        var value = this.calcValue(offset);
        if (value <= this._minBound) {
            value = this._minBound;
        }
        if (value >= this._maxBound) {
            value = this._maxBound;
        }
        /** @type {?} */
        var closestPoint = this.getClosestPoint(value);
        return this._step === null ? closestPoint : parseFloat(closestPoint.toFixed(this.getPrecision(this._step)));
    };
    /**
     * @param {?} offset
     * @return {?}
     */
    SliderHandleComponent.prototype.calcValue = /**
     * @param {?} offset
     * @return {?}
     */
    function (offset) {
        /** @type {?} */
        var ratio = Math.abs(Math.max(offset, 0) / this._sliderLength);
        /** @type {?} */
        var value = ratio * (this._max - this._min) + this._min;
        return value;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    SliderHandleComponent.prototype.getClosestPoint = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var points = Object.keys(this._marks).map(parseFloat);
        if (this._step !== null) {
            /** @type {?} */
            var closestStep = Math.round((val - this._min) / this._step) * this._step + this._min;
            points.push(closestStep);
        }
        /** @type {?} */
        var diffs = points.map((/**
         * @param {?} point
         * @return {?}
         */
        function (point) {
            return Math.abs(val - point);
        }));
        return points[diffs.indexOf(Math.min.apply(Math, this.toConsumableArray(diffs)))];
    };
    /**
     * @param {?} step
     * @return {?}
     */
    SliderHandleComponent.prototype.getPrecision = /**
     * @param {?} step
     * @return {?}
     */
    function (step) {
        /** @type {?} */
        var stepString = step.toString();
        /** @type {?} */
        var precision = 0;
        if (stepString.indexOf('.') >= 0) {
            precision = stepString.length - stepString.indexOf('.') - 1;
        }
        return precision;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    SliderHandleComponent.prototype.calcOffset = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var ratio = (value - this._min) / (this._max - this._min);
        return ratio * 100;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SliderHandleComponent.prototype.pauseEvent = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        e.preventDefault();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    SliderHandleComponent.prototype.isMouseTarget = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var target = event.target;
        /** @type {?} */
        var parentFound = false;
        while (target !== null && !parentFound) {
            if (target === this._elf.nativeElement) {
                parentFound = true;
            }
            target = target.parentElement;
        }
        return parentFound;
    };
    /**
     * @param {?} arr
     * @return {?}
     */
    SliderHandleComponent.prototype.toConsumableArray = /**
     * @param {?} arr
     * @return {?}
     */
    function (arr) {
        if (Array.isArray(arr)) {
            /** @type {?} */
            var arr2 = Array(arr.length);
            for (var i = 0; i < arr.length; i++) {
                arr2[i] = arr[i];
            }
            return arr2;
        }
    };
    /**
     * @return {?}
     */
    SliderHandleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        this._elf.nativeElement.addEventListener('mousedown', this.mouseDown, false);
        this._handleOffsetX = this._elf.nativeElement.getBoundingClientRect().x;
        this.left = this.calcOffset(this._value);
        this._minBound = this._minBound === undefined ? this._min : this._minBound;
        this._maxBound = this._maxBound === undefined ? this._max : this._maxBound;
    };
    /**
     * @return {?}
     */
    SliderHandleComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        document.removeEventListener('mousemove', this.mouseMove, false);
        document.removeEventListener('mouseup', this.mouseUp, false);
    };
    SliderHandleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'SliderHandle, nzm-slider-handle',
                    template: "<div role=\"slider\"\n     class=\"am-slider-handle\"\n     [ngStyle]=\"handleStyle\"\n     [style.left.%]=\"left\">\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SliderHandleComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DomSanitizer }
    ]; };
    SliderHandleComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        minBound: [{ type: Input }],
        maxBound: [{ type: Input }],
        step: [{ type: Input }],
        value: [{ type: Input }],
        disabled: [{ type: Input }],
        sliderLength: [{ type: Input }],
        sliderStart: [{ type: Input }],
        handleStyle: [{ type: Input }],
        onChange: [{ type: Output }],
        onAfterChange: [{ type: Output }],
        panstart: [{ type: HostListener, args: ['touchstart', ['$event'],] }],
        panmove: [{ type: HostListener, args: ['touchmove', ['$event'],] }],
        panend: [{ type: HostListener, args: ['touchend', ['$event'],] }]
    };
    return SliderHandleComponent;
}());
export { SliderHandleComponent };
if (false) {
    /** @type {?} */
    SliderHandleComponent.prototype.left;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._min;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._max;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._step;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._marks;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._handleStyle;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._sliderLength;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._sliderStart;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._minBound;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._maxBound;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._startX;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._isDraging;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._handleStatus;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._handleOffsetX;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._oldValue;
    /** @type {?} */
    SliderHandleComponent.prototype.onChange;
    /** @type {?} */
    SliderHandleComponent.prototype.onAfterChange;
    /** @type {?} */
    SliderHandleComponent.prototype.mouseDown;
    /** @type {?} */
    SliderHandleComponent.prototype.mouseMove;
    /** @type {?} */
    SliderHandleComponent.prototype.mouseUp;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._elf;
    /**
     * @type {?}
     * @private
     */
    SliderHandleComponent.prototype._sanitizer;
}
