// ng-zorro-antd-mobile/esm5/card/card-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding, TemplateRef } from '@angular/core';
var CardHeaderComponent = /** @class */ (function () {
    function CardHeaderComponent() {
        this.prefixCls = 'am-card-header';
        this.thumb = null;
        this.thumbStyle = null;
        this.title = null;
        this.extra = null;
        this.cardBodyWrapper = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    CardHeaderComponent.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof TemplateRef;
    };
    CardHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CardHeader, nzm-card-header',
                    template: "<div class=\"{{prefixCls}}-content\">\n  <img *ngIf='thumb && !isTemplateRef(thumb)' src=\"{{thumb}}\" [ngStyle]=\"thumbStyle\">\n  <ng-container *ngIf=\"thumb && isTemplateRef(thumb)\" [ngTemplateOutlet]=\"thumb\"></ng-container>\n  <ng-container *ngIf=\"!isTemplateRef(title);else titleTemplate\">{{title}}</ng-container>\n</div>\n<div *ngIf=\"extra\" class=\"{{prefixCls}}-extra\">\n  <ng-container *ngIf=\"!isTemplateRef(extra);else extraTemplate\">{{extra}}</ng-container>\n</div>\n<ng-template #titleTemplate>\n  <ng-template [ngTemplateOutlet]=\"title\"></ng-template>\n</ng-template>\n<ng-template #extraTemplate>\n  <ng-template [ngTemplateOutlet]=\"extra\"></ng-template>\n</ng-template>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    CardHeaderComponent.ctorParameters = function () { return []; };
    CardHeaderComponent.propDecorators = {
        thumb: [{ type: Input }],
        thumbStyle: [{ type: Input }],
        title: [{ type: Input }],
        extra: [{ type: Input }],
        cardBodyWrapper: [{ type: HostBinding, args: ['class.am-card-header',] }]
    };
    return CardHeaderComponent;
}());
export { CardHeaderComponent };
if (false) {
    /** @type {?} */
    CardHeaderComponent.prototype.prefixCls;
    /** @type {?} */
    CardHeaderComponent.prototype.thumb;
    /** @type {?} */
    CardHeaderComponent.prototype.thumbStyle;
    /** @type {?} */
    CardHeaderComponent.prototype.title;
    /** @type {?} */
    CardHeaderComponent.prototype.extra;
    /** @type {?} */
    CardHeaderComponent.prototype.cardBodyWrapper;
}
