// ng-zorro-antd-mobile/esm2015/tab-bar/tab-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, ContentChildren, QueryList, HostBinding, EventEmitter } from '@angular/core';
import { TabBarItemComponent } from './tab-bar-item.component';
/**
 * @record
 */
export function TabBarOnPressEvent() { }
if (false) {
    /** @type {?} */
    TabBarOnPressEvent.prototype.index;
    /** @type {?} */
    TabBarOnPressEvent.prototype.title;
    /** @type {?} */
    TabBarOnPressEvent.prototype.key;
}
export class TabBarComponent {
    constructor() {
        this.prefixCls = 'am-tab-bar';
        this._activeTab = 0;
        this._tintColor = '#108ee9';
        this._unselectedTintColor = '#888';
        this.hidden = false;
        this.prerenderingSiblingsNumber = -1;
        this.barTintColor = 'white';
        this.tabBarPosition = 'bottom';
        this.onPress = new EventEmitter();
        this.tabBar = true;
    }
    /**
     * @return {?}
     */
    get activeTab() {
        return this._activeTab;
    }
    /**
     * @param {?} tab
     * @return {?}
     */
    set activeTab(tab) {
        this._activeTab = tab;
        if (this.tabBarItems && this.tabBarItems.length > 0) {
            this.selectTabBarItem(tab);
        }
    }
    /**
     * @return {?}
     */
    get tintColor() {
        return this._tintColor;
    }
    /**
     * @param {?} color
     * @return {?}
     */
    set tintColor(color) {
        this._tintColor = color;
        if (this.tabBarItems && this.tabBarItems.length > 0) {
            this.tabBarItems.forEach((/**
             * @param {?} tabBarItem
             * @return {?}
             */
            (tabBarItem) => {
                tabBarItem.tintColor = this._tintColor;
            }));
        }
    }
    /**
     * @return {?}
     */
    get unselectedTintColor() {
        return this._unselectedTintColor;
    }
    /**
     * @param {?} color
     * @return {?}
     */
    set unselectedTintColor(color) {
        this._unselectedTintColor = color;
        if (this.tabBarItems && this.tabBarItems.length > 0) {
            this.tabBarItems.forEach((/**
             * @param {?} tabBarItem
             * @return {?}
             */
            (tabBarItem) => {
                tabBarItem.unselectedTintColor = this._unselectedTintColor;
            }));
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    selectTabBarItem(index) {
        if (this.tabBarItems && this.tabBarItems.length > 0) {
            this.tabBarItems.forEach((/**
             * @param {?} tabBarItem
             * @return {?}
             */
            (tabBarItem) => {
                tabBarItem.selected = false;
            }));
            this.tabBarItems.toArray()[index].selected = true;
        }
    }
    /**
     * @param {?} pressParam
     * @return {?}
     */
    tabBarTabOnPress(pressParam) {
        this.onPress.emit(pressParam);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this.tabBarItems && this.tabBarItems.length > 0) {
            this.tabBarItems.forEach((/**
             * @param {?} tabBarItem
             * @return {?}
             */
            (tabBarItem) => {
                tabBarItem.tintColor = this._tintColor;
                tabBarItem.unselectedTintColor = this._unselectedTintColor;
            }));
        }
        this.selectTabBarItem(this.activeTab);
    }
}
TabBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'TabBar, nzm-tab-bar',
                template: "<Tabs [activeTab]=\"activeTab\"\n      [swipeable]=\"false\"\n      [useOnPan]=\"false\"\n      [animated]=\"false\"\n      [tabDirection]=\"'horizontal'\"\n      [tabBarPosition]='tabBarPosition'\n      [renderTabBar]=\"TabBarBar\"\n      [prerenderingSiblingsNumber]=\"prerenderingSiblingsNumber\"\n      [tabPanesContent]=\"tabBarItems\"\n></Tabs>\n\n<ng-template #TabBarBar>\n  <div class=\"am-tabs-tab-bar-wrap\">\n    <div class=\"{{prefixCls}}-bar\"\n         [ngClass]=\"{'am-tab-bar-bar-hidden-top': 'top' === tabBarPosition && hidden, 'am-tab-bar-bar-hidden-bottom': 'bottom' === tabBarPosition && hidden}\"\n         [style.background-color]=\"barTintColor\">\n      <div class=\"am-tab-bar-tab\" *ngFor=\"let tabBarItem of tabBarItems; let i = index;\" (click)=\"tabBarTabOnPress({'index': i, 'key': tabBarItem.key, 'title': tabBarItem.title})\">\n        <ng-container [ngTemplateOutlet]=\"tabBarItem.tabBarTab\"></ng-container>\n      </div>\n    </div>\n  </div>\n</ng-template>"
            }] }
];
/** @nocollapse */
TabBarComponent.ctorParameters = () => [];
TabBarComponent.propDecorators = {
    tabBarItems: [{ type: ContentChildren, args: [TabBarItemComponent, { descendants: true },] }],
    hidden: [{ type: Input }],
    prerenderingSiblingsNumber: [{ type: Input }],
    activeTab: [{ type: Input }],
    barTintColor: [{ type: Input }],
    tabBarPosition: [{ type: Input }],
    tintColor: [{ type: Input }],
    unselectedTintColor: [{ type: Input }],
    onPress: [{ type: Output }],
    tabBar: [{ type: HostBinding, args: ['class.am-tab-bar',] }]
};
if (false) {
    /** @type {?} */
    TabBarComponent.prototype.prefixCls;
    /**
     * @type {?}
     * @private
     */
    TabBarComponent.prototype._activeTab;
    /**
     * @type {?}
     * @private
     */
    TabBarComponent.prototype._tintColor;
    /**
     * @type {?}
     * @private
     */
    TabBarComponent.prototype._unselectedTintColor;
    /** @type {?} */
    TabBarComponent.prototype.tabBarItems;
    /** @type {?} */
    TabBarComponent.prototype.hidden;
    /** @type {?} */
    TabBarComponent.prototype.prerenderingSiblingsNumber;
    /** @type {?} */
    TabBarComponent.prototype.barTintColor;
    /** @type {?} */
    TabBarComponent.prototype.tabBarPosition;
    /** @type {?} */
    TabBarComponent.prototype.onPress;
    /** @type {?} */
    TabBarComponent.prototype.tabBar;
}
