// ng-zorro-antd-mobile/esm5/radio/radio-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectorRef, Input, ChangeDetectionStrategy, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
var RadioItemComponent = /** @class */ (function () {
    function RadioItemComponent(cdr) {
        this.cdr = cdr;
        this.select$ = new Subject();
        this.prefixCls = 'am-radio';
        this._checked = false;
        this._disabled = false;
        this.wrap = false;
        this.error = false;
        this.multipleLine = false;
        this.platform = 'ios';
        this.align = 'middle';
    }
    Object.defineProperty(RadioItemComponent.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            return this._checked;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._checked = value;
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioItemComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    RadioItemComponent.prototype.onRadioItemClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) { };
    /**
     * @param {?} event
     * @return {?}
     */
    RadioItemComponent.prototype.change = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.disabled && !this.checked) {
            this.select$.next(this);
        }
    };
    /**
     * @return {?}
     */
    RadioItemComponent.prototype.markForCheck = /**
     * @return {?}
     */
    function () {
        this.cdr.markForCheck();
    };
    RadioItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'RadioItem, nzm-radio-item',
                    template: "<ListItem [className]=\"'am-radio-item ' + (disabled ? 'am-radio-item-disabled' : '')\"\n          [wrap]=\"wrap\"\n          [align]=\"align\"\n          [arrow]=\"arrow\"\n          [error]=\"error\"\n          [extra]=\"radio\"\n          [thumb]=\"thumb\"\n          [disabled]=\"disabled\"\n          [platform]=\"platform\"\n          [multipleLine]=\"multipleLine\"\n          [disabled]=\"disabled\"\n          [wrap]=\"wrap\"\n          (onClick)=\"onRadioItemClick($event)\"\n>\n  <ng-content></ng-content>\n</ListItem>\n<ng-template #radio>\n  <label Radio\n         [name]=\"name\"\n         [value]=\"value\"\n         [checked]=\"checked\"\n         [disabled]=\"disabled\"\n         (onChange)=\"change($event)\"\n  ></label>\n</ng-template>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    RadioItemComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    RadioItemComponent.propDecorators = {
        name: [{ type: Input }],
        value: [{ type: Input }],
        arrow: [{ type: Input }],
        thumb: [{ type: Input }],
        wrap: [{ type: Input }],
        error: [{ type: Input }],
        multipleLine: [{ type: Input }],
        platform: [{ type: Input }],
        align: [{ type: Input }],
        disabled: [{ type: Input }]
    };
    return RadioItemComponent;
}());
export { RadioItemComponent };
if (false) {
    /** @type {?} */
    RadioItemComponent.prototype.select$;
    /** @type {?} */
    RadioItemComponent.prototype.prefixCls;
    /**
     * @type {?}
     * @private
     */
    RadioItemComponent.prototype._checked;
    /**
     * @type {?}
     * @private
     */
    RadioItemComponent.prototype._disabled;
    /** @type {?} */
    RadioItemComponent.prototype.name;
    /** @type {?} */
    RadioItemComponent.prototype.value;
    /** @type {?} */
    RadioItemComponent.prototype.arrow;
    /** @type {?} */
    RadioItemComponent.prototype.thumb;
    /** @type {?} */
    RadioItemComponent.prototype.wrap;
    /** @type {?} */
    RadioItemComponent.prototype.error;
    /** @type {?} */
    RadioItemComponent.prototype.multipleLine;
    /** @type {?} */
    RadioItemComponent.prototype.platform;
    /** @type {?} */
    RadioItemComponent.prototype.align;
    /**
     * @type {?}
     * @private
     */
    RadioItemComponent.prototype.cdr;
}
