// ng-zorro-antd-mobile/esm5/tag/tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, EventEmitter } from '@angular/core';
var TagComponent = /** @class */ (function () {
    function TagComponent() {
        this.prefixCls = 'am-tag';
        this.closed = false;
        this.wrapCls = {};
        this._small = false;
        this._closable = false;
        this._selected = false;
        this._disabled = false;
        this.onChange = new EventEmitter();
        this.onClose = new EventEmitter();
        this.afterClose = new EventEmitter();
    }
    Object.defineProperty(TagComponent.prototype, "small", {
        get: /**
         * @return {?}
         */
        function () {
            return this._small;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._small = v;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TagComponent.prototype, "closable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._closable;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._closable = v;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TagComponent.prototype, "selected", {
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._selected = v;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TagComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._disabled = v;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TagComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (this._disabled) {
            return;
        }
        this._selected = !this._selected;
        this.onChange.emit(this._selected);
        this.setClassMap();
    };
    /**
     * @return {?}
     */
    TagComponent.prototype.onTagClose = /**
     * @return {?}
     */
    function () {
        this.onClose.emit();
        this.closed = true;
        this.afterClose.emit();
    };
    /**
     * @return {?}
     */
    TagComponent.prototype.setClassMap = /**
     * @return {?}
     */
    function () {
        var _a;
        this.wrapCls = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-normal"] = !this._disabled && (!this._selected || this._small || this._closable),
            _a[this.prefixCls + "-small"] = this._small,
            _a[this.prefixCls + "-active"] = this._selected && !this._disabled && !this._small && !this._closable,
            _a[this.prefixCls + "-disabled"] = this._disabled,
            _a[this.prefixCls + "-closable"] = this._closable,
            _a);
    };
    /**
     * @return {?}
     */
    TagComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClassMap();
    };
    TagComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Tag, nzm-tag',
                    template: "<div *ngIf=\"!closed\" [ngClass]=\"wrapCls\" (click)=\"onClick()\">\n  <div class=\"{{prefixCls}}-text\">\n    <ng-content></ng-content>\n  </div>\n  <div *ngIf=\"closable && !disabled && !small\"\n       role=\"button\"\n       class=\"{{prefixCls}}-close\"\n       aria-label=\"remove tag\"\n       (click)=\"onTagClose()\">\n    <Icon aria-hidden=\"true\" [type]=\"'cross-circle'\" [size]=\"'xs'\"></Icon>\n  </div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    TagComponent.ctorParameters = function () { return []; };
    TagComponent.propDecorators = {
        small: [{ type: Input }],
        closable: [{ type: Input }],
        selected: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        onClose: [{ type: Output }],
        afterClose: [{ type: Output }]
    };
    return TagComponent;
}());
export { TagComponent };
if (false) {
    /** @type {?} */
    TagComponent.prototype.prefixCls;
    /** @type {?} */
    TagComponent.prototype.closed;
    /** @type {?} */
    TagComponent.prototype.wrapCls;
    /**
     * @type {?}
     * @private
     */
    TagComponent.prototype._small;
    /**
     * @type {?}
     * @private
     */
    TagComponent.prototype._closable;
    /**
     * @type {?}
     * @private
     */
    TagComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    TagComponent.prototype._disabled;
    /** @type {?} */
    TagComponent.prototype.onChange;
    /** @type {?} */
    TagComponent.prototype.onClose;
    /** @type {?} */
    TagComponent.prototype.afterClose;
}
