// ng-zorro-antd-mobile/esm2015/date-picker/date-picker-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
/**
 * @record
 */
export function DatePickerOptionsInterface() { }
if (false) {
    /** @type {?} */
    DatePickerOptionsInterface.prototype.mode;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.value;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.minDate;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.maxDate;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.use12Hours;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.minuteStep;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.locale;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.disabled;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.mask;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.title;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.okText;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.dismissText;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.appendToBody;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.showErrorToast;
    /** @type {?} */
    DatePickerOptionsInterface.prototype.showErrorToastInterval;
}
export class DatePickerOptions {
    constructor() {
        this.mode = 'date';
        this.value = new Date();
        this.minDate = new Date(2000, 6, 1, 0, 0, 0);
        this.maxDate = new Date(2030, 1, 1, 23, 59, 59);
        this.use12Hours = false;
        this.minuteStep = 1;
        this.data = [];
        this.mask = true;
        this.title = '';
        this.okText = '确定';
        this.dismissText = '取消';
        this.disabled = false;
        this.appendToBody = false;
        this.showErrorToast = true;
        this.showErrorToastInterval = 2000;
        this.onOk = new EventEmitter();
        this.onDismiss = new EventEmitter();
        this.onValueChange = new EventEmitter();
        this.onChange = new EventEmitter();
    }
}
if (false) {
    /** @type {?} */
    DatePickerOptions.prototype.mode;
    /** @type {?} */
    DatePickerOptions.prototype.value;
    /** @type {?} */
    DatePickerOptions.prototype.minDate;
    /** @type {?} */
    DatePickerOptions.prototype.maxDate;
    /** @type {?} */
    DatePickerOptions.prototype.use12Hours;
    /** @type {?} */
    DatePickerOptions.prototype.minuteStep;
    /** @type {?} */
    DatePickerOptions.prototype.data;
    /** @type {?} */
    DatePickerOptions.prototype.mask;
    /** @type {?} */
    DatePickerOptions.prototype.title;
    /** @type {?} */
    DatePickerOptions.prototype.okText;
    /** @type {?} */
    DatePickerOptions.prototype.dismissText;
    /** @type {?} */
    DatePickerOptions.prototype.disabled;
    /** @type {?} */
    DatePickerOptions.prototype.locale;
    /** @type {?} */
    DatePickerOptions.prototype.appendToBody;
    /** @type {?} */
    DatePickerOptions.prototype.showErrorToast;
    /** @type {?} */
    DatePickerOptions.prototype.showErrorToastInterval;
    /** @type {?} */
    DatePickerOptions.prototype.onOk;
    /** @type {?} */
    DatePickerOptions.prototype.onDismiss;
    /** @type {?} */
    DatePickerOptions.prototype.onValueChange;
    /** @type {?} */
    DatePickerOptions.prototype.onChange;
    /** @type {?} */
    DatePickerOptions.prototype.hidePicker;
    /** @type {?} */
    DatePickerOptions.prototype.updateNgModel;
}
