// ng-zorro-antd-mobile/esm2015/card/card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
export class CardComponent {
    constructor() {
        this.full = false;
        this.cardWrapper = true;
    }
    /**
     * @return {?}
     */
    get cardFull() {
        return this.full;
    }
}
CardComponent.decorators = [
    { type: Component, args: [{
                selector: 'Card, nzm-card',
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
CardComponent.ctorParameters = () => [];
CardComponent.propDecorators = {
    full: [{ type: Input }],
    cardWrapper: [{ type: HostBinding, args: ['class.am-card',] }],
    cardFull: [{ type: HostBinding, args: ['class.am-card-full',] }]
};
if (false) {
    /** @type {?} */
    CardComponent.prototype.full;
    /** @type {?} */
    CardComponent.prototype.cardWrapper;
}
