// ng-zorro-antd-mobile/esm2015/picker/picker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { PickerComponent } from './picker.component';
import { PickerOptions } from './picker-options.provider';
import { PopupService } from '../core/services/popup.service';
export class PickerService extends PopupService {
    /**
     * @param {?=} config
     * @param {?=} confirmCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    static showPicker(config = PickerService.defaultOptions, confirmCallback, cancelCallback) {
        /** @type {?} */
        const options = new PickerOptions();
        Object.assign(options, config, {
            hidePicker: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                this.hidePicker();
            }),
            confirm: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (confirmCallback) {
                    confirmCallback(event);
                }
            }),
            cancel: (/**
             * @return {?}
             */
            () => {
                if (cancelCallback) {
                    cancelCallback();
                }
            })
        });
        /** @type {?} */
        const childInjector = Injector.create([
            {
                provide: PickerOptions,
                useValue: options
            }
        ]);
        PickerService.showPopup('PickerComponent', PickerComponent, childInjector);
    }
    /**
     * @return {?}
     */
    static hidePicker() {
        PickerService.hidePopup('PickerComponent');
    }
}
PickerService.defaultOptions = new PickerOptions();
PickerService.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    PickerService.defaultOptions;
}
