// ng-zorro-antd-mobile/esm2015/action-sheet/action-sheet.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, TemplateRef, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { LocaleProviderService } from '../locale-provider/locale-provider.service';
export class ActionSheetComponent {
    /**
     * @param {?} localeProviderService
     */
    constructor(localeProviderService) {
        this.localeProviderService = localeProviderService;
        this.unsubscribe$ = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.localeProvider();
    }
    /**
     * @return {?}
     */
    localeProvider() {
        /** @type {?} */
        const self = this;
        if (self.option.locale || self.option.locale !== undefined) {
            self.localeProviderService.setLocale(self.option.locale);
        }
        self.localeProviderService.localeChange.pipe(takeUntil(self.unsubscribe$)).subscribe((/**
         * @param {?} _
         * @return {?}
         */
        _ => {
            if (self.option.cancelButtonText) {
                self.option.cancelButtonText = self.localeProviderService.getLocaleSubObj('ActionSheet')['dismissText'];
            }
        }));
    }
    /**
     * @param {?} index
     * @param {?=} rowIndex
     * @param {?=} event
     * @return {?}
     */
    onPress(index, rowIndex = 0, event) { }
    /**
     * @param {?} option
     * @return {?}
     */
    showShare(option) {
        /** @type {?} */
        const cls = { [`${option.prefixCls}-share`]: option.flag === 'SHARE' };
        return cls;
    }
    /**
     * @param {?} option
     * @param {?} suffix
     * @return {?}
     */
    setActiveClassName(option, suffix) {
        return [`${option.prefixCls}-${suffix}-active`];
    }
    /**
     * @param {?} value
     * @return {?}
     */
    isNoTitle(value) {
        return value === '' || value === null || value === undefined;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    isTemplateRef(value) {
        if (value) {
            return value instanceof TemplateRef;
        }
        return false;
    }
    /**
     * @param {?} options
     * @param {?} value
     * @return {?}
     */
    isArray(options, value) {
        if (options.length > 0 && value) {
            return value instanceof Array;
        }
        return false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
}
ActionSheetComponent.decorators = [
    { type: Component, args: [{
                selector: 'ActionSheet',
                template: "<div class=\"{{option.prefixCls}}-mask {{option.maskTransitionName}}\"></div>\n<div role=\"dialog\" class=\"{{option.prefixCls}}-wrap {{option.transitionName}}\" (click)=\"option.maskClose(-1, 0, $event)\">\n  <div role=\"document\"\n       class=\"{{option.prefixCls}}\"\n       [ngClass]=\"showShare(option)\">\n    <div class=\"{{option.prefixCls}}-content\">\n      <button aria-label=\"Close\" class=\"{{option.prefixCls}}-close\">\n        <span class=\"{{option.prefixCls}}-close-x\"></span>\n      </button>\n      <div class=\"{{option.prefixCls}}-body\">\n        <div>\n          <ng-container *ngIf=\"!isNoTitle(option.title)\">\n            <ng-template *ngIf=\"isTemplateRef(option.title)\" [ngTemplateOutlet]=\"option.title\"></ng-template>\n            <h3 *ngIf=\"!isTemplateRef(option.title)\" class=\"{{option.prefixCls}}-title\">{{option.title}}</h3>\n          </ng-container>\n          <ng-container *ngIf=\"!isNoTitle(option.message)\">\n            <ng-template *ngIf=\"isTemplateRef(option.message)\" [ngTemplateOutlet]=\"option.message\"></ng-template>\n            <div *ngIf=\"!isTemplateRef(option.message)\" class=\"{{option.prefixCls}}-message\">{{option.message}}</div>\n          </ng-container>\n          <ng-container [ngSwitch]=\"option.flag\">\n            <div *ngSwitchCase=\"'NORMAL'\" class=\"{{option.prefixCls}}-button-list\" role=\"group\">\n              <ng-container *ngFor=\"let item of option.options; let i = index;\">\n                  <div TouchFeedbackDirective class=\"{{option.prefixCls}}-button-list-item\" [className]=\"setActiveClassName(option, 'button-list-item')\">\n                      <div *ngIf=\"option.destructiveButtonIndex !== i && option.cancelButtonIndex !== i\"\n                           class=\"{{option.prefixCls}}-button-list-item\"\n                           (click)=\"option.onPress(i, 0, $event)\"\n                      >\n                        {{item}}\n                      </div>\n                      <div *ngIf=\"option.destructiveButtonIndex === i\"\n                           class=\"{{option.prefixCls}}-button-list-item {{option.prefixCls}}-destructive-button\"\n                           (click)=\"option.onPress(i, 0, $event)\"\n                      >\n                        {{item}}\n                      </div>\n                      <div *ngIf=\"option.cancelButtonIndex === i\"\n                           class=\"{{option.prefixCls}}-button-list-item {{option.prefixCls}}-cancel-button\"\n                           (click)=\"option.onPress(i, 0, $event)\"\n                      >\n                        {{item}}\n                        <span class=\"{{option.prefixCls}}-cancel-button-mask\"></span>\n                      </div>\n                  </div>\n              </ng-container>\n            </div>\n            <div *ngSwitchCase=\"'SHARE'\" class=\"{{option.prefixCls}}-share {{option.prefixCls}}-share-content\">\n              <div *ngIf=\"!isArray(option.options, option.options[0])\" class=\"{{option.prefixCls}}-share-list\">\n                <ng-container *ngFor=\"let item of option.options; let i = index;\">\n                  <div class=\"{{option.prefixCls}}-share-list-item\" (click)=\"option.onPress(i, 0, $event)\">\n                    <div class=\"{{option.prefixCls}}-share-list-item-icon\">\n                      <ng-template *ngIf=\"isTemplateRef(item.icon)\" [ngTemplateOutlet]=\"item.icon\"></ng-template>\n                      <div *ngIf=\"!isTemplateRef(item.icon)\" [innerHTML]=\"item.icon | safeHTML\"></div>\n                    </div>\n                    <div class=\"{{option.prefixCls}}-share-list-item-title\">{{item.title}}</div>\n                  </div>\n                </ng-container>\n              </div>\n              <ng-container *ngIf=\"isArray(option.options, option.options[0])\">\n                <div *ngFor=\"let items of option.options; let rowIndex = index\" class=\"{{option.prefixCls}}-share-list\">\n                  <ng-container *ngFor=\"let item of items; let i = index;\">\n                    <div class=\"{{option.prefixCls}}-share-list-item\" (click)=\"option.onPress(i, rowIndex, $event)\">\n                      <div class=\"{{option.prefixCls}}-share-list-item-icon\">\n                        <ng-template *ngIf=\"isTemplateRef(item.icon)\" [ngTemplateOutlet]=\"item.icon\"></ng-template>\n                        <div *ngIf=\"!isTemplateRef(item.icon)\" [innerHTML]=\"item.icon | safeHTML\"></div>\n                      </div>\n                      <div class=\"{{option.prefixCls}}-share-list-item-title\">{{item.title}}</div>\n                    </div>\n                  </ng-container>\n                </div>\n              </ng-container>\n              <div TouchFeedbackDirective [className]=\"setActiveClassName(option, 'share-cancel-button')\" class=\"{{option.prefixCls}}-share-cancel-button\">{{option.cancelButtonText}}</div>\n            </div>\n          </ng-container>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
ActionSheetComponent.ctorParameters = () => [
    { type: LocaleProviderService }
];
if (false) {
    /** @type {?} */
    ActionSheetComponent.prototype.unsubscribe$;
    /** @type {?} */
    ActionSheetComponent.prototype.option;
    /**
     * @type {?}
     * @private
     */
    ActionSheetComponent.prototype.localeProviderService;
}
