// ng-zorro-antd-mobile/esm2015/action-sheet/action-sheet-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class ActionSheetOptions {
    constructor() {
        this.prefixCls = 'am-action-sheet';
        this.maskClosable = true;
        this.transitionName = 'am-slide-up';
        this.maskTransitionName = 'am-fade';
    }
}
ActionSheetOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    ActionSheetOptions.prototype.prefixCls;
    /** @type {?} */
    ActionSheetOptions.prototype.maskClosable;
    /** @type {?} */
    ActionSheetOptions.prototype.cancelButtonIndex;
    /** @type {?} */
    ActionSheetOptions.prototype.destructiveButtonIndex;
    /** @type {?} */
    ActionSheetOptions.prototype.title;
    /** @type {?} */
    ActionSheetOptions.prototype.message;
    /** @type {?} */
    ActionSheetOptions.prototype.className;
    /** @type {?} */
    ActionSheetOptions.prototype.transitionName;
    /** @type {?} */
    ActionSheetOptions.prototype.maskTransitionName;
}
export class ShareOption {
}
ShareOption.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    ShareOption.prototype.icon;
    /** @type {?} */
    ShareOption.prototype.title;
}
export class ShareActionSheetWithOptions extends ActionSheetOptions {
    constructor() {
        super(...arguments);
        this.cancelButtonText = 'Cancel';
    }
}
ShareActionSheetWithOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.options;
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.cancelButtonText;
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.locale;
}
export class ActionSheetWithOptions extends ActionSheetOptions {
}
ActionSheetWithOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    ActionSheetWithOptions.prototype.options;
}
