// ng-zorro-antd-mobile/esm5/toast/toast.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, TemplateRef, NgZone } from '@angular/core';
var ToastComponent = /** @class */ (function () {
    function ToastComponent(_zone) {
        this._zone = _zone;
        this.prefixCls = 'am-toast';
        this.isContentString = true;
        this.transitionName = 'am-fade-enter am-fade-enter-active';
        this._iconType = '';
        this._content = '';
        this.mask = true;
        this.position = 'middle';
    }
    Object.defineProperty(ToastComponent.prototype, "content", {
        get: /**
         * @return {?}
         */
        function () {
            return this._content;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            if (value instanceof TemplateRef) {
                this.isContentString = false;
            }
            else {
                this.isContentString = true;
            }
            this._zone.run((/**
             * @return {?}
             */
            function () {
                _this._content = value;
            }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastComponent.prototype, "iconType", {
        get: /**
         * @return {?}
         */
        function () {
            return this._iconType;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            this._zone.run((/**
             * @return {?}
             */
            function () {
                _this._iconType = value;
            }));
        },
        enumerable: true,
        configurable: true
    });
    ToastComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Toast',
                    encapsulation: ViewEncapsulation.None,
                    template: "<div class=\"{{prefixCls}}-notice {{prefixCls}}-notice-closable {{transitionName}}\">\n  <div class=\"{{prefixCls}}-notice-content\">\n    <div role=\"alert\"\n         *ngIf=\"iconType\"\n         class=\"{{prefixCls}}-text {{prefixCls}}-text-icon\"\n         aria-live=\"assertive\"\n    >\n      <Icon [type]=\"iconType\" [size]=\"'lg'\"></Icon>\n      <div *ngIf=\"isContentString\" class=\"{{prefixCls}}-text-info\">{{content}}</div>\n    </div>\n    <div *ngIf=\"!iconType\" class=\"{{prefixCls}}-text\" role=\"alert\" aria-live=\"assertive\">\n      <div *ngIf=\"isContentString\" class=\"{{prefixCls}}-text-info\">{{content}}</div>\n      <ng-template *ngIf=\"!isContentString\" [ngTemplateOutlet]=\"content\"></ng-template>\n    </div>\n  </div>\n  <a class=\"{{prefixCls}}-notice-close\">\n    <span class=\"{{prefixCls}}-notice-close-x\"></span>\n  </a>\n</div>\n",
                    host: {
                        '[class.am-toast]': 'true',
                        '[class.am-toast-mask]': 'mask',
                        '[class.am-toast-mask-top]': "mask && position === 'top'",
                        '[class.am-toast-mask-middle]': "mask && position === 'middle'",
                        '[class.am-toast-mask-bottom]': "mask && position === 'bottom'",
                        '[class.am-toast-nomask]': '!mask',
                        '[class.am-toast-nomask-top]': "!mask && position === 'top'",
                        '[class.am-toast-nomask-middle]': "!mask && position === 'middle'",
                        '[class.am-toast-nomask-bottom]': "!mask && position === 'bottom'",
                    }
                }] }
    ];
    /** @nocollapse */
    ToastComponent.ctorParameters = function () { return [
        { type: NgZone }
    ]; };
    ToastComponent.propDecorators = {
        mask: [{ type: Input }],
        content: [{ type: Input }],
        iconType: [{ type: Input }],
        position: [{ type: Input }]
    };
    return ToastComponent;
}());
export { ToastComponent };
if (false) {
    /** @type {?} */
    ToastComponent.prototype.prefixCls;
    /** @type {?} */
    ToastComponent.prototype.isContentString;
    /** @type {?} */
    ToastComponent.prototype.maskClassMap;
    /** @type {?} */
    ToastComponent.prototype.transitionName;
    /**
     * @type {?}
     * @private
     */
    ToastComponent.prototype._iconType;
    /**
     * @type {?}
     * @private
     */
    ToastComponent.prototype._content;
    /** @type {?} */
    ToastComponent.prototype.mask;
    /** @type {?} */
    ToastComponent.prototype.position;
    /**
     * @type {?}
     * @private
     */
    ToastComponent.prototype._zone;
}
