// ng-zorro-antd-mobile/esm2015/locale-provider/locale-provider.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, Optional, SkipSelf } from '@angular/core';
import { LOCAL_PROVIDER_TOKEN } from './locale-provider.token';
import { BehaviorSubject } from 'rxjs';
import zh_CN from './locale/zh_CN';
export class LocaleProviderService {
    /**
     * @param {?} locale
     */
    constructor(locale) {
        this._change = new BehaviorSubject(this._locale);
        this.setLocale(locale || zh_CN);
    }
    /**
     * @return {?}
     */
    get localeChange() {
        return this._change.asObservable();
    }
    /**
     * @param {?} keyPath
     * @return {?}
     */
    getLocaleValue(keyPath) {
        /** @type {?} */
        const content = (/** @type {?} */ (this._getObjectPath(this._locale, keyPath)));
        if (typeof content === 'string') {
            return content;
        }
        return '';
    }
    /**
     * @param {?} keyPath
     * @return {?}
     */
    getLocaleSubObj(keyPath) {
        /** @type {?} */
        const content = (/** @type {?} */ (this._getObjectPath(this._locale, keyPath)));
        if (typeof content === 'object') {
            return content;
        }
        return null;
    }
    /**
     * @param {?} locale
     * @return {?}
     */
    setLocale(locale) {
        if (!locale || (this._locale && this._locale.locale === locale.locale)) {
            return;
        }
        this._locale = locale;
        this._change.next(locale);
    }
    /**
     * @return {?}
     */
    getLocaleId() {
        return this._locale && this._locale.locale ? this._locale.locale : '';
    }
    /**
     * @return {?}
     */
    getLocale() {
        return this._locale;
    }
    /**
     * @private
     * @param {?} obj
     * @param {?} path
     * @return {?}
     */
    _getObjectPath(obj, path) {
        /** @type {?} */
        let res = obj;
        /** @type {?} */
        const paths = path.split('.');
        /** @type {?} */
        const depth = paths.length;
        /** @type {?} */
        let index = 0;
        while (res && index < depth) {
            res = res[paths[index++]];
        }
        return index === depth ? res : null;
    }
}
LocaleProviderService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LocaleProviderService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [LOCAL_PROVIDER_TOKEN,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LocaleProviderService.prototype._locale;
    /**
     * @type {?}
     * @private
     */
    LocaleProviderService.prototype._change;
}
/**
 * @param {?} exist
 * @param {?} locale
 * @return {?}
 */
export function LOCALE_PROVIDER_SERVICE_FACTORY(exist, locale) {
    return exist || new LocaleProviderService(locale);
}
/** @type {?} */
export const LOCALE_PROVIDER_SERVICE_PROVIDER = {
    provide: LocaleProviderService,
    useFactory: LOCALE_PROVIDER_SERVICE_FACTORY,
    deps: [[new Optional(), new SkipSelf(), LocaleProviderService], LOCAL_PROVIDER_TOKEN]
};
