// ng-zorro-antd-mobile/esm5/picker/picker.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PickerComponent } from './picker.component';
import { PickerDirective } from './picker.directive';
import { PickerOptions } from './picker-options.provider';
import { PopupService } from '../core/services/popup.service';
import { PickerService } from './picker.service';
import { OverlayModule } from '@angular/cdk/overlay';
var PickerModule = /** @class */ (function () {
    function PickerModule() {
    }
    PickerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, OverlayModule],
                    exports: [PickerComponent, PickerDirective],
                    declarations: [PickerComponent, PickerDirective],
                    providers: [PickerOptions, PopupService, PickerService],
                    entryComponents: [PickerComponent]
                },] }
    ];
    return PickerModule;
}());
export { PickerModule };
