// ng-zorro-antd-mobile/esm5/picker/picker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, NgZone, Injector, Renderer2, Directive, ElementRef, forwardRef, HostListener, EventEmitter, ViewContainerRef, ComponentFactoryResolver } from '@angular/core';
import { PickerComponent } from './picker.component';
import { PickerOptions } from './picker-options.provider';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var PickerDirective = /** @class */ (function () {
    function PickerDirective(_viewContainerRef, _elm, _defaultOptions, _cfr, _renderer, _zone) {
        this._viewContainerRef = _viewContainerRef;
        this._elm = _elm;
        this._defaultOptions = _defaultOptions;
        this._cfr = _cfr;
        this._renderer = _renderer;
        this._zone = _zone;
        this._eventListeners = [];
        this.onVisibleChange = new EventEmitter(true);
        this.onPickerChange = new EventEmitter();
        this.onDismiss = new EventEmitter();
        this.onChange = (/**
         * @return {?}
         */
        function () { return null; });
        this.onTouched = (/**
         * @return {?}
         */
        function () { return null; });
    }
    /**
     * @return {?}
     */
    PickerDirective.prototype.togglePicker = /**
     * @return {?}
     */
    function () {
        if (!this.picker) {
            this.showPicker();
        }
        else {
            this.hidePicker();
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.onVisibleChange.emit(false);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    PickerDirective.prototype.ngOnChanges = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value.cols && this.picker) {
            this.picker.instance.options.cols = value.cols.currentValue;
        }
        if (value.data && this.picker) {
            this.picker.instance.options.data = value.data.currentValue;
            this.showPicker();
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.hidePicker();
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PickerDirective.prototype.onDocumentClick = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.picker &&
            !this._elm.nativeElement.contains(event.target) &&
            !this.picker.location.nativeElement.contains(event.target)) {
            this.hidePicker();
        }
    };
    /**
     * @private
     * @return {?}
     */
    PickerDirective.prototype.showPicker = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.picker) {
            this._zone.run((/**
             * @return {?}
             */
            function () {
                _this.picker.instance.init();
            }));
        }
        else if (!this.picker && !this.disabled) {
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this._eventListeners = [
                    _this._renderer.listen('document', 'click', (/**
                     * @param {?} event
                     * @return {?}
                     */
                    function (event) { return _this.onDocumentClick(event); })),
                    _this._renderer.listen('document', 'touchend', (/**
                     * @param {?} event
                     * @return {?}
                     */
                    function (event) { return _this.onDocumentClick(event); }))
                ];
            }));
            /** @type {?} */
            var options_1 = new PickerOptions();
            Object.assign(options_1, this._defaultOptions, {
                hidePicker: (/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) {
                    _this.hidePicker();
                }),
                updateNgModel: (/**
                 * @param {?} value
                 * @return {?}
                 */
                function (value) {
                    _this.onChange(value);
                })
            });
            /** @type {?} */
            var optionalParams = [
                'data',
                'value',
                'cols',
                'mask',
                'title',
                'okText',
                'dismissText',
                'disabled',
                'cascade',
                'appendToBody',
                'indicatorStyle',
                'onPickerChange',
                'onVisibleChange',
                'onDismiss'
            ];
            optionalParams.forEach((/**
             * @param {?} param
             * @return {?}
             */
            function (param) {
                if (typeof _this[param] !== 'undefined') {
                    ((/** @type {?} */ (options_1)))[param] = _this[param];
                }
            }));
            /** @type {?} */
            var componentFactory = this._cfr.resolveComponentFactory(PickerComponent);
            /** @type {?} */
            var childInjector = Injector.create([
                {
                    provide: PickerOptions,
                    useValue: options_1
                }
            ]);
            this.picker = this._viewContainerRef.createComponent(componentFactory, this._viewContainerRef.length, childInjector);
            if (options_1.appendToBody) {
                document.body.appendChild(this.picker.location.nativeElement);
            }
            this.onVisibleChange.emit(true);
        }
    };
    /**
     * @private
     * @return {?}
     */
    PickerDirective.prototype.hidePicker = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.picker) {
            this.picker.destroy();
            delete this.picker;
            this.onVisibleChange.emit(false);
            this._eventListeners.forEach((/**
             * @param {?} fn
             * @return {?}
             */
            function (fn) { return fn(); }));
            this._eventListeners = [];
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    PickerDirective.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = Array.isArray(value) ? value : [];
        if (this.picker) {
            this.picker.instance.options.value = this.value;
            this.showPicker();
            this.picker.instance.reloadPicker();
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PickerDirective.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PickerDirective.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    PickerDirective.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    PickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[Picker], [nzm-picker]',
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return PickerDirective; })),
                            multi: true
                        }
                    ]
                },] }
    ];
    /** @nocollapse */
    PickerDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: PickerOptions },
        { type: ComponentFactoryResolver },
        { type: Renderer2 },
        { type: NgZone }
    ]; };
    PickerDirective.propDecorators = {
        data: [{ type: Input }],
        cols: [{ type: Input }],
        mask: [{ type: Input }],
        title: [{ type: Input }],
        visible: [{ type: Input }],
        okText: [{ type: Input }],
        dismissText: [{ type: Input }],
        disabled: [{ type: Input }],
        cascade: [{ type: Input }],
        appendToBody: [{ type: Input }],
        indicatorStyle: [{ type: Input }],
        onVisibleChange: [{ type: Output }],
        onPickerChange: [{ type: Output }],
        onDismiss: [{ type: Output }],
        togglePicker: [{ type: HostListener, args: ['click',] }]
    };
    return PickerDirective;
}());
export { PickerDirective };
if (false) {
    /** @type {?} */
    PickerDirective.prototype.picker;
    /** @type {?} */
    PickerDirective.prototype.value;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._eventListeners;
    /** @type {?} */
    PickerDirective.prototype.data;
    /** @type {?} */
    PickerDirective.prototype.cols;
    /** @type {?} */
    PickerDirective.prototype.mask;
    /** @type {?} */
    PickerDirective.prototype.title;
    /** @type {?} */
    PickerDirective.prototype.visible;
    /** @type {?} */
    PickerDirective.prototype.okText;
    /** @type {?} */
    PickerDirective.prototype.dismissText;
    /** @type {?} */
    PickerDirective.prototype.disabled;
    /** @type {?} */
    PickerDirective.prototype.cascade;
    /** @type {?} */
    PickerDirective.prototype.appendToBody;
    /** @type {?} */
    PickerDirective.prototype.indicatorStyle;
    /** @type {?} */
    PickerDirective.prototype.onVisibleChange;
    /** @type {?} */
    PickerDirective.prototype.onPickerChange;
    /** @type {?} */
    PickerDirective.prototype.onDismiss;
    /** @type {?} */
    PickerDirective.prototype.onChange;
    /** @type {?} */
    PickerDirective.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._elm;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._defaultOptions;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._cfr;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    PickerDirective.prototype._zone;
}
