// ng-zorro-antd-mobile/esm5/picker/picker-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
/**
 * @record
 */
export function PickerOptionsInterface() { }
if (false) {
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.data;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.cols;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.mask;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.title;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.okText;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.dismissText;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.disabled;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.cascade;
    /** @type {?|undefined} */
    PickerOptionsInterface.prototype.appendToBody;
}
var PickerOptions = /** @class */ (function () {
    function PickerOptions() {
        this.data = [];
        this.value = [];
        this.cols = 3;
        this.mask = true;
        this.title = '';
        this.okText = '确定';
        this.dismissText = '取消';
        this.disabled = false;
        this.cascade = true;
        this.appendToBody = false;
        this.onDismiss = new EventEmitter();
        this.onPickerChange = new EventEmitter();
        this.indicatorStyle = {};
    }
    return PickerOptions;
}());
export { PickerOptions };
if (false) {
    /** @type {?} */
    PickerOptions.prototype.data;
    /** @type {?} */
    PickerOptions.prototype.value;
    /** @type {?} */
    PickerOptions.prototype.cols;
    /** @type {?} */
    PickerOptions.prototype.mask;
    /** @type {?} */
    PickerOptions.prototype.title;
    /** @type {?} */
    PickerOptions.prototype.okText;
    /** @type {?} */
    PickerOptions.prototype.dismissText;
    /** @type {?} */
    PickerOptions.prototype.disabled;
    /** @type {?} */
    PickerOptions.prototype.cascade;
    /** @type {?} */
    PickerOptions.prototype.appendToBody;
    /** @type {?} */
    PickerOptions.prototype.onDismiss;
    /** @type {?} */
    PickerOptions.prototype.onPickerChange;
    /** @type {?} */
    PickerOptions.prototype.indicatorStyle;
    /** @type {?} */
    PickerOptions.prototype.hidePicker;
    /** @type {?} */
    PickerOptions.prototype.confirm;
    /** @type {?} */
    PickerOptions.prototype.cancel;
    /** @type {?} */
    PickerOptions.prototype.updateNgModel;
}
