// ng-zorro-antd-mobile/esm5/checkbox/agree-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, Output, EventEmitter, HostBinding, ChangeDetectorRef, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var AgreeItemComponent = /** @class */ (function () {
    function AgreeItemComponent(cdr) {
        this.cdr = cdr;
        this.prefixCls = 'am-checkbox';
        this.checked = false;
        this._disabled = false;
        this.onChange = new EventEmitter();
        this.checkboxAgree = true;
    }
    Object.defineProperty(AgreeItemComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    AgreeItemComponent.prototype.change = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.checked = event.checked;
        this._ngModelOnChange(event.checked);
        this.onChange.emit(event);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    AgreeItemComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.checked = value;
        this.cdr.markForCheck();
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AgreeItemComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AgreeItemComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnTouched = fn;
    };
    AgreeItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'AgreeItem, nzm-agree-item',
                    template: "<label Checkbox\n       class=\"{{prefixCls}}-agree-label\"\n       [name]=\"name\"\n       [value]=\"value\"\n       [checked]=\"checked\"\n       [disabled]=\"disabled\"\n       (onChange)=\"change($event)\"\n>\n  <ng-content></ng-content>\n</label>",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return AgreeItemComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    AgreeItemComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    AgreeItemComponent.propDecorators = {
        name: [{ type: Input }],
        value: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        checkboxAgree: [{ type: HostBinding, args: ['class.am-checkbox-agree',] }]
    };
    return AgreeItemComponent;
}());
export { AgreeItemComponent };
if (false) {
    /** @type {?} */
    AgreeItemComponent.prototype.prefixCls;
    /** @type {?} */
    AgreeItemComponent.prototype.checked;
    /**
     * @type {?}
     * @private
     */
    AgreeItemComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    AgreeItemComponent.prototype._ngModelOnChange;
    /**
     * @type {?}
     * @private
     */
    AgreeItemComponent.prototype._ngModelOnTouched;
    /** @type {?} */
    AgreeItemComponent.prototype.name;
    /** @type {?} */
    AgreeItemComponent.prototype.value;
    /** @type {?} */
    AgreeItemComponent.prototype.onChange;
    /** @type {?} */
    AgreeItemComponent.prototype.checkboxAgree;
    /**
     * @type {?}
     * @private
     */
    AgreeItemComponent.prototype.cdr;
}
