// ng-zorro-antd-mobile/esm5/calendar/confirm-panel/confirm-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
import { formatDate } from '../util/index';
var CalendarConfirmPanelComponent = /** @class */ (function () {
    function CalendarConfirmPanelComponent() {
        var _this = this;
        this.props = (/** @type {?} */ ({
            formatStr: 'yyyy-MM-dd hh:mm'
        }));
        this.confirmPane = true;
        this.triggerConfirm = (/**
         * @return {?}
         */
        function () {
            var _a = _this.props, onConfirm = _a.onConfirm, disableBtn = _a.disableBtn;
            !disableBtn && onConfirm();
        });
    }
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "propsData", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props = tslib_1.__assign({}, this.props, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "disableBtn", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.disableBtn = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "formatStr", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.formatStr = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "startDateTime", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.startDateTime = value;
            this.formatTime();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "endDateTime", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.endDateTime = value;
            this.formatTime();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalendarConfirmPanelComponent.prototype, "onConfirm", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.onConfirm = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CalendarConfirmPanelComponent.prototype.formatTime = /**
     * @return {?}
     */
    function () {
        var _a = this.props, type = _a.type, locale = _a.locale, disableBtn = _a.disableBtn;
        var _b = this.props, startDateTime = _b.startDateTime, endDateTime = _b.endDateTime;
        if (startDateTime && endDateTime && +startDateTime > +endDateTime) {
            /** @type {?} */
            var tmp = startDateTime;
            startDateTime = endDateTime;
            endDateTime = tmp;
        }
        this.startTimeStr = startDateTime ? this.selfFormatDate(startDateTime) : locale.noChoose;
        this.endTimeStr = endDateTime ? this.selfFormatDate(endDateTime) : locale.noChoose;
        /** @type {?} */
        var btnCls = disableBtn ? 'button button-disable' : 'button';
        if (type === 'one') {
            btnCls += ' button-full';
        }
        this.btnCls = btnCls;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarConfirmPanelComponent.prototype.selfFormatDate = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        var _a = this.props, _b = _a.formatStr, formatStr = _b === void 0 ? '' : _b, locale = _a.locale;
        return formatDate(date, formatStr, locale);
    };
    CalendarConfirmPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CalendarConfirmPanel, nzm-calendar-confirm-panel',
                    template: "<div *ngIf=\"props.type === 'range'\" class=\"info\">\n  <p>{{props.locale.start}}: <span class=\"{{!props.startDateTime ? 'grey' : ''}}\">{{startTimeStr}}</span></p>\n  <p>{{props.locale.end}}: <span class=\"{{!props.endDateTime ? 'grey' : ''}}\">{{endTimeStr}}</span></p>\n</div>\n<div [ngClass]=\"btnCls\" (click)=\"triggerConfirm()\">\n  {{props.locale.confirm}}\n</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    CalendarConfirmPanelComponent.ctorParameters = function () { return []; };
    CalendarConfirmPanelComponent.propDecorators = {
        propsData: [{ type: Input }],
        disableBtn: [{ type: Input }],
        formatStr: [{ type: Input }],
        startDateTime: [{ type: Input }],
        endDateTime: [{ type: Input }],
        onConfirm: [{ type: Input }],
        confirmPane: [{ type: HostBinding, args: ['class.confirm-panel',] }]
    };
    return CalendarConfirmPanelComponent;
}());
export { CalendarConfirmPanelComponent };
if (false) {
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.props;
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.startTimeStr;
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.endTimeStr;
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.btnCls;
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.confirmPane;
    /** @type {?} */
    CalendarConfirmPanelComponent.prototype.triggerConfirm;
}
