// ng-zorro-antd-mobile/esm2015/white-space/white-space.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from '@angular/core';
export class WhiteSpaceComponent {
    constructor() {
        this.prefixCls = 'am-whitespace';
        this.size = 'md';
        this.amWhiteSpace = true;
    }
    /**
     * @return {?}
     */
    get amWhitespaceXs() {
        return this.size === 'xs';
    }
    /**
     * @return {?}
     */
    get amWhitespaceSm() {
        return this.size === 'sm';
    }
    /**
     * @return {?}
     */
    get amWhitespaceMd() {
        return this.size === 'md';
    }
    /**
     * @return {?}
     */
    get amWhitespaceLg() {
        return this.size === 'lg';
    }
    /**
     * @return {?}
     */
    get amWhitespaceXl() {
        return this.size === 'xl';
    }
}
WhiteSpaceComponent.decorators = [
    { type: Component, args: [{
                selector: 'WhiteSpace, nzm-whitespace',
                template: ``
            }] }
];
/** @nocollapse */
WhiteSpaceComponent.ctorParameters = () => [];
WhiteSpaceComponent.propDecorators = {
    size: [{ type: Input }],
    amWhiteSpace: [{ type: HostBinding, args: ['class.am-whitespace',] }],
    amWhitespaceXs: [{ type: HostBinding, args: ['class.am-whitespace-xs',] }],
    amWhitespaceSm: [{ type: HostBinding, args: ['class.am-whitespace-sm',] }],
    amWhitespaceMd: [{ type: HostBinding, args: ['class.am-whitespace-md',] }],
    amWhitespaceLg: [{ type: HostBinding, args: ['class.am-whitespace-lg',] }],
    amWhitespaceXl: [{ type: HostBinding, args: ['class.am-whitespace-xl',] }]
};
if (false) {
    /** @type {?} */
    WhiteSpaceComponent.prototype.prefixCls;
    /** @type {?} */
    WhiteSpaceComponent.prototype.size;
    /** @type {?} */
    WhiteSpaceComponent.prototype.amWhiteSpace;
}
