// ng-zorro-antd-mobile/esm5/carousel/dotindicator/dotindicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, HostBinding } from '@angular/core';
var DotIndicatorComponent = /** @class */ (function () {
    function DotIndicatorComponent() {
        this.items = [];
        this._page = 0;
        this._pageCount = 0;
        this.dotStyle = {};
        this.dotActiveStyle = {};
        this.dotColor = 'white';
        this.dotIndicator = true;
    }
    Object.defineProperty(DotIndicatorComponent.prototype, "page", {
        set: /**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            this._page = p;
            this.updateSelected();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DotIndicatorComponent.prototype, "pageCount", {
        set: /**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            this._pageCount = p || 0;
            this.updateItems();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @return {?}
     */
    DotIndicatorComponent.prototype.updateItems = /**
     * @private
     * @return {?}
     */
    function () {
        this.items = new Array(this._pageCount);
        for (var i = 0; i < this._pageCount; i++) {
            this.items[i] = { active: i == this._page };
        }
    };
    /**
     * @private
     * @return {?}
     */
    DotIndicatorComponent.prototype.updateSelected = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.items.length != this._pageCount) {
            return this.updateItems();
        }
        if (this.items.length == 0) {
            return;
        }
        for (var i = 0; i < this._pageCount; i++) {
            this.items[i].active = false;
        }
        this.items[this._page].active = true;
    };
    DotIndicatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'DotIndicator, nzm-dot-indicator',
                    template: "<div class=\"am-carousel-wrap\">\n  <div *ngFor=\"let item of items\"\n       class=\"am-carousel-wrap-dot\"\n       [class.am-carousel-wrap-dot-active]=\"item.active\"\n  >\n    <span [ngStyle]=\"item.active ? dotActiveStyle : dotStyle\"></span>\n  </div>\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    DotIndicatorComponent.propDecorators = {
        page: [{ type: Input }],
        pageCount: [{ type: Input }],
        dotStyle: [{ type: Input }],
        dotActiveStyle: [{ type: Input }],
        dotColor: [{ type: Input }],
        dotIndicator: [{ type: HostBinding, args: ['class.dot-indicator',] }]
    };
    return DotIndicatorComponent;
}());
export { DotIndicatorComponent };
if (false) {
    /** @type {?} */
    DotIndicatorComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    DotIndicatorComponent.prototype._page;
    /**
     * @type {?}
     * @private
     */
    DotIndicatorComponent.prototype._pageCount;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotStyle;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotActiveStyle;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotColor;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotIndicator;
}
