// ng-zorro-antd-mobile/esm2015/accordion/accordion-group/accordion-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectorRef, Output, EventEmitter, ViewEncapsulation, HostBinding } from '@angular/core';
import { animate, state, style, transition, trigger } from '@angular/animations';
import { AccordionService } from '../accordion.service';
import { isTemplateRef } from '../../core/util/check';
export class AccordionGroupComponent {
    /**
     * @param {?} _accordionService
     * @param {?} _cdr
     */
    constructor(_accordionService, _cdr) {
        this._accordionService = _accordionService;
        this._cdr = _cdr;
        this.isShowChild = true;
        this.isOpened = false;
        this.disabled = false;
        this.onOpen = new EventEmitter();
        this.onClose = new EventEmitter();
        this.onChange = new EventEmitter();
        this.isTemplateRef = isTemplateRef;
        this.amItem = true;
        this.isActive = this.isOpened;
        this.addon = true;
    }
    /**
     * @return {?}
     */
    checkAndToggle() {
        this.toggle();
    }
    /**
     * @return {?}
     */
    get slide() {
        return this.isOpened ? 'down' : 'up';
    }
    /**
     * @return {?}
     */
    toggle() {
        if (this.disabled) {
            return;
        }
        this.isShowChild = true;
        /** @type {?} */
        const isOpenedBeforeWeChange = this.isOpened;
        if (this._accordionService.accordion) {
            this._accordionService.component.closeAll();
        }
        this.isOpened = !isOpenedBeforeWeChange;
        if (this.isOpened) {
            this.onOpen.emit();
        }
        else {
            this.onClose.emit();
        }
        this.onChange.emit(this.isOpened);
    }
    /**
     * @return {?}
     */
    openOnInitialization() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.isOpened = true;
            this._cdr.detectChanges();
        }), 0);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    slideAnimationDoen(event) {
        if (event.fromState === 'down' && event.toState === 'up') {
            this.isShowChild = false;
        }
    }
}
AccordionGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'AccordionPanel',
                template: "<div role=\"tab\"\n     class=\"am-accordion-header\"\n     data-scale=\"true\"\n     [attr.aria-expanded]=\"isOpened\"\n     (click)=\"checkAndToggle()\"\n>\n  <i class=\"arrow\"></i>\n  <div *ngIf=\"!isTemplateRef(header)\" [innerHTML]=\"header | safeHTML\"></div>\n  <ng-template *ngIf=\"isTemplateRef(header)\" [ngTemplateOutlet]=\"header\"></ng-template>\n</div>\n<div role=\"tabpanel\"\n     class=\"am-accordion-content\"\n     [ngClass]=\"{'am-accordion-content-active':isOpened}\"\n     [@slide]=\"slide\"\n     (@slide.done)=\"slideAnimationDoen($event)\"\n>\n  <div *ngIf=\"isShowChild\" class=\"am-accordion-content-box\">\n    <ng-content></ng-content>\n  </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                animations: [
                    trigger('slide', [
                        state('up', style({ height: 0 })),
                        state('down', style({ height: '*' })),
                        transition('down => up', [animate(200, style({ height: 0 }))]),
                        transition('up => down', [
                            animate(200, style({
                                height: '*'
                            }))
                        ])
                    ])
                ]
            }] }
];
/** @nocollapse */
AccordionGroupComponent.ctorParameters = () => [
    { type: AccordionService },
    { type: ChangeDetectorRef }
];
AccordionGroupComponent.propDecorators = {
    key: [{ type: Input }],
    header: [{ type: Input }],
    isOpened: [{ type: Input }],
    disabled: [{ type: Input }],
    onOpen: [{ type: Output }],
    onClose: [{ type: Output }],
    onChange: [{ type: Output }],
    amItem: [{ type: HostBinding, args: ['class.am-accordion-item',] }],
    isActive: [{ type: HostBinding, args: ['class.am-accordion-item-active',] }],
    addon: [{ type: HostBinding, args: ['class.addon',] }]
};
if (false) {
    /** @type {?} */
    AccordionGroupComponent.prototype.isShowChild;
    /** @type {?} */
    AccordionGroupComponent.prototype.key;
    /** @type {?} */
    AccordionGroupComponent.prototype.header;
    /** @type {?} */
    AccordionGroupComponent.prototype.isOpened;
    /** @type {?} */
    AccordionGroupComponent.prototype.disabled;
    /** @type {?} */
    AccordionGroupComponent.prototype.onOpen;
    /** @type {?} */
    AccordionGroupComponent.prototype.onClose;
    /** @type {?} */
    AccordionGroupComponent.prototype.onChange;
    /** @type {?} */
    AccordionGroupComponent.prototype.isTemplateRef;
    /** @type {?} */
    AccordionGroupComponent.prototype.amItem;
    /** @type {?} */
    AccordionGroupComponent.prototype.isActive;
    /** @type {?} */
    AccordionGroupComponent.prototype.addon;
    /**
     * @type {?}
     * @private
     */
    AccordionGroupComponent.prototype._accordionService;
    /**
     * @type {?}
     * @private
     */
    AccordionGroupComponent.prototype._cdr;
}
