// ng-zorro-antd-mobile/esm5/range/range.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ElementRef, HostBinding, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var RangeComponent = /** @class */ (function () {
    function RangeComponent(_elf) {
        this._elf = _elf;
        this.prefixCls = 'am-slider';
        this.offset = [];
        this.length = [];
        this._min = 0;
        this._max = 100;
        this._step = 1;
        this._defaultValue = [0, 0, 0];
        this._disabled = false;
        this._marks = {};
        this._dots = false;
        this._included = true;
        this._count = 1;
        this._allowCross = true;
        this._handleStyle = [];
        this._trackStyle = [];
        this.onChange = new EventEmitter();
        this.onAfterChange = new EventEmitter();
        this.amWrapper = true;
        this._ngModelOnChange = (/**
         * @return {?}
         */
        function () { });
        this._ngModelOnTouched = (/**
         * @return {?}
         */
        function () { });
    }
    Object.defineProperty(RangeComponent.prototype, "min", {
        get: /**
         * @return {?}
         */
        function () {
            return this._min;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._min = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "max", {
        get: /**
         * @return {?}
         */
        function () {
            return this._max;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._max = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "step", {
        get: /**
         * @return {?}
         */
        function () {
            return this._step;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._step = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.setValue(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "defaultValue", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._defaultValue = value;
            this._value = this._defaultValue;
            this.setValue(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "marks", {
        get: /**
         * @return {?}
         */
        function () {
            return this._marks;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._marks = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "dots", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dots;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._dots = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "included", {
        get: /**
         * @return {?}
         */
        function () {
            return this._included;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._included = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "count", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._count = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "allowCross", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._allowCross = value;
            this.setValueBound();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "pushable", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._pushable = value;
            if (this.verifyPushable()) {
                this.setValueBound();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "handleStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._handleStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._handleStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "trackStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._trackStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._trackStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RangeComponent.prototype, "railStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._railStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._railStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    RangeComponent.prototype.setCls = /**
     * @return {?}
     */
    function () {
        var _a;
        this.sliderCls = (_a = {},
            _a[this.prefixCls + "-disabled"] = this._disabled,
            _a);
    };
    /**
     * @return {?}
     */
    RangeComponent.prototype.initialValue = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var minTemp = this._min;
        if (!this.verifyPushable()) {
            this._pushable = 0;
            console.log('pushable设置无效，已大于有些value间隔，被强制设为0');
        }
        /** @type {?} */
        var initialValue = Array.apply(null, Array(this._count + 1)).map((/**
         * @return {?}
         */
        function () {
            return minTemp;
        }));
        this._defaultValue = this._defaultValue !== undefined ? this._defaultValue : initialValue;
        this._value = this._value !== undefined ? this._value : this._defaultValue;
        // 验证count值
        this._count = this._value.length - 1;
        // 验证value区间
        for (var i = 0; i < this._value.length; i++) {
            if (this._value[i] < this._min) {
                this._value[i] = this._min;
            }
            else if (this._value[i] > this._max) {
                this._value[i] = this._max;
            }
        }
        if (this._count > 0) {
            this.upperBound = Math.max.apply(Math, tslib_1.__spread(this._value));
            this.lowerBound = Math.min.apply(Math, tslib_1.__spread(this._value));
        }
    };
    /**
     * @param {?} e
     * @param {?} i
     * @return {?}
     */
    RangeComponent.prototype.handleChange = /**
     * @param {?} e
     * @param {?} i
     * @return {?}
     */
    function (e, i) {
        /** @type {?} */
        var temp = tslib_1.__spread(this._value);
        temp[i] = e;
        this.upperBound = Math.max.apply(Math, tslib_1.__spread(temp));
        this.lowerBound = Math.min.apply(Math, tslib_1.__spread(temp));
        this.setTrackStyle(temp);
        this.onChange.emit(temp);
    };
    /**
     * @param {?} e
     * @param {?} i
     * @return {?}
     */
    RangeComponent.prototype.handleAfterChange = /**
     * @param {?} e
     * @param {?} i
     * @return {?}
     */
    function (e, i) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this._value[i] = e;
            _this.upperBound = Math.max.apply(Math, tslib_1.__spread(_this._value));
            _this.lowerBound = Math.min.apply(Math, tslib_1.__spread(_this._value));
            _this.setTrackStyle(_this._value);
            _this.onAfterChange.emit(_this._value);
            _this._ngModelOnChange(_this._value);
            _this.setValueBound();
        }), 0);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RangeComponent.prototype.setTrackStyle = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value && value.length === this._count + 1) {
            value.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) { return a - b; }));
            for (var i = 0; i < this._count; i++) {
                this.offset[i] = (value[i] * 100) / (this._max - this._min);
                this.length[i] = ((value[i + 1] - value[i]) * 100) / (this._max - this._min);
            }
        }
    };
    /**
     * @return {?}
     */
    RangeComponent.prototype.setValueBound = /**
     * @return {?}
     */
    function () {
        this.maxBound = [];
        this.minBound = [];
        if ((this._allowCross && this._pushable === undefined) || this._handleCount <= 1) {
            for (var i = 0; i < this._handleCount; i++) {
                this.maxBound[i] = this._max;
                this.minBound[i] = this._min;
            }
        }
        else {
            if (this._pushable === undefined) {
                this._pushable = 0;
            }
            for (var i = 0; i < this._handleCount; i++) {
                this.maxBound[i] = i === this._handleCount - 1 ? this._max : this._value[i + 1] - this._pushable;
                this.minBound[i] = i === 0 ? this._min : this._value[i - 1] + this._pushable;
            }
        }
    };
    /**
     * @return {?}
     */
    RangeComponent.prototype.verifyPushable = /**
     * @return {?}
     */
    function () {
        for (var i = 1; i < this._handleCount; i++) {
            /** @type {?} */
            var diff = this._value[i] - this._value[i - 1];
            if (diff < this._pushable) {
                return false;
            }
        }
        return true;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RangeComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.setValue(value, true);
    };
    /**
     * @param {?} value
     * @param {?=} isWriteValue
     * @return {?}
     */
    RangeComponent.prototype.setValue = /**
     * @param {?} value
     * @param {?=} isWriteValue
     * @return {?}
     */
    function (value, isWriteValue) {
        if (isWriteValue === void 0) { isWriteValue = false; }
        if (value) {
            this._value = value;
            this._handleCount = this._value.length + 1;
            this.initialValue();
            this.setValueBound();
            this.setCls();
            this.setTrackStyle(this._value);
            if (isWriteValue) {
                this._ngModelOnChange(this._value);
            }
            else {
                this.onAfterChange.emit(this._value);
            }
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RangeComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RangeComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnTouched = fn;
    };
    /**
     * @return {?}
     */
    RangeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initialValue();
        this.setValueBound();
        this._handleCount = this._count + 1;
        this.setCls();
        /** @type {?} */
        var sliderCoords = this._elf.nativeElement.getElementsByClassName('am-slider')[0].getBoundingClientRect();
        this.sliderLength = sliderCoords.width;
        this.sliderStart = sliderCoords.left;
    };
    RangeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Range , nzm-range',
                    template: "<div class=\"am-slider\" [ngClass]=\"sliderCls\">\n  <div class=\"am-slider-rail\" [ngStyle]=\"railStyle\"></div>\n  <SliderTrack *ngFor=\"let off of offset ; let i = index\"\n               [className]=\"'am-slider-track'\"\n               [included]=\"included\"\n               [style]=\"trackStyle[i]\"\n               [offset]=\"off\"\n               [length]=\"length[i]\"\n  ></SliderTrack>\n  <SliderSteps [max]=\"max\"\n               [min]=\"min\"\n               [dots]=\"dots\"\n               [step]=\"step\"\n               [marks]=\"marks\"\n               [upperBound]=\"upperBound\"\n               [lowerBound]=\"lowerBound\"\n  ></SliderSteps>\n  <SliderHandle *ngFor=\"let val of value ; let i = index\"\n                [max]=\"max\"\n                [min]=\"min\"\n                [maxBound]=\"maxBound[i]\"\n                [minBound]=\"minBound[i]\"\n                [value]=\"val\"\n                [step]=\"step\"\n                [disabled]=\"disabled\"\n                [sliderLength]=\"sliderLength\"\n                [sliderStart]=\"sliderStart\"\n                [handleStyle]=\"handleStyle[i]\"\n                (onChange)=\"handleChange($event , i)\"\n                (onAfterChange)=\"handleAfterChange($event , i)\"\n  ></SliderHandle>\n  <SliderMarks [max]=\"max\"\n               [min]=\"min\"\n               [marks]=\"marks\"\n               [upperBound]=\"upperBound\"\n               [lowerBound]=\"lowerBound\"\n  ></SliderMarks>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return RangeComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    RangeComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    RangeComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        step: [{ type: Input }],
        value: [{ type: Input }],
        defaultValue: [{ type: Input }],
        disabled: [{ type: Input }],
        marks: [{ type: Input }],
        dots: [{ type: Input }],
        included: [{ type: Input }],
        count: [{ type: Input }],
        allowCross: [{ type: Input }],
        pushable: [{ type: Input }],
        handleStyle: [{ type: Input }],
        trackStyle: [{ type: Input }],
        railStyle: [{ type: Input }],
        onChange: [{ type: Output }],
        onAfterChange: [{ type: Output }],
        amWrapper: [{ type: HostBinding, args: ['class.am-slider-wrapper',] }]
    };
    return RangeComponent;
}());
export { RangeComponent };
if (false) {
    /** @type {?} */
    RangeComponent.prototype.prefixCls;
    /** @type {?} */
    RangeComponent.prototype.offset;
    /** @type {?} */
    RangeComponent.prototype.length;
    /** @type {?} */
    RangeComponent.prototype.upperBound;
    /** @type {?} */
    RangeComponent.prototype.lowerBound;
    /** @type {?} */
    RangeComponent.prototype.maxBound;
    /** @type {?} */
    RangeComponent.prototype.minBound;
    /** @type {?} */
    RangeComponent.prototype.sliderCls;
    /** @type {?} */
    RangeComponent.prototype.sliderLength;
    /** @type {?} */
    RangeComponent.prototype.sliderStart;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._min;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._max;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._step;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._defaultValue;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._marks;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._dots;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._included;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._count;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._allowCross;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._pushable;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._handleStyle;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._trackStyle;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._railStyle;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._handleCount;
    /** @type {?} */
    RangeComponent.prototype.onChange;
    /** @type {?} */
    RangeComponent.prototype.onAfterChange;
    /** @type {?} */
    RangeComponent.prototype.amWrapper;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._ngModelOnChange;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._ngModelOnTouched;
    /**
     * @type {?}
     * @private
     */
    RangeComponent.prototype._elf;
}
