// ng-zorro-antd-mobile/esm5/calendar/date/DataTypes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export var DateModels;
(function (DateModels) {
    var SelectType;
    (function (SelectType) {
        SelectType[SelectType["None"] = 0] = "None";
        SelectType[SelectType["Single"] = 1] = "Single";
        SelectType[SelectType["All"] = 2] = "All";
        SelectType[SelectType["Only"] = 3] = "Only";
        SelectType[SelectType["Start"] = 4] = "Start";
        SelectType[SelectType["Middle"] = 5] = "Middle";
        SelectType[SelectType["End"] = 6] = "End";
    })(SelectType = DateModels.SelectType || (DateModels.SelectType = {}));
    /**
     * @record
     */
    function Locale() { }
    DateModels.Locale = Locale;
    if (false) {
        /** @type {?} */
        Locale.prototype.title;
        /** @type {?} */
        Locale.prototype.today;
        /** @type {?} */
        Locale.prototype.month;
        /** @type {?} */
        Locale.prototype.year;
        /** @type {?} */
        Locale.prototype.am;
        /** @type {?} */
        Locale.prototype.pm;
        /** @type {?} */
        Locale.prototype.dateFormat;
        /** @type {?} */
        Locale.prototype.dateTimeFormat;
        /** @type {?} */
        Locale.prototype.noChoose;
        /** @type {?} */
        Locale.prototype.week;
        /** @type {?} */
        Locale.prototype.clear;
        /** @type {?} */
        Locale.prototype.selectTime;
        /** @type {?} */
        Locale.prototype.selectStartTime;
        /** @type {?} */
        Locale.prototype.selectEndTime;
        /** @type {?} */
        Locale.prototype.start;
        /** @type {?} */
        Locale.prototype.end;
        /** @type {?} */
        Locale.prototype.begin;
        /** @type {?} */
        Locale.prototype.over;
        /** @type {?} */
        Locale.prototype.begin_over;
        /** @type {?} */
        Locale.prototype.confirm;
        /** @type {?} */
        Locale.prototype.monthTitle;
        /** @type {?} */
        Locale.prototype.loadPrevMonth;
        /** @type {?} */
        Locale.prototype.yesterday;
        /** @type {?} */
        Locale.prototype.lastWeek;
        /** @type {?} */
        Locale.prototype.lastMonth;
    }
    /**
     * @record
     */
    function CellData() { }
    DateModels.CellData = CellData;
    if (false) {
        /** @type {?} */
        CellData.prototype.tick;
        /** @type {?} */
        CellData.prototype.dayOfMonth;
        /** @type {?} */
        CellData.prototype.selected;
        /** @type {?} */
        CellData.prototype.isFirstOfMonth;
        /** @type {?} */
        CellData.prototype.isLastOfMonth;
        /** @type {?} */
        CellData.prototype.outOfDate;
    }
    /**
     * @record
     */
    function ExtraData() { }
    DateModels.ExtraData = ExtraData;
    if (false) {
        /** @type {?|undefined} */
        ExtraData.prototype.info;
        /** @type {?|undefined} */
        ExtraData.prototype.disable;
        /** @type {?|undefined} */
        ExtraData.prototype.cellCls;
        /** @type {?|undefined} */
        ExtraData.prototype.cellRender;
    }
    /**
     * @record
     */
    function MonthData() { }
    DateModels.MonthData = MonthData;
    if (false) {
        /** @type {?} */
        MonthData.prototype.title;
        /** @type {?} */
        MonthData.prototype.firstDate;
        /** @type {?} */
        MonthData.prototype.lastDate;
        /** @type {?} */
        MonthData.prototype.weeks;
        /** @type {?|undefined} */
        MonthData.prototype.component;
        /** @type {?|undefined} */
        MonthData.prototype.height;
        /** @type {?|undefined} */
        MonthData.prototype.y;
        /** @type {?|undefined} */
        MonthData.prototype.updateLayout;
        /** @type {?|undefined} */
        MonthData.prototype.componentRef;
    }
})(DateModels || (DateModels = {}));
