// ng-zorro-antd-mobile/esm5/flex/flex.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, HostBinding } from '@angular/core';
var FlexComponent = /** @class */ (function () {
    function FlexComponent() {
        this.defaultProps = {
            prefixCls: 'am-flexbox',
            align: 'center'
        };
        // _wrapCls = {};
        this._direction = '';
        this._justify = '';
        this._alignContent = '';
        this.amFlexbox = true;
    }
    Object.defineProperty(FlexComponent.prototype, "direction", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._direction = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "wrap", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._wrap = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "justify", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._justify = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "align", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.defaultProps.align = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "alignContent", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._alignContent = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxDirRow", {
        get: /**
         * @return {?}
         */
        function () {
            return this._direction === 'row';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxDirRowReverse", {
        get: /**
         * @return {?}
         */
        function () {
            return this._direction === 'row-reverse';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxDirColumn", {
        get: /**
         * @return {?}
         */
        function () {
            return this._direction === 'column';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxDirColumnReverse", {
        get: /**
         * @return {?}
         */
        function () {
            return this._direction === 'column-reverse';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxNowrap", {
        get: /**
         * @return {?}
         */
        function () {
            return this._wrap === 'nowrap';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxWrap", {
        get: /**
         * @return {?}
         */
        function () {
            return this._wrap === 'wrap';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxWrapReverse", {
        get: /**
         * @return {?}
         */
        function () {
            return this._wrap === 'wrap-reverse';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxJustifyStart", {
        get: /**
         * @return {?}
         */
        function () {
            return this._justify === 'start';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxJustifyCenter", {
        get: /**
         * @return {?}
         */
        function () {
            return this._justify === 'center';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxJustifyEnd", {
        get: /**
         * @return {?}
         */
        function () {
            return this._justify === 'end';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxJustifyBetween", {
        get: /**
         * @return {?}
         */
        function () {
            return this._justify === 'between';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignAround", {
        get: /**
         * @return {?}
         */
        function () {
            return this._justify === 'around';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignStart", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.align === 'start';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignCenter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.align === 'center';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignEnd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.align === 'end';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignBaseline", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.align === 'baseline';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignStretch", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.align === 'stretch';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignContentStart", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'start';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignCotentCenter", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'center';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignContentEnd", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'end';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignContentBetween", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'between';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignContentAround", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'around';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexComponent.prototype, "amFlexboxAlignContentStretch", {
        get: /**
         * @return {?}
         */
        function () {
            return this._alignContent === 'stretch';
        },
        enumerable: true,
        configurable: true
    });
    FlexComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Flex, nzm-flex',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    FlexComponent.ctorParameters = function () { return []; };
    FlexComponent.propDecorators = {
        direction: [{ type: Input }],
        wrap: [{ type: Input }],
        justify: [{ type: Input }],
        align: [{ type: Input }],
        alignContent: [{ type: Input }],
        amFlexbox: [{ type: HostBinding, args: ['class.am-flexbox',] }],
        amFlexboxDirRow: [{ type: HostBinding, args: ['class.am-flexbox-dir-row',] }],
        amFlexboxDirRowReverse: [{ type: HostBinding, args: ['class.am-flexbox-dir-row-reverse',] }],
        amFlexboxDirColumn: [{ type: HostBinding, args: ['class.am-flexbox-dir-column',] }],
        amFlexboxDirColumnReverse: [{ type: HostBinding, args: ['class.am-flexbox-dir-column-reverse',] }],
        amFlexboxNowrap: [{ type: HostBinding, args: ['class.am-flexbox-nowrap',] }],
        amFlexboxWrap: [{ type: HostBinding, args: ['class.am-flexbox-wrap',] }],
        amFlexboxWrapReverse: [{ type: HostBinding, args: ['class.am-flexbox-wrap-reverse',] }],
        amFlexboxJustifyStart: [{ type: HostBinding, args: ['class.am-flexbox-justify-start',] }],
        amFlexboxJustifyCenter: [{ type: HostBinding, args: ['class.am-flexbox-justify-center',] }],
        amFlexboxJustifyEnd: [{ type: HostBinding, args: ['class.am-flexbox-justify-end',] }],
        amFlexboxJustifyBetween: [{ type: HostBinding, args: ['class.am-flexbox-justify-between',] }],
        amFlexboxAlignAround: [{ type: HostBinding, args: ['class.am-flexbox-justify-around',] }],
        amFlexboxAlignStart: [{ type: HostBinding, args: ['class.am-flexbox-align-start',] }],
        amFlexboxAlignCenter: [{ type: HostBinding, args: ['class.am-flexbox-align-center',] }],
        amFlexboxAlignEnd: [{ type: HostBinding, args: ['class.am-flexbox-align-end',] }],
        amFlexboxAlignBaseline: [{ type: HostBinding, args: ['class.am-flexbox-align-baseline',] }],
        amFlexboxAlignStretch: [{ type: HostBinding, args: ['class.am-flexbox-align-stretch',] }],
        amFlexboxAlignContentStart: [{ type: HostBinding, args: ['class.am-flexbox-align-content-start',] }],
        amFlexboxAlignCotentCenter: [{ type: HostBinding, args: ['class.am-flexbox-align-content-center',] }],
        amFlexboxAlignContentEnd: [{ type: HostBinding, args: ['class.am-flexbox-align-content-end',] }],
        amFlexboxAlignContentBetween: [{ type: HostBinding, args: ['class.am-flexbox-align-content-between',] }],
        amFlexboxAlignContentAround: [{ type: HostBinding, args: ['class.am-flexbox-align-content-around',] }],
        amFlexboxAlignContentStretch: [{ type: HostBinding, args: ['class.am-flexbox-align-content-stretch',] }]
    };
    return FlexComponent;
}());
export { FlexComponent };
if (false) {
    /** @type {?} */
    FlexComponent.prototype.defaultProps;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._direction;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._wrap;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._justify;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._alignContent;
    /** @type {?} */
    FlexComponent.prototype.amFlexbox;
}
var FlexItemComponent = /** @class */ (function () {
    function FlexItemComponent() {
        this.defaultProps = {
            prefixCls: 'am-flexbox',
            align: 'center'
        };
        this.flexboxItem = true;
    }
    FlexItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'FlexItem, nzm-flex-item',
                    template: "\n    <ng-content></ng-content>\n  ",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    FlexItemComponent.propDecorators = {
        flexboxItem: [{ type: HostBinding, args: ['class.am-flexbox-item',] }]
    };
    return FlexItemComponent;
}());
export { FlexItemComponent };
if (false) {
    /** @type {?} */
    FlexItemComponent.prototype.defaultProps;
    /** @type {?} */
    FlexItemComponent.prototype.flexboxItem;
}
