// ng-zorro-antd-mobile/esm2015/carousel/dotindicator/dotindicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, HostBinding } from '@angular/core';
export class DotIndicatorComponent {
    constructor() {
        this.items = [];
        this._page = 0;
        this._pageCount = 0;
        this.dotStyle = {};
        this.dotActiveStyle = {};
        this.dotColor = 'white';
        this.dotIndicator = true;
    }
    /**
     * @param {?} p
     * @return {?}
     */
    set page(p) {
        this._page = p;
        this.updateSelected();
    }
    /**
     * @param {?} p
     * @return {?}
     */
    set pageCount(p) {
        this._pageCount = p || 0;
        this.updateItems();
    }
    /**
     * @private
     * @return {?}
     */
    updateItems() {
        this.items = new Array(this._pageCount);
        for (let i = 0; i < this._pageCount; i++) {
            this.items[i] = { active: i == this._page };
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateSelected() {
        if (this.items.length != this._pageCount) {
            return this.updateItems();
        }
        if (this.items.length == 0) {
            return;
        }
        for (let i = 0; i < this._pageCount; i++) {
            this.items[i].active = false;
        }
        this.items[this._page].active = true;
    }
}
DotIndicatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'DotIndicator, nzm-dot-indicator',
                template: "<div class=\"am-carousel-wrap\">\n  <div *ngFor=\"let item of items\"\n       class=\"am-carousel-wrap-dot\"\n       [class.am-carousel-wrap-dot-active]=\"item.active\"\n  >\n    <span [ngStyle]=\"item.active ? dotActiveStyle : dotStyle\"></span>\n  </div>\n</div>",
                encapsulation: ViewEncapsulation.None
            }] }
];
DotIndicatorComponent.propDecorators = {
    page: [{ type: Input }],
    pageCount: [{ type: Input }],
    dotStyle: [{ type: Input }],
    dotActiveStyle: [{ type: Input }],
    dotColor: [{ type: Input }],
    dotIndicator: [{ type: HostBinding, args: ['class.dot-indicator',] }]
};
if (false) {
    /** @type {?} */
    DotIndicatorComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    DotIndicatorComponent.prototype._page;
    /**
     * @type {?}
     * @private
     */
    DotIndicatorComponent.prototype._pageCount;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotStyle;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotActiveStyle;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotColor;
    /** @type {?} */
    DotIndicatorComponent.prototype.dotIndicator;
}
