// ng-zorro-antd-mobile/esm2015/toast/toast.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ApplicationRef, Compiler, NgZone, ComponentFactoryResolver, } from '@angular/core';
import { ToastComponent } from './toast.component';
import { ToastOptions } from './toast-options.provider';
/**
 * @record
 */
export function ConfigInterface() { }
if (false) {
    /** @type {?|undefined} */
    ConfigInterface.prototype.content;
    /** @type {?|undefined} */
    ConfigInterface.prototype.iconType;
    /** @type {?|undefined} */
    ConfigInterface.prototype.mask;
}
export class ToastService {
    /**
     * @param {?} _appRef
     * @param {?} _compiler
     * @param {?} _cfr
     * @param {?} _zone
     */
    constructor(_appRef, _compiler, _cfr, _zone) {
        this._appRef = _appRef;
        this._compiler = _compiler;
        this._cfr = _cfr;
        this._zone = _zone;
        ToastService._zone = this._zone;
        ToastService._appRef = this._appRef;
        ToastService._toastCompFactory = this._cfr.resolveComponentFactory(ToastComponent);
    }
    /**
     * @param {?} config
     * @param {?} options
     * @return {?}
     */
    static _initConfig(config, options) {
        /** @type {?} */
        const props = {};
        /** @type {?} */
        const optionalParams = ['content', 'iconType', 'mask', 'position'];
        config = Object.assign(options, config);
        optionalParams.forEach((/**
         * @param {?} key
         * @return {?}
         */
        key => {
            if (config[key] !== undefined) {
                props[key] = config[key];
            }
        }));
        /** @type {?} */
        const iconType = {
            info: '',
            success: 'success',
            fail: 'fail',
            offline: 'dislike',
            loading: 'loading'
        }[options.iconType];
        props['iconType'] = iconType;
        props['mask'] = options.mask;
        props['position'] = options.position;
        return props;
    }
    /**
     * @param {?} config
     * @param {?} type
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static notice(config, type, timeInterval = 2000, onClose, mask = true, position = 'middle') {
        // 如果已经存在，在没有遮罩层的情况下，会响应别的toast，需要清除原来的
        if (ToastService.compRef) {
            ToastService.hide();
        }
        /** @type {?} */
        const options = new ToastOptions();
        options.iconType = type;
        options.mask = mask;
        options.position = position;
        /** @type {?} */
        const props = ToastService._initConfig(config, options);
        document.body.insertBefore(document.createElement(ToastService._toastCompFactory.selector), document.body.firstChild);
        /** @type {?} */
        let instance;
        /** @type {?} */
        let subject;
        ToastService.compRef = ToastService._appRef.bootstrap(ToastService._toastCompFactory);
        instance = ToastService.compRef.instance;
        subject = instance.subject;
        if (timeInterval) {
            ToastService.timeout = setTimeout((/**
             * @return {?}
             */
            () => {
                if (onClose) {
                    onClose();
                }
                ToastService.hide();
            }), timeInterval);
        }
        Object.assign(instance, props);
        return subject;
    }
    /**
     * Open info dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static info(content, timeInterval, onClose, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'info',
            content: content
        });
        return ToastService.notice(config, 'info', timeInterval, onClose, mask, position);
    }
    /**
     * Open success dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static success(content, timeInterval, onClose, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'success',
            content: content
        });
        return ToastService.notice(config, 'success', timeInterval, onClose, mask, position);
    }
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static show(content, timeInterval, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'info',
            content: content
        });
        return ToastService.notice(config, 'info', timeInterval, (/**
         * @return {?}
         */
        () => { }), mask, position);
    }
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static fail(content, timeInterval, onClose, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'fail',
            content: content
        });
        return ToastService.notice(config, 'fail', timeInterval, onClose, mask, position);
    }
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static offline(content, timeInterval, onClose, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'offline',
            content: content
        });
        return ToastService.notice(config, 'offline', timeInterval, onClose, mask, position);
    }
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @param {?=} position
     * @return {?}
     */
    static loading(content, timeInterval, onClose, mask, position) {
        /** @type {?} */
        const config = Object.assign({
            iconType: 'loading',
            content: content
        });
        return ToastService.notice(config, 'loading', timeInterval, onClose, mask, position);
    }
    /**
     * @return {?}
     */
    static hide() {
        if (ToastService.timeout) {
            clearTimeout(ToastService.timeout);
        }
        if (ToastService.compRef) {
            ToastService._zone.run((/**
             * @return {?}
             */
            () => {
                ToastService.compRef.destroy();
            }));
            ToastService.compRef = null;
        }
    }
}
ToastService.timeout = null;
ToastService._zone = null;
ToastService.compRef = null;
ToastService._toastCompFactory = null;
ToastService._appRef = null;
ToastService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
    { type: Injectable }
];
/** @nocollapse */
ToastService.ctorParameters = () => [
    { type: ApplicationRef },
    { type: Compiler },
    { type: ComponentFactoryResolver },
    { type: NgZone }
];
if (false) {
    /** @type {?} */
    ToastService.timeout;
    /** @type {?} */
    ToastService._zone;
    /** @type {?} */
    ToastService.compRef;
    /** @type {?} */
    ToastService._toastCompFactory;
    /** @type {?} */
    ToastService._appRef;
    /**
     * @type {?}
     * @private
     */
    ToastService.prototype._appRef;
    /**
     * @type {?}
     * @private
     */
    ToastService.prototype._compiler;
    /**
     * @type {?}
     * @private
     */
    ToastService.prototype._cfr;
    /**
     * @type {?}
     * @private
     */
    ToastService.prototype._zone;
}
