// ng-zorro-antd-mobile/esm5/wing-blank/wing-blank.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, ViewEncapsulation } from '@angular/core';
var WingBlankComponent = /** @class */ (function () {
    function WingBlankComponent() {
        this.prefixCls = 'am-wingblank';
        this.size = 'lg';
        this.amWingBlank = true;
    }
    Object.defineProperty(WingBlankComponent.prototype, "amWingBlnkSm", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'sm';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WingBlankComponent.prototype, "amWingBlnkMd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'md';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WingBlankComponent.prototype, "amWingBlnkLg", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'lg';
        },
        enumerable: true,
        configurable: true
    });
    WingBlankComponent.decorators = [
        { type: Component, args: [{
                    selector: 'WingBlank, nzm-wingblank',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    WingBlankComponent.ctorParameters = function () { return []; };
    WingBlankComponent.propDecorators = {
        size: [{ type: Input }],
        amWingBlank: [{ type: HostBinding, args: ['class.am-wingblank',] }],
        amWingBlnkSm: [{ type: HostBinding, args: ['class.am-wingblank-sm',] }],
        amWingBlnkMd: [{ type: HostBinding, args: ['class.am-wingblank-md',] }],
        amWingBlnkLg: [{ type: HostBinding, args: ['class.am-wingblank-lg',] }]
    };
    return WingBlankComponent;
}());
export { WingBlankComponent };
if (false) {
    /** @type {?} */
    WingBlankComponent.prototype.prefixCls;
    /** @type {?} */
    WingBlankComponent.prototype.size;
    /** @type {?} */
    WingBlankComponent.prototype.amWingBlank;
}
