// ng-zorro-antd-mobile/esm5/ng-zorro-antd-mobile.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ActivityIndicatorModule } from './activity-indicator/activity-indicator.module';
import { CardModule } from './card/card.module';
import { BadgeModule } from './badge/badge.module';
import { InputItemModule } from './input-item/input-item.module';
import { ButtonModule } from './button/button.module';
import { SwitchModule } from './switch/switch.module';
import { SearchBarModule } from './search-bar/search-bar.module';
import { StepperModule } from './stepper/stepper.module';
import { StepsModule } from './steps/steps.module';
import { CarouselModule } from './carousel/carousel.module';
import { CheckboxModule } from './checkbox/checkbox.module';
import { RadioModule } from './radio/radio.module';
import { ProgressModule } from './progress/progress.module';
import { AccordionModule } from './accordion/accordion.module';
import { NoticeBarModule } from './notice-bar/notice-bar.module';
import { IconModule } from './icon/icon.module';
import { ToastModule } from './toast/toast.module';
import { ModalModule } from './modal/modal.module';
import { PopoverModule } from './popover/popover.module';
import { NavBarModule } from './nav-bar/nav-bar.module';
import { ListModule } from './list/list.module';
import { CalendarModule } from './calendar/calendar.module';
import { PopoverItemModule } from './popover/popover-item/popover-item.module';
import { FlexModule } from './flex/flex.module';
import { GridModule } from './grid/grid.module';
import { ActionSheetModule } from './action-sheet/action-sheet.module';
import { SegmentedControlModule } from './segmented-control/segmented-control.module';
import { TextareaItemModule } from './textarea-item/textarea-item.module';
import { NgZorroAntdMobilePipesModule } from './pipes/ng-zorro-antd-mobile.pipes.module';
import { TabsModule } from './tabs/tabs.module';
import { TabBarModule } from './tab-bar/tab-bar.module';
import { PickerModule } from './picker/picker.module';
import { PickerViewModule } from './picker-view/picker-view.module';
import { ImagePickerModule } from './image-picker/image-picker.module';
import { ResultModule } from './result/result.module';
import { WhiteSpaceModule } from './white-space/white-space.module';
import { WingBlankModule } from './wing-blank/wing-blank.module';
import { RangeModule } from './range/range.module';
import { SliderModule } from './slider/slider.module';
import { PaginationModule } from './pagination/pagination.module';
import { TagModule } from './tag/tag.module';
import { MenuModule } from './menu/menu.module';
import { DatePickerModule } from './date-picker/date-picker.module';
import { DatePickerViewModule } from './date-picker-view/date-picker-view.module';
import { DrawerModule } from './drawer/drawer.module';
import { SwipeActionModule } from './swipe-action/swipe-action.module';
import { PullToRefreshModule } from './pull-to-refresh/pull-to-refresh.module';
import { LocaleProviderModule } from './locale-provider/locale-provider.module';
export { TagModule, TagComponent } from './tag/public-api';
export { CardComponent, CardModule, CardBodyComponent, CardFooterComponent, CardHeaderComponent } from './card/public-api';
export { FlexModule, FlexComponent, FlexItemComponent } from './flex/public-api';
export { GridModule, GridComponent } from './grid/public-api';
export { IconModule, IconComponent } from './icon/public-api';
export { ListModule, ListComponent, BriefComponent, ListItemComponent } from './list/public-api';
export { MenuModule, MenuComponent, SubMenuComponent } from './menu/public-api';
export { TabsModule, TabsComponent, TabPaneComponent, TabPaneBodyComponent, DefaultTabBarComponent } from './tabs/public-api';
export { BadgeModule, BadgeComponent } from './badge/public-api';
export { ModalModule, ModalService, Modal, ModalComponent, ModalServiceComponent, ModalBaseOptions, ModalOptions, AlertOptions, Action } from './modal/public-api';
export { SafeHTMLPipe, NgZorroAntdMobilePipesModule } from './pipes/public-api';
export { RadioModule, RadioComponent, RadioItemComponent, RadioItemGroupComponent } from './radio/public-api';
export { RangeModule, RangeComponent } from './range/public-api';
export { StepsModule, StepsComponent, StepComponent, StepStatusEnum, StepDirectionEnum } from './steps/public-api';
export { ToastModule, ToastComponent, ToastOptions, ToastService, Toast } from './toast/public-api';
export { ButtonModule, ButtonComponent } from './button/public-api';
export { DrawerModule, DrawerComponent } from './drawer/public-api';
export { PickerModule, PickerService, Picker, PickerComponent, PickerDirective, PickerOptions } from './picker/public-api';
export { ResultModule, ResultComponent } from './result/public-api';
export { SliderModule, SliderComponent, SliderMarksComponent, SliderStepsComponent, SliderTrackComponent, SliderHandleComponent } from './slider/public-api';
export { SwitchModule, SwitchComponent } from './switch/public-api';
export { NavBarModule, NavBarComponent } from './nav-bar/public-api';
export { PopoverDirective, PopoverComponent, PopoverItemModule, PopoverOptionsFactory, PopoverModule, PopoverItemComponent, PopoverComponentOptions, PopoverOptions } from './popover/public-api';
export { StepperModule, StepperComponent } from './stepper/public-api';
export { TabBarModule, TabBarComponent, TabBarItemComponent } from './tab-bar/public-api';
export { DateModels, CalendarModule, mergeDateTime, formatDate, CalendarHeaderComponent, CalendarComponent, CalendarDatePickerComponent, CalendarWeekPanelComponent, CalendarTimePickerComponent, CalendarSingleMonthComponent, CalendarConfirmPanelComponent, CalendarShortcutPanelComponent } from './calendar/public-api';
export { CarouselModule, CarouselComponent, DotIndicatorComponent, CarouselSlideComponent } from './carousel/public-api';
export { CheckboxModule, CheckboxComponent, AgreeItemComponent, CheckboxItemComponent } from './checkbox/public-api';
export { ProgressModule, ProgressComponent } from './progress/public-api';
export { AccordionModule, AccordionService, AccordionComponent, AccordionGroupComponent } from './accordion/public-api';
export { InputItemModule, InputItemComponent, CustomInputService, CustomInputComponent, CustomKeyboardComponent } from './input-item/public-api';
export { NoticeBarModule, NoticeBarComponent } from './notice-bar/public-api';
export { PaginationModule, PaginationComponent } from './pagination/public-api';
export { SearchBarModule, SearchBarComponent } from './search-bar/public-api';
export { WingBlankModule, WingBlankComponent } from './wing-blank/public-api';
export { DatePickerModule, DatePickerComponent, DatePickerDirective, DatePickerOptions } from './date-picker/public-api';
export { WhiteSpaceModule, WhiteSpaceComponent } from './white-space/public-api';
export { PickerViewModule, PickerViewComponent } from './picker-view/public-api';
export { ActionSheetModule, ActionSheetComponent, ActionSheetSubjectService, ActionSheetService, ActionSheet, ActionSheetOptions, ShareOption, ShareActionSheetWithOptions, ActionSheetWithOptions } from './action-sheet/public-api';
export { ImagePickerModule, ImagePickerComponent } from './image-picker/public-api';
export { SwipeActionModule, SwipeActionComponent } from './swipe-action/public-api';
export { TextareaItemModule, TextareaItemComponent } from './textarea-item/public-api';
export { en_US, ru_RU, sv_SE, zh_CN, da_DK, LocaleProviderPipe, LOCAL_PROVIDER_TOKEN, LocaleProviderModule, LocaleProviderService } from './locale-provider/public-api';
export { PullToRefreshModule, PullToRefreshComponent } from './pull-to-refresh/public-api';
export { DatePickerViewModule, DatePickerViewComponent } from './date-picker-view/public-api';
export { SegmentedControlModule, SegmentedControlComponent } from './segmented-control/public-api';
export { ActivityIndicatorModule, ActivityIndicatorComponent } from './activity-indicator/public-api';
var NgZorroAntdMobileModule = /** @class */ (function () {
    function NgZorroAntdMobileModule() {
    }
    /**
     * @return {?}
     */
    NgZorroAntdMobileModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgZorroAntdMobileModule
        };
    };
    NgZorroAntdMobileModule.decorators = [
        { type: NgModule, args: [{
                    providers: [],
                    exports: [
                        ActivityIndicatorModule,
                        CardModule,
                        TabsModule,
                        TabBarModule,
                        ButtonModule,
                        SwitchModule,
                        SearchBarModule,
                        StepperModule,
                        StepsModule,
                        CheckboxModule,
                        ProgressModule,
                        SegmentedControlModule,
                        BadgeModule,
                        CarouselModule,
                        ActionSheetModule,
                        AccordionModule,
                        NoticeBarModule,
                        IconModule,
                        ToastModule,
                        RadioModule,
                        ModalModule,
                        PopoverModule,
                        PopoverItemModule,
                        NavBarModule,
                        ListModule,
                        CalendarModule,
                        InputItemModule,
                        FlexModule,
                        GridModule,
                        RangeModule,
                        SliderModule,
                        TextareaItemModule,
                        NgZorroAntdMobilePipesModule,
                        PickerModule,
                        PickerViewModule,
                        ImagePickerModule,
                        ResultModule,
                        WhiteSpaceModule,
                        WingBlankModule,
                        LocaleProviderModule,
                        NgZorroAntdMobilePipesModule,
                        PaginationModule,
                        TagModule,
                        MenuModule,
                        DatePickerModule,
                        DatePickerViewModule,
                        DrawerModule,
                        PullToRefreshModule,
                        SwipeActionModule
                    ]
                },] }
    ];
    return NgZorroAntdMobileModule;
}());
export { NgZorroAntdMobileModule };
