// ng-zorro-antd-mobile/esm5/slider/slider-marks/slider-marks.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, Output, EventEmitter, HostBinding, ViewEncapsulation } from '@angular/core';
var SliderMarksComponent = /** @class */ (function () {
    function SliderMarksComponent(_elf) {
        this._elf = _elf;
        this.markArray = [];
        this._min = 0;
        this._max = 100;
        this._marks = {};
        this._included = true;
        this._className = 'am-slider-mark';
        this.onChange = new EventEmitter();
        this.onAfterChange = new EventEmitter();
    }
    Object.defineProperty(SliderMarksComponent.prototype, "min", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value <= this._max) {
                this._min = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "max", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value >= this._min) {
                this._max = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "marks", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._marks = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "included", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._included = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "upperBound", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value !== this._upperBound) {
                this._upperBound = value;
                this.setActiveCls();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "lowerBound", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value !== this.lowerBound) {
                this._lowerBound = value;
                this.setActiveCls();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderMarksComponent.prototype, "class", {
        get: /**
         * @return {?}
         */
        function () {
            return this._className;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} marksKeys
     * @return {?}
     */
    SliderMarksComponent.prototype.getMarks = /**
     * @param {?} marksKeys
     * @return {?}
     */
    function (marksKeys) {
        var _this = this;
        this.markArray = [];
        marksKeys
            .map(parseFloat)
            .sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) { return a - b; }))
            .map((/**
         * @param {?} point
         * @return {?}
         */
        function (point) {
            var _a;
            /** @type {?} */
            var markItem = {
                markLabel: '',
                point: '',
                className: {},
                style: {}
            };
            /** @type {?} */
            var markPoint = _this._marks[point];
            /** @type {?} */
            var markPointIsObject = typeof markPoint === 'object';
            /** @type {?} */
            var markLabel = markPointIsObject ? markPoint.label : markPoint;
            if (!markLabel && markLabel !== 0) {
                return null;
            }
            /** @type {?} */
            var isActive = (!_this._included && point === _this._upperBound) ||
                (_this._included && point <= _this._upperBound && point >= _this._lowerBound);
            /** @type {?} */
            var markClassName = (_a = {},
                _a[_this._className + "-text"] = true,
                _a[_this._className + "-text-active"] = isActive,
                _a);
            /** @type {?} */
            var bottomStyle = {
                marginBottom: '-50%',
                bottom: ((point - _this._min) / _this._range) * 100 + "%"
            };
            /** @type {?} */
            var leftStyle = {
                width: _this._markWidth + "%",
                marginLeft: -_this._markWidth / 2 + "%",
                left: ((point - _this._min) / _this._range) * 100 + "%"
            };
            /** @type {?} */
            var style = leftStyle;
            /** @type {?} */
            var markStyle = markPointIsObject ? tslib_1.__assign({}, style, markPoint.style) : style;
            markItem.markLabel = markLabel;
            markItem.point = point;
            markItem.className = Object.keys(markClassName).join(' ');
            markItem.style = markStyle;
            _this.markArray.push(markItem);
        }));
    };
    /**
     * @return {?}
     */
    SliderMarksComponent.prototype.setActiveCls = /**
     * @return {?}
     */
    function () {
        var _a;
        for (var i = 0; i < this.markArray.length; i++) {
            /** @type {?} */
            var point = this.markArray[i].point;
            /** @type {?} */
            var isActive = (!this._included && point === this._upperBound) ||
                (this._included && point <= this._upperBound && point >= this._lowerBound);
            this.markArray[i].className = (_a = {},
                _a[this._className + "-text"] = true,
                _a[this._className + "-text-active"] = isActive,
                _a);
        }
    };
    /**
     * @return {?}
     */
    SliderMarksComponent.prototype.setMarksLable = /**
     * @return {?}
     */
    function () {
        for (var i = 0; i < this.markArray.length; i++) {
            /** @type {?} */
            var markEle = this._elf.nativeElement.getElementsByClassName(this._className + '-text')[i];
            markEle.innerHTML = this.markArray[i].markLabel;
        }
    };
    /**
     * @return {?}
     */
    SliderMarksComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var marksKeys = Object.keys(this._marks);
        /** @type {?} */
        var marksCount = marksKeys.length;
        /** @type {?} */
        var unit = marksCount > 1 ? 100 / (marksCount - 1) : 100;
        this._markWidth = unit * 0.9;
        this._range = this._max - this._min;
        this.getMarks(marksKeys);
    };
    /**
     * @return {?}
     */
    SliderMarksComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.setMarksLable();
    };
    SliderMarksComponent.decorators = [
        { type: Component, args: [{
                    selector: 'SliderMarks, nzm-slider-marks',
                    template: "<span *ngFor=\"let item of markArray\"\n      [ngClass]=\"item.className\"\n      [ngStyle]=\"item.style\">\n</span>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SliderMarksComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    SliderMarksComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        marks: [{ type: Input }],
        included: [{ type: Input }],
        upperBound: [{ type: Input }],
        lowerBound: [{ type: Input }],
        onChange: [{ type: Output }],
        onAfterChange: [{ type: Output }],
        class: [{ type: HostBinding }]
    };
    return SliderMarksComponent;
}());
export { SliderMarksComponent };
if (false) {
    /** @type {?} */
    SliderMarksComponent.prototype.markArray;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._min;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._max;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._marks;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._included;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._className;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._upperBound;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._lowerBound;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._range;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._markWidth;
    /** @type {?} */
    SliderMarksComponent.prototype.onChange;
    /** @type {?} */
    SliderMarksComponent.prototype.onAfterChange;
    /**
     * @type {?}
     * @private
     */
    SliderMarksComponent.prototype._elf;
}
