// ng-zorro-antd-mobile/esm5/tab-bar/tab-bar-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, TemplateRef } from '@angular/core';
import { TabPaneComponent } from '../tabs/tab-pane.component';
var TabBarItemComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TabBarItemComponent, _super);
    function TabBarItemComponent() {
        var _this = _super.call(this) || this;
        _this.prefixCls = 'am-tab-bar-tab';
        _this.selected = false;
        _this.tintColor = '#108ee9';
        _this.unselectedTintColor = '#888';
        _this.key = '';
        _this.title = '';
        _this.dot = false;
        _this.badge = null;
        _this.icon = null;
        _this.selectedIcon = null;
        return _this;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TabBarItemComponent.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof TemplateRef;
    };
    TabBarItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'TabBarItem, nzm-tab-bar-item',
                    template: "<ng-template #content>\n  <ng-content></ng-content>\n</ng-template>\n\n<ng-template #tabBarTab>\n  <div class=\"{{prefixCls}}-icon\"\n       [style.color]=\"selected ? tintColor : unselectedTintColor\">\n    <Badge *ngIf=\"badge\" className=\"{{prefixCls}}-badge tab-badge\" [text]=\"badge\">\n      <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n      </ng-container>\n    </Badge>\n    <Badge className=\"{{prefixCls}}-badge tab-badge\" [dot]=\"dot\" *ngIf=\"dot\">\n      <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n      </ng-container>\n    </Badge>\n    <ng-container *ngIf=\"!badge && !dot\" >\n      <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n      </ng-container>\n    </ng-container>\n  </div>\n  <p class=\"{{prefixCls}}-title\" [style.color]=\"selected ? tintColor : unselectedTintColor\">\n    {{title}}\n  </p>\n</ng-template>\n\n<ng-template #domTemplate>\n  <ng-template [ngTemplateOutlet]=\"(selected ? selectedIcon : icon)\"></ng-template>\n</ng-template>\n\n<ng-template #imgTemplate >\n  <img src=\"{{selected ? selectedIcon : icon}}\" alt=\"{{title}}\">\n</ng-template>"
                }] }
    ];
    /** @nocollapse */
    TabBarItemComponent.ctorParameters = function () { return []; };
    TabBarItemComponent.propDecorators = {
        tabBarTab: [{ type: ViewChild, args: ['tabBarTab',] }],
        key: [{ type: Input }],
        title: [{ type: Input }],
        dot: [{ type: Input }],
        badge: [{ type: Input }],
        icon: [{ type: Input }],
        selectedIcon: [{ type: Input }]
    };
    return TabBarItemComponent;
}(TabPaneComponent));
export { TabBarItemComponent };
if (false) {
    /** @type {?} */
    TabBarItemComponent.prototype.prefixCls;
    /** @type {?} */
    TabBarItemComponent.prototype.selected;
    /** @type {?} */
    TabBarItemComponent.prototype.tintColor;
    /** @type {?} */
    TabBarItemComponent.prototype.unselectedTintColor;
    /** @type {?} */
    TabBarItemComponent.prototype.tabBarTab;
    /** @type {?} */
    TabBarItemComponent.prototype.key;
    /** @type {?} */
    TabBarItemComponent.prototype.title;
    /** @type {?} */
    TabBarItemComponent.prototype.dot;
    /** @type {?} */
    TabBarItemComponent.prototype.badge;
    /** @type {?} */
    TabBarItemComponent.prototype.icon;
    /** @type {?} */
    TabBarItemComponent.prototype.selectedIcon;
}
