// ng-zorro-antd-mobile/esm5/swipe-action/swipe-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
var SwipeActionComponent = /** @class */ (function () {
    function SwipeActionComponent() {
        var _this = this;
        this.prefixCls = 'am-swipe';
        this.wrapCls = {};
        this._swiping = false;
        this._openedLeft = false;
        this._openedRight = false;
        this.left = [];
        this.right = [];
        this.autoClose = false;
        this.disabled = false;
        this.onOpen = new EventEmitter();
        this.onClose = new EventEmitter();
        this.onCloseSwipe = (/**
         * @param {?} ev
         * @return {?}
         */
        function (ev) {
            if (!(_this._openedLeft || _this._openedRight)) {
                return;
            }
            /** @type {?} */
            var pNode = ev.target.closest("." + _this.prefixCls + "-actions");
            if (!pNode) {
                _this.close();
            }
        });
    }
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.setClassMap = /**
     * @return {?}
     */
    function () {
        var _a;
        this.wrapCls = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-swiping"] = this._swiping,
            _a);
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        if (this._openedLeft || this._openedRight) {
            this.onClose.emit();
        }
        this.setBtnStyle(0);
        this._openedLeft = false;
        this._openedRight = false;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    SwipeActionComponent.prototype.setBtnStyle = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var limit = value > 0 ? this._btnsLeftWidth : -this._btnsRightWidth;
        /** @type {?} */
        var contentLeft = this.getContentEasing(value, limit);
        this.content.nativeElement.style.left = contentLeft + "px";
        this.cover.nativeElement.style.display = Math.abs(value) > 0 ? 'block' : 'none';
        this.cover.nativeElement.style.left = contentLeft + "px";
    };
    /**
     * @param {?} value
     * @param {?} limit
     * @return {?}
     */
    SwipeActionComponent.prototype.getContentEasing = /**
     * @param {?} value
     * @param {?} limit
     * @return {?}
     */
    function (value, limit) {
        return Math.abs(value) - Math.abs(limit) > 0 ? limit : value;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SwipeActionComponent.prototype.onTouchStart = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this._startX = e.changedTouches[0].clientX;
        this._swiping = true;
    };
    ;
    /**
     * @param {?} e
     * @return {?}
     */
    SwipeActionComponent.prototype.onTouchMove = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        /** @type {?} */
        var deltaX = e.changedTouches[0].clientX - this._startX;
        this._needShowRight = deltaX < -5 && this.right.length > 0;
        this._needShowLeft = deltaX > 5 && this.left.length > 0;
        if (this.leftBtnRef) {
            this.leftBtnRef.nativeElement.style.visibility = this._needShowRight ? 'hidden' : 'visible';
        }
        if (this.rightBtnRef) {
            this.rightBtnRef.nativeElement.style.visibility = this._needShowLeft ? 'hidden' : 'visible';
        }
        this.setBtnStyle(deltaX);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SwipeActionComponent.prototype.onTouchEnd = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        /** @type {?} */
        var deltaX = e.changedTouches[0].clientX - this._startX;
        /** @type {?} */
        var needOpenRight = this._needShowRight && Math.abs(deltaX) > this._btnsRightWidth / 2;
        /** @type {?} */
        var needOpenLeft = this._needShowLeft && Math.abs(deltaX) > this._btnsLeftWidth / 2;
        if (needOpenRight) {
            this.doOpenRight();
        }
        else if (needOpenLeft) {
            this.doOpenLeft();
        }
        else {
            this.close();
        }
        this._swiping = false;
        this._needShowLeft = false;
        this._needShowRight = false;
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.doOpenLeft = /**
     * @return {?}
     */
    function () {
        this.open(this._btnsLeftWidth, true, false);
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.doOpenRight = /**
     * @return {?}
     */
    function () {
        this.open(-this._btnsRightWidth, false, true);
    };
    /**
     * @param {?} ev
     * @param {?} btn
     * @return {?}
     */
    SwipeActionComponent.prototype.onBtnClick = /**
     * @param {?} ev
     * @param {?} btn
     * @return {?}
     */
    function (ev, btn) {
        /** @type {?} */
        var onPress = btn.onPress;
        if (onPress) {
            onPress(ev);
        }
        if (this.autoClose) {
            this.close();
        }
    };
    /**
     * @param {?} value
     * @param {?} openedLeft
     * @param {?} openedRight
     * @return {?}
     */
    SwipeActionComponent.prototype.open = /**
     * @param {?} value
     * @param {?} openedLeft
     * @param {?} openedRight
     * @return {?}
     */
    function (value, openedLeft, openedRight) {
        this.onOpen.emit();
        this._openedLeft = openedLeft;
        this._openedRight = openedRight;
        this.setBtnStyle(value);
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClassMap();
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this._btnsLeftWidth = this.leftBtnRef ? this.leftBtnRef.nativeElement.offsetWidth : 0;
        this._btnsRightWidth = this.rightBtnRef ? this.rightBtnRef.nativeElement.offsetWidth : 0;
        document.body.addEventListener('touchstart', this.onCloseSwipe, true);
    };
    /**
     * @return {?}
     */
    SwipeActionComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        document.body.removeEventListener('touchstart', this.onCloseSwipe, true);
    };
    SwipeActionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'SwipeAction, nzm-swipe-action',
                    template: "<div *ngIf=\"(left.length != 0 || right.length != 0) && !disabled\" [ngClass]=\"wrapCls\">\n  <div class=\"{{prefixCls}}-cover\" #coverRef></div>\n  <div *ngIf=\"left && left.length > 0\"\n       class=\"{{prefixCls}}-actions {{prefixCls}}-actions-left\"\n       #leftBtnRef>\n    <div *ngFor=\"let btn of left\"\n         class=\"{{prefixCls}}-btn {{btn.className}}\"\n         [ngStyle]=\"btn.style\"\n         role=\"button\"\n         (click)=\"onBtnClick($event, btn)\">\n      <div class=\"{{prefixCls}}-btn-text\">\n        {{btn.text || 'Click'}}\n      </div>\n    </div>\n  </div>\n  <div *ngIf=\"right && right.length > 0\"\n       class=\"{{prefixCls}}-actions {{prefixCls}}-actions-right\"\n       #rightBtnRef>\n    <div *ngFor=\"let btn of right\"\n         class=\"{{prefixCls}}-btn {{btn.className}}\"\n         [ngStyle]=\"btn.style\"\n         role=\"button\"\n         (click)=\"onBtnClick($event, btn)\"\n    >\n      <div class=\"{{prefixCls}}-btn-text\">\n        {{btn.text || 'Click'}}\n      </div>\n    </div>\n  </div>\n  <div class=\"{{prefixCls}}-content\"\n       #contentRef\n       (touchstart)=\"onTouchStart($event)\"\n       (touchmove)=\"onTouchMove($event)\"\n       (touchend)=\"onTouchEnd($event)\">\n    <ng-template [ngTemplateOutlet]=\"content\"></ng-template>\n  </div>\n</div>\n<div *ngIf=\"!((left.length != 0 || right.length != 0) && !disabled)\"\n     class=\"{{prefixCls}}-content\"\n     #contentRef>\n  <ng-template [ngTemplateOutlet]=\"content\"></ng-template>\n</div>\n\n<ng-template #content>\n  <ng-content></ng-content>\n</ng-template>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SwipeActionComponent.ctorParameters = function () { return []; };
    SwipeActionComponent.propDecorators = {
        leftBtnRef: [{ type: ViewChild, args: ['leftBtnRef',] }],
        rightBtnRef: [{ type: ViewChild, args: ['rightBtnRef',] }],
        content: [{ type: ViewChild, args: ['contentRef',] }],
        cover: [{ type: ViewChild, args: ['coverRef',] }],
        left: [{ type: Input }],
        right: [{ type: Input }],
        autoClose: [{ type: Input }],
        disabled: [{ type: Input }],
        onOpen: [{ type: Output }],
        onClose: [{ type: Output }]
    };
    return SwipeActionComponent;
}());
export { SwipeActionComponent };
if (false) {
    /** @type {?} */
    SwipeActionComponent.prototype.prefixCls;
    /** @type {?} */
    SwipeActionComponent.prototype.wrapCls;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._swiping;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._openedLeft;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._openedRight;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._btnsLeftWidth;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._btnsRightWidth;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._needShowLeft;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._needShowRight;
    /**
     * @type {?}
     * @private
     */
    SwipeActionComponent.prototype._startX;
    /** @type {?} */
    SwipeActionComponent.prototype.leftBtnRef;
    /** @type {?} */
    SwipeActionComponent.prototype.rightBtnRef;
    /** @type {?} */
    SwipeActionComponent.prototype.content;
    /** @type {?} */
    SwipeActionComponent.prototype.cover;
    /** @type {?} */
    SwipeActionComponent.prototype.left;
    /** @type {?} */
    SwipeActionComponent.prototype.right;
    /** @type {?} */
    SwipeActionComponent.prototype.autoClose;
    /** @type {?} */
    SwipeActionComponent.prototype.disabled;
    /** @type {?} */
    SwipeActionComponent.prototype.onOpen;
    /** @type {?} */
    SwipeActionComponent.prototype.onClose;
    /** @type {?} */
    SwipeActionComponent.prototype.onCloseSwipe;
    /* Skipping unhandled member: ;*/
}
