// ng-zorro-antd-mobile/esm5/picker-view/picker-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, Input, forwardRef } from '@angular/core';
import { PickerComponent } from '../picker/picker.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var PickerViewComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PickerViewComponent, _super);
    function PickerViewComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.data = [];
        _this.cols = 3;
        _this.indicatorStyle = {};
        _this.itemStyle = {};
        return _this;
    }
    /**
     * @return {?}
     */
    PickerViewComponent.prototype.pickerViewInit = /**
     * @return {?}
     */
    function () {
        this.options.data = this.data;
        this.options.cols = this.cols;
        this.options.cascade = this.cascade;
        this.init();
    };
    /**
     * @return {?}
     */
    PickerViewComponent.prototype.init = /**
     * @return {?}
     */
    function () {
        this.selectedTarget = [];
        if (this.dataForRender.length === 0 && this.generateArrayData(this.options.data).length > 0) {
            this.dataForRender.push(this.generateArrayData(this.options.data));
        }
        if (this.options.value.length > 0) {
            this.getInitValueIndex(this.dataForRender);
        }
        else {
            for (var index = 0; index < this.dataForRender.length; index++) {
                this.selectedTarget.push({ targetId: "" + index, currentY: 0 });
            }
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    PickerViewComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            this.options.value = value;
            this.init();
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PickerViewComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PickerViewComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) { };
    /**
     * @return {?}
     */
    PickerViewComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.pickerViewInit();
    };
    /**
     * @return {?}
     */
    PickerViewComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.currentPicker = this.elementRef.nativeElement;
        this.reloadPicker();
    };
    PickerViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'PickerView, nzm-picker-view',
                    template: "<div #picker class=\"am-picker\" style=\"flex-direction: row; align-items: center;\">\n  <div *ngFor=\"let item of dataForRender; let i = index\" class=\"am-picker-col\">\n    <div #indicator class=\"am-picker-col-indicator \" [ngStyle]=\"indicatorStyle\"></div>\n    <div id={{i}} class=\"am-picker-col-mask\" style=\"background-size: 100% 102px;\"></div>\n    <div class=\"am-picker-col-content\">\n      <div *ngFor=\"let val of item; let i = index\"\n           id={{i}}\n           class=\"am-picker-col-item\"\n           [ngStyle]=\"itemStyle\"\n      >\n        {{val.label?val.label:val}}\n      </div>\n    </div>\n  </div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return PickerViewComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    PickerViewComponent.propDecorators = {
        data: [{ type: Input }],
        cols: [{ type: Input }],
        cascade: [{ type: Input }],
        indicatorStyle: [{ type: Input }],
        itemStyle: [{ type: Input }]
    };
    return PickerViewComponent;
}(PickerComponent));
export { PickerViewComponent };
if (false) {
    /** @type {?} */
    PickerViewComponent.prototype.options;
    /** @type {?} */
    PickerViewComponent.prototype.data;
    /** @type {?} */
    PickerViewComponent.prototype.cols;
    /** @type {?} */
    PickerViewComponent.prototype.cascade;
    /** @type {?} */
    PickerViewComponent.prototype.indicatorStyle;
    /** @type {?} */
    PickerViewComponent.prototype.itemStyle;
}
