// ng-zorro-antd-mobile/esm2015/checkbox/checkbox-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, Output, EventEmitter, ChangeDetectionStrategy, ChangeDetectorRef, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class CheckboxItemComponent {
    /**
     * @param {?} cdr
     */
    constructor(cdr) {
        this.cdr = cdr;
        this.prefixCls = 'am-checkbox';
        this.checked = false;
        this._disabled = false;
        this.wrap = false;
        this.error = false;
        this.multipleLine = false;
        this.platform = 'ios';
        this.align = 'middle';
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = value;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onCheckboxClick(event) { }
    /**
     * @param {?} event
     * @return {?}
     */
    change(event) {
        this.checked = event.checked;
        this._ngModelOnChange(event.checked);
        this.onChange.emit(event);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.checked = value;
        this.cdr.markForCheck();
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._ngModelOnChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._ngModelOnTouched = fn;
    }
}
CheckboxItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'CheckboxItem, nzm-checkbox-item',
                template: "<ListItem [className]=\"'am-checkbox-item ' + (disabled ? 'am-checkbox-item-disabled' : '')\"\n          [wrap]=\"wrap\"\n          [align]=\"align\"\n          [arrow]=\"arrow\"\n          [error]=\"error\"\n          [extra]=\"extra\"\n          [thumb]=\"checkbox\"\n          [disabled]=\"disabled\"\n          [platform]=\"platform\"\n          [multipleLine]=\"multipleLine\"\n          (onClick)=\"onCheckboxClick($event)\"\n>\n  <ng-content></ng-content>\n</ListItem>\n<ng-template #checkbox>\n  <label Checkbox\n         [name]=\"name\"\n         [value]=\"value\"\n         [checked]=\"checked\"\n         [disabled]=\"disabled\"\n         (onChange)=\"change($event)\"\n  >\n  </label>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => CheckboxItemComponent)),
                        multi: true
                    }
                ]
            }] }
];
/** @nocollapse */
CheckboxItemComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
CheckboxItemComponent.propDecorators = {
    name: [{ type: Input }],
    value: [{ type: Input }],
    arrow: [{ type: Input }],
    extra: [{ type: Input }],
    wrap: [{ type: Input }],
    error: [{ type: Input }],
    multipleLine: [{ type: Input }],
    platform: [{ type: Input }],
    align: [{ type: Input }],
    disabled: [{ type: Input }],
    onChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CheckboxItemComponent.prototype.prefixCls;
    /** @type {?} */
    CheckboxItemComponent.prototype.checked;
    /**
     * @type {?}
     * @private
     */
    CheckboxItemComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    CheckboxItemComponent.prototype._ngModelOnChange;
    /**
     * @type {?}
     * @private
     */
    CheckboxItemComponent.prototype._ngModelOnTouched;
    /** @type {?} */
    CheckboxItemComponent.prototype.name;
    /** @type {?} */
    CheckboxItemComponent.prototype.value;
    /** @type {?} */
    CheckboxItemComponent.prototype.arrow;
    /** @type {?} */
    CheckboxItemComponent.prototype.extra;
    /** @type {?} */
    CheckboxItemComponent.prototype.wrap;
    /** @type {?} */
    CheckboxItemComponent.prototype.error;
    /** @type {?} */
    CheckboxItemComponent.prototype.multipleLine;
    /** @type {?} */
    CheckboxItemComponent.prototype.platform;
    /** @type {?} */
    CheckboxItemComponent.prototype.align;
    /** @type {?} */
    CheckboxItemComponent.prototype.onChange;
    /**
     * @type {?}
     * @private
     */
    CheckboxItemComponent.prototype.cdr;
}
