// ng-zorro-antd-mobile/esm2015/flex/flex.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, HostBinding } from '@angular/core';
export class FlexComponent {
    constructor() {
        this.defaultProps = {
            prefixCls: 'am-flexbox',
            align: 'center'
        };
        // _wrapCls = {};
        this._direction = '';
        this._justify = '';
        this._alignContent = '';
        this.amFlexbox = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set direction(value) {
        this._direction = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set wrap(value) {
        this._wrap = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set justify(value) {
        this._justify = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set align(value) {
        this.defaultProps.align = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set alignContent(value) {
        this._alignContent = value;
    }
    /**
     * @return {?}
     */
    get amFlexboxDirRow() {
        return this._direction === 'row';
    }
    /**
     * @return {?}
     */
    get amFlexboxDirRowReverse() {
        return this._direction === 'row-reverse';
    }
    /**
     * @return {?}
     */
    get amFlexboxDirColumn() {
        return this._direction === 'column';
    }
    /**
     * @return {?}
     */
    get amFlexboxDirColumnReverse() {
        return this._direction === 'column-reverse';
    }
    /**
     * @return {?}
     */
    get amFlexboxNowrap() {
        return this._wrap === 'nowrap';
    }
    /**
     * @return {?}
     */
    get amFlexboxWrap() {
        return this._wrap === 'wrap';
    }
    /**
     * @return {?}
     */
    get amFlexboxWrapReverse() {
        return this._wrap === 'wrap-reverse';
    }
    /**
     * @return {?}
     */
    get amFlexboxJustifyStart() {
        return this._justify === 'start';
    }
    /**
     * @return {?}
     */
    get amFlexboxJustifyCenter() {
        return this._justify === 'center';
    }
    /**
     * @return {?}
     */
    get amFlexboxJustifyEnd() {
        return this._justify === 'end';
    }
    /**
     * @return {?}
     */
    get amFlexboxJustifyBetween() {
        return this._justify === 'between';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignAround() {
        return this._justify === 'around';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignStart() {
        return this.defaultProps.align === 'start';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignCenter() {
        return this.defaultProps.align === 'center';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignEnd() {
        return this.defaultProps.align === 'end';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignBaseline() {
        return this.defaultProps.align === 'baseline';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignStretch() {
        return this.defaultProps.align === 'stretch';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignContentStart() {
        return this._alignContent === 'start';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignCotentCenter() {
        return this._alignContent === 'center';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignContentEnd() {
        return this._alignContent === 'end';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignContentBetween() {
        return this._alignContent === 'between';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignContentAround() {
        return this._alignContent === 'around';
    }
    /**
     * @return {?}
     */
    get amFlexboxAlignContentStretch() {
        return this._alignContent === 'stretch';
    }
}
FlexComponent.decorators = [
    { type: Component, args: [{
                selector: 'Flex, nzm-flex',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
FlexComponent.ctorParameters = () => [];
FlexComponent.propDecorators = {
    direction: [{ type: Input }],
    wrap: [{ type: Input }],
    justify: [{ type: Input }],
    align: [{ type: Input }],
    alignContent: [{ type: Input }],
    amFlexbox: [{ type: HostBinding, args: ['class.am-flexbox',] }],
    amFlexboxDirRow: [{ type: HostBinding, args: ['class.am-flexbox-dir-row',] }],
    amFlexboxDirRowReverse: [{ type: HostBinding, args: ['class.am-flexbox-dir-row-reverse',] }],
    amFlexboxDirColumn: [{ type: HostBinding, args: ['class.am-flexbox-dir-column',] }],
    amFlexboxDirColumnReverse: [{ type: HostBinding, args: ['class.am-flexbox-dir-column-reverse',] }],
    amFlexboxNowrap: [{ type: HostBinding, args: ['class.am-flexbox-nowrap',] }],
    amFlexboxWrap: [{ type: HostBinding, args: ['class.am-flexbox-wrap',] }],
    amFlexboxWrapReverse: [{ type: HostBinding, args: ['class.am-flexbox-wrap-reverse',] }],
    amFlexboxJustifyStart: [{ type: HostBinding, args: ['class.am-flexbox-justify-start',] }],
    amFlexboxJustifyCenter: [{ type: HostBinding, args: ['class.am-flexbox-justify-center',] }],
    amFlexboxJustifyEnd: [{ type: HostBinding, args: ['class.am-flexbox-justify-end',] }],
    amFlexboxJustifyBetween: [{ type: HostBinding, args: ['class.am-flexbox-justify-between',] }],
    amFlexboxAlignAround: [{ type: HostBinding, args: ['class.am-flexbox-justify-around',] }],
    amFlexboxAlignStart: [{ type: HostBinding, args: ['class.am-flexbox-align-start',] }],
    amFlexboxAlignCenter: [{ type: HostBinding, args: ['class.am-flexbox-align-center',] }],
    amFlexboxAlignEnd: [{ type: HostBinding, args: ['class.am-flexbox-align-end',] }],
    amFlexboxAlignBaseline: [{ type: HostBinding, args: ['class.am-flexbox-align-baseline',] }],
    amFlexboxAlignStretch: [{ type: HostBinding, args: ['class.am-flexbox-align-stretch',] }],
    amFlexboxAlignContentStart: [{ type: HostBinding, args: ['class.am-flexbox-align-content-start',] }],
    amFlexboxAlignCotentCenter: [{ type: HostBinding, args: ['class.am-flexbox-align-content-center',] }],
    amFlexboxAlignContentEnd: [{ type: HostBinding, args: ['class.am-flexbox-align-content-end',] }],
    amFlexboxAlignContentBetween: [{ type: HostBinding, args: ['class.am-flexbox-align-content-between',] }],
    amFlexboxAlignContentAround: [{ type: HostBinding, args: ['class.am-flexbox-align-content-around',] }],
    amFlexboxAlignContentStretch: [{ type: HostBinding, args: ['class.am-flexbox-align-content-stretch',] }]
};
if (false) {
    /** @type {?} */
    FlexComponent.prototype.defaultProps;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._direction;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._wrap;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._justify;
    /**
     * @type {?}
     * @private
     */
    FlexComponent.prototype._alignContent;
    /** @type {?} */
    FlexComponent.prototype.amFlexbox;
}
export class FlexItemComponent {
    constructor() {
        this.defaultProps = {
            prefixCls: 'am-flexbox',
            align: 'center'
        };
        this.flexboxItem = true;
    }
}
FlexItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'FlexItem, nzm-flex-item',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            }] }
];
FlexItemComponent.propDecorators = {
    flexboxItem: [{ type: HostBinding, args: ['class.am-flexbox-item',] }]
};
if (false) {
    /** @type {?} */
    FlexItemComponent.prototype.defaultProps;
    /** @type {?} */
    FlexItemComponent.prototype.flexboxItem;
}
