// ng-zorro-antd-mobile/esm5/tabs/tab-pane-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, TemplateRef, ViewEncapsulation } from '@angular/core';
var TabPaneBodyComponent = /** @class */ (function () {
    function TabPaneBodyComponent() {
        this._prerender = false;
        this.active = false;
        this.loaded = false;
        this.paneWrap = true;
    }
    Object.defineProperty(TabPaneBodyComponent.prototype, "prerender", {
        get: /**
         * @return {?}
         */
        function () {
            return this._prerender;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._prerender = value;
            if (value) {
                this.loaded = true;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPaneBodyComponent.prototype, "wrapActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TabPaneBodyComponent.prototype, "wrapInactive", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.active;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TabPaneBodyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    TabPaneBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: '[tab-pane-body]',
                    template: "<ng-container *ngIf=\"loaded || prerender\">\n  <ng-template [ngTemplateOutlet]=\"content\"></ng-template>\n</ng-container>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    TabPaneBodyComponent.ctorParameters = function () { return []; };
    TabPaneBodyComponent.propDecorators = {
        active: [{ type: Input }],
        loaded: [{ type: Input }],
        content: [{ type: Input }],
        prerender: [{ type: Input }],
        paneWrap: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap',] }],
        wrapActive: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap-active',] }],
        wrapInactive: [{ type: HostBinding, args: ['class.am-tabs-pane-wrap-inactive',] }]
    };
    return TabPaneBodyComponent;
}());
export { TabPaneBodyComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TabPaneBodyComponent.prototype._prerender;
    /** @type {?} */
    TabPaneBodyComponent.prototype.active;
    /** @type {?} */
    TabPaneBodyComponent.prototype.loaded;
    /** @type {?} */
    TabPaneBodyComponent.prototype.content;
    /** @type {?} */
    TabPaneBodyComponent.prototype.paneWrap;
}
