// ng-zorro-antd-mobile/esm5/popover/popover-item/popover-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, ElementRef, TemplateRef, HostBinding, HostListener, EventEmitter, ViewEncapsulation } from '@angular/core';
var PopoverItemComponent = /** @class */ (function () {
    function PopoverItemComponent(_elementRef) {
        this._elementRef = _elementRef;
        this.defaultProps = {
            prefixCls: 'am-popover',
            disabled: false
        };
        this.isActive = false;
        this.select = new EventEmitter();
        this.amPopoverItem = true;
    }
    Object.defineProperty(PopoverItemComponent.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._icon;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._icon = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverItemComponent.prototype, "style", {
        get: /**
         * @return {?}
         */
        function () {
            return this._style;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._style = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverItemComponent.prototype, "disabled", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.defaultProps.disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverItemComponent.prototype, "amPopoverItemActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isActive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverItemComponent.prototype, "amPopoverItemDisabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.defaultProps.disabled;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} e
     * @return {?}
     */
    PopoverItemComponent.prototype.touchStart = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.select.emit();
        this.isActive = true;
    };
    /**
     * @return {?}
     */
    PopoverItemComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () { };
    PopoverItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'PopoverItem, nzm-popover-item',
                    template: "<div class=\"{{defaultProps.prefixCls}}-item-container\">\n  <span class=\"{{defaultProps.prefixCls}}-item-icon\" aria-hidden=\"true\">\n    <ng-template [ngTemplateOutlet]=\"icon\"></ng-template>\n  </span>\n  <span class=\"{{defaultProps.prefixCls}}-item-content\">\n    <ng-content></ng-content>\n  </span>\n</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    PopoverItemComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    PopoverItemComponent.propDecorators = {
        icon: [{ type: Input }],
        style: [{ type: Input }],
        disabled: [{ type: Input }],
        select: [{ type: Output }],
        amPopoverItem: [{ type: HostBinding, args: ['class.am-popover-item',] }],
        amPopoverItemActive: [{ type: HostBinding, args: ['class.am-popover-item-active',] }],
        amPopoverItemDisabled: [{ type: HostBinding, args: ['class.am-popover-item-disabled',] }],
        touchStart: [{ type: HostListener, args: ['touchstart', ['$event'],] }, { type: HostListener, args: ['mousedown', ['$event'],] }]
    };
    return PopoverItemComponent;
}());
export { PopoverItemComponent };
if (false) {
    /** @type {?} */
    PopoverItemComponent.prototype.defaultProps;
    /** @type {?} */
    PopoverItemComponent.prototype.isActive;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._style;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._icon;
    /** @type {?} */
    PopoverItemComponent.prototype.select;
    /** @type {?} */
    PopoverItemComponent.prototype.amPopoverItem;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._elementRef;
}
