// ng-zorro-antd-mobile/esm5/core/directive/touch-feedback.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostListener, ViewEncapsulation, ElementRef, Renderer2 } from '@angular/core';
var TouchFeedBackComponent = /** @class */ (function () {
    function TouchFeedBackComponent(_ele, _render) {
        this._ele = _ele;
        this._render = _render;
        this.activeClassName = '';
        this.activeStyle = true;
    }
    /**
     * @return {?}
     */
    TouchFeedBackComponent.prototype.touchStart = /**
     * @return {?}
     */
    function () {
        if (this.activeStyle) {
            this._render.addClass(this._ele.nativeElement, this.activeClassName);
        }
    };
    /**
     * @return {?}
     */
    TouchFeedBackComponent.prototype.touchEnd = /**
     * @return {?}
     */
    function () {
        if (this.activeStyle) {
            this._render.removeClass(this._ele.nativeElement, this.activeClassName);
        }
    };
    TouchFeedBackComponent.decorators = [
        { type: Component, args: [{
                    selector: 'TouchFeedBack',
                    template: '<ng-content></ng-content>',
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    TouchFeedBackComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    TouchFeedBackComponent.propDecorators = {
        activeClassName: [{ type: Input }],
        activeStyle: [{ type: Input }],
        touchStart: [{ type: HostListener, args: ['touchstart',] }, { type: HostListener, args: ['mousedown',] }],
        touchEnd: [{ type: HostListener, args: ['touchend',] }, { type: HostListener, args: ['mouseup',] }]
    };
    return TouchFeedBackComponent;
}());
export { TouchFeedBackComponent };
if (false) {
    /** @type {?} */
    TouchFeedBackComponent.prototype.activeClassName;
    /** @type {?} */
    TouchFeedBackComponent.prototype.activeStyle;
    /**
     * @type {?}
     * @private
     */
    TouchFeedBackComponent.prototype._ele;
    /**
     * @type {?}
     * @private
     */
    TouchFeedBackComponent.prototype._render;
}
