// ng-zorro-antd-mobile/esm2015/popover/popover-item/popover-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, ElementRef, TemplateRef, HostBinding, HostListener, EventEmitter, ViewEncapsulation } from '@angular/core';
export class PopoverItemComponent {
    /**
     * @param {?} _elementRef
     */
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        this.defaultProps = {
            prefixCls: 'am-popover',
            disabled: false
        };
        this.isActive = false;
        this.select = new EventEmitter();
        this.amPopoverItem = true;
    }
    /**
     * @return {?}
     */
    get icon() {
        return this._icon;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set icon(value) {
        this._icon = value;
    }
    /**
     * @return {?}
     */
    get style() {
        return this._style;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set style(value) {
        this._style = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this.defaultProps.disabled = value;
    }
    /**
     * @return {?}
     */
    get amPopoverItemActive() {
        return this.isActive;
    }
    /**
     * @return {?}
     */
    get amPopoverItemDisabled() {
        return this.defaultProps.disabled;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    touchStart(e) {
        this.select.emit();
        this.isActive = true;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() { }
}
PopoverItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'PopoverItem, nzm-popover-item',
                template: "<div class=\"{{defaultProps.prefixCls}}-item-container\">\n  <span class=\"{{defaultProps.prefixCls}}-item-icon\" aria-hidden=\"true\">\n    <ng-template [ngTemplateOutlet]=\"icon\"></ng-template>\n  </span>\n  <span class=\"{{defaultProps.prefixCls}}-item-content\">\n    <ng-content></ng-content>\n  </span>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
PopoverItemComponent.ctorParameters = () => [
    { type: ElementRef }
];
PopoverItemComponent.propDecorators = {
    icon: [{ type: Input }],
    style: [{ type: Input }],
    disabled: [{ type: Input }],
    select: [{ type: Output }],
    amPopoverItem: [{ type: HostBinding, args: ['class.am-popover-item',] }],
    amPopoverItemActive: [{ type: HostBinding, args: ['class.am-popover-item-active',] }],
    amPopoverItemDisabled: [{ type: HostBinding, args: ['class.am-popover-item-disabled',] }],
    touchStart: [{ type: HostListener, args: ['touchstart', ['$event'],] }, { type: HostListener, args: ['mousedown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    PopoverItemComponent.prototype.defaultProps;
    /** @type {?} */
    PopoverItemComponent.prototype.isActive;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._style;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._icon;
    /** @type {?} */
    PopoverItemComponent.prototype.select;
    /** @type {?} */
    PopoverItemComponent.prototype.amPopoverItem;
    /**
     * @type {?}
     * @private
     */
    PopoverItemComponent.prototype._elementRef;
}
