// ng-zorro-antd-mobile/esm2015/list/list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding, TemplateRef } from '@angular/core';
export class ListComponent {
    constructor() {
        this.defaultProps = {
            prefixCls: 'am-list'
        };
        this.renderHeaderType = '';
        this.renderFooterType = '';
        this._renderHeader = '';
        this._renderFooter = '';
        this._className = '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set className(value) {
        this._className = value;
    }
    /**
     * @return {?}
     */
    get renderHeader() {
        return this._renderHeader;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set renderHeader(value) {
        if (value instanceof TemplateRef) {
            this.renderHeaderType = 'templateRef';
        }
        else {
            this.renderHeaderType = typeof value;
        }
        console.log('this isi rend', this.renderHeaderType);
        this._renderHeader = value;
    }
    /**
     * @return {?}
     */
    get renderFooter() {
        return this._renderFooter;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set renderFooter(value) {
        if (value instanceof TemplateRef) {
            this.renderFooterType = 'templateRef';
        }
        else {
            this.renderFooterType = typeof value;
        }
        this._renderFooter = value;
    }
    /**
     * @return {?}
     */
    get hostClassName() {
        return 'am-list ' + this._className;
    }
}
ListComponent.decorators = [
    { type: Component, args: [{
                selector: 'List, nzm-list',
                template: "<div *ngIf=\"renderHeader && renderHeaderType === 'string'\"\n     class=\"{{defaultProps.prefixCls}}-header\"\n     [innerHTML]=\"renderHeader\">\n</div>\n<div *ngIf=\"renderHeader && renderHeaderType === 'function'\"\n     class=\"{{defaultProps.prefixCls}}-header\">{{renderHeader()}}\n</div>\n<ng-template *ngIf=\"renderHeader && renderHeaderType === 'templateRef'\" [ngTemplateOutlet]=\"renderHeader\">\n</ng-template>\n<div class=\"{{defaultProps.prefixCls}}-body\">\n  <ng-content></ng-content>\n</div>\n<div *ngIf=\"renderFooter && renderFooterType === 'string'\"\n     class=\"{{defaultProps.prefixCls}}-footer\"\n     [innerHTML]=\"renderFooter\">\n</div>\n<div *ngIf=\"renderFooter && renderFooterType === 'function'\"\n     class=\"{{defaultProps.prefixCls}}-footer\">{{renderFooter()}}\n</div>\n<ng-template *ngIf=\"renderFooter && renderFooterType === 'templateRef'\" [ngTemplateOutlet]=\"renderFooter\">\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
ListComponent.ctorParameters = () => [];
ListComponent.propDecorators = {
    className: [{ type: Input }],
    renderHeader: [{ type: Input }],
    renderFooter: [{ type: Input }],
    hostClassName: [{ type: HostBinding, args: ['class',] }]
};
if (false) {
    /** @type {?} */
    ListComponent.prototype.defaultProps;
    /** @type {?} */
    ListComponent.prototype.renderHeaderType;
    /** @type {?} */
    ListComponent.prototype.renderFooterType;
    /**
     * @type {?}
     * @private
     */
    ListComponent.prototype._renderHeader;
    /**
     * @type {?}
     * @private
     */
    ListComponent.prototype._renderFooter;
    /**
     * @type {?}
     * @private
     */
    ListComponent.prototype._className;
}
