// ng-zorro-antd-mobile/esm5/calendar/week-panel/week-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation, Input } from '@angular/core';
var CalendarWeekPanelComponent = /** @class */ (function () {
    function CalendarWeekPanelComponent() {
        this.week = ['日', '一', '二', '三', '四', '五', '六'];
        this.weekPanel = true;
    }
    Object.defineProperty(CalendarWeekPanelComponent.prototype, "locale", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._locale = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CalendarWeekPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.week = this._locale.week;
    };
    CalendarWeekPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CalendarWeekPanel, nzm-calendar-week-panel',
                    template: "<div class=\"cell cell-grey\">{{week[0]}}</div>\n<div class=\"cell\">{{week[1]}}</div>\n<div class=\"cell\">{{week[2]}}</div>\n<div class=\"cell\">{{week[3]}}</div>\n<div class=\"cell\">{{week[4]}}</div>\n<div class=\"cell\">{{week[5]}}</div>\n<div class=\"cell cell-grey\">{{week[6]}}</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    CalendarWeekPanelComponent.ctorParameters = function () { return []; };
    CalendarWeekPanelComponent.propDecorators = {
        locale: [{ type: Input }],
        weekPanel: [{ type: HostBinding, args: ['class.week-panel',] }]
    };
    return CalendarWeekPanelComponent;
}());
export { CalendarWeekPanelComponent };
if (false) {
    /** @type {?} */
    CalendarWeekPanelComponent.prototype.week;
    /**
     * @type {?}
     * @private
     */
    CalendarWeekPanelComponent.prototype._locale;
    /** @type {?} */
    CalendarWeekPanelComponent.prototype.weekPanel;
}
