// ng-zorro-antd-mobile/esm2015/notice-bar/util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} rule
 * @param {?} className
 * @return {?}
 */
function insertKeyFrame(rule, className) {
    /** @type {?} */
    const style = document.createElement('style');
    style.setAttribute('class', className);
    style.innerHTML = rule;
    document.body.appendChild(style);
}
/**
 * @param {?} className
 * @return {?}
 */
function deleteKeyFrame(className) {
    /** @type {?} */
    const styleDom = document.getElementsByClassName(className);
    while (styleDom.length > 0) {
        styleDom[0].remove();
    }
}
/**
 * @return {?}
 */
function getWidthHeight() {
    /** @type {?} */
    const w = window;
    /** @type {?} */
    const d = document;
    /** @type {?} */
    const e = d.documentElement;
    /** @type {?} */
    const g = d.getElementsByTagName('body')[0];
    return {
        width: w.innerWidth || e.clientWidth || g.clientWidth,
        height: w.innerHeight || e.clientHeight || g.clientHeight
    };
}
/**
 * @param {?} text
 * @param {?} font
 * @return {?}
 */
function getTextWidth(text, font) {
    /** @type {?} */
    const _dom = document.createElement('div');
    _dom.innerHTML = text;
    _dom.style.position = 'absolute';
    _dom.style.left = '-9999';
    _dom.style.whiteSpace = 'nowrap';
    _dom.style.fontSize = font;
    document.body.appendChild(_dom);
    /** @type {?} */
    const _w = _dom.clientWidth + 10;
    document.body.removeChild(_dom);
    return _w;
}
export { insertKeyFrame, deleteKeyFrame, getWidthHeight, getTextWidth };
