// ng-zorro-antd-mobile/esm5/locale-provider/locale-provider.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, Optional, SkipSelf } from '@angular/core';
import { LOCAL_PROVIDER_TOKEN } from './locale-provider.token';
import { BehaviorSubject } from 'rxjs';
import zh_CN from './locale/zh_CN';
var LocaleProviderService = /** @class */ (function () {
    function LocaleProviderService(locale) {
        this._change = new BehaviorSubject(this._locale);
        this.setLocale(locale || zh_CN);
    }
    Object.defineProperty(LocaleProviderService.prototype, "localeChange", {
        get: /**
         * @return {?}
         */
        function () {
            return this._change.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} keyPath
     * @return {?}
     */
    LocaleProviderService.prototype.getLocaleValue = /**
     * @param {?} keyPath
     * @return {?}
     */
    function (keyPath) {
        /** @type {?} */
        var content = (/** @type {?} */ (this._getObjectPath(this._locale, keyPath)));
        if (typeof content === 'string') {
            return content;
        }
        return '';
    };
    /**
     * @param {?} keyPath
     * @return {?}
     */
    LocaleProviderService.prototype.getLocaleSubObj = /**
     * @param {?} keyPath
     * @return {?}
     */
    function (keyPath) {
        /** @type {?} */
        var content = (/** @type {?} */ (this._getObjectPath(this._locale, keyPath)));
        if (typeof content === 'object') {
            return content;
        }
        return null;
    };
    /**
     * @param {?} locale
     * @return {?}
     */
    LocaleProviderService.prototype.setLocale = /**
     * @param {?} locale
     * @return {?}
     */
    function (locale) {
        if (!locale || (this._locale && this._locale.locale === locale.locale)) {
            return;
        }
        this._locale = locale;
        this._change.next(locale);
    };
    /**
     * @return {?}
     */
    LocaleProviderService.prototype.getLocaleId = /**
     * @return {?}
     */
    function () {
        return this._locale && this._locale.locale ? this._locale.locale : '';
    };
    /**
     * @return {?}
     */
    LocaleProviderService.prototype.getLocale = /**
     * @return {?}
     */
    function () {
        return this._locale;
    };
    /**
     * @private
     * @param {?} obj
     * @param {?} path
     * @return {?}
     */
    LocaleProviderService.prototype._getObjectPath = /**
     * @private
     * @param {?} obj
     * @param {?} path
     * @return {?}
     */
    function (obj, path) {
        /** @type {?} */
        var res = obj;
        /** @type {?} */
        var paths = path.split('.');
        /** @type {?} */
        var depth = paths.length;
        /** @type {?} */
        var index = 0;
        while (res && index < depth) {
            res = res[paths[index++]];
        }
        return index === depth ? res : null;
    };
    LocaleProviderService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LocaleProviderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [LOCAL_PROVIDER_TOKEN,] }] }
    ]; };
    return LocaleProviderService;
}());
export { LocaleProviderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LocaleProviderService.prototype._locale;
    /**
     * @type {?}
     * @private
     */
    LocaleProviderService.prototype._change;
}
/**
 * @param {?} exist
 * @param {?} locale
 * @return {?}
 */
export function LOCALE_PROVIDER_SERVICE_FACTORY(exist, locale) {
    return exist || new LocaleProviderService(locale);
}
/** @type {?} */
export var LOCALE_PROVIDER_SERVICE_PROVIDER = {
    provide: LocaleProviderService,
    useFactory: LOCALE_PROVIDER_SERVICE_FACTORY,
    deps: [[new Optional(), new SkipSelf(), LocaleProviderService], LOCAL_PROVIDER_TOKEN]
};
