// ng-zorro-antd-mobile/esm5/locale-provider/locale-provider.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import zh_CN from './locale/zh_CN';
import { LocaleProviderPipe } from './locale-provider.pipe';
import { LOCAL_PROVIDER_TOKEN } from './locale-provider.token';
import { LOCALE_PROVIDER_SERVICE_PROVIDER } from './locale-provider.service';
var ɵ0 = zh_CN;
var LocaleProviderModule = /** @class */ (function () {
    function LocaleProviderModule() {
    }
    LocaleProviderModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [LocaleProviderPipe],
                    exports: [LocaleProviderPipe],
                    providers: [{ provide: LOCAL_PROVIDER_TOKEN, useValue: ɵ0 }, LOCALE_PROVIDER_SERVICE_PROVIDER]
                },] }
    ];
    return LocaleProviderModule;
}());
export { LocaleProviderModule };
export { ɵ0 };
