// ng-zorro-antd-mobile/esm5/popover/popover.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { PopoverComponentOptions } from './popover-component-options.provider';
var PopoverComponent = /** @class */ (function () {
    function PopoverComponent(options) {
        this.options = options;
        this.defaultProps = {
            prefixCls: 'am-popover'
        };
        this.maskCls = {};
        this.popoverCls = {};
    }
    /**
     * @return {?}
     */
    PopoverComponent.prototype.setClassMap = /**
     * @return {?}
     */
    function () {
        var _a, _b;
        this.maskCls = (_a = {},
            _a[this.defaultProps.prefixCls + "-mask"] = this.options.mask,
            _a[this.defaultProps.prefixCls + "-mask-hidden"] = !this.options.mask,
            _a);
        this.popoverCls = (_b = {},
            _b["" + this.defaultProps.prefixCls] = true,
            _b[this.defaultProps.prefixCls + "-placement-" + this.options.placement] = true,
            _b[this.defaultProps.prefixCls + "-" + this.options.className] = true,
            _b);
    };
    /**
     * @return {?}
     */
    PopoverComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClassMap();
    };
    /**
     * @return {?}
     */
    PopoverComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.options.onAfterViewInit();
    };
    PopoverComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Popover',
                    template: "<ng-content></ng-content>\n<div [ngClass]=\"maskCls\" (click)=\"options.hidePopover()\"></div>\n<div [ngClass]=\"popoverCls\" style=\"color: currentcolor;\">\n  <div class=\"{{defaultProps.prefixCls}}-content\">\n    <div *ngIf=\"options.showArrow\" class=\"{{defaultProps.prefixCls}}-arrow\"></div>\n    <div class=\"{{defaultProps.prefixCls}}-inner\">\n      <div class=\"{{defaultProps.prefixCls}}-inner-wrapper\">\n        <ng-template [ngTemplateOutlet]=\"options.overlay\" [ngTemplateOutletContext]=\"{options: options}\">\n        </ng-template>\n      </div>\n    </div>\n  </div>\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    PopoverComponent.ctorParameters = function () { return [
        { type: PopoverComponentOptions }
    ]; };
    return PopoverComponent;
}());
export { PopoverComponent };
if (false) {
    /** @type {?} */
    PopoverComponent.prototype.defaultProps;
    /** @type {?} */
    PopoverComponent.prototype.maskCls;
    /** @type {?} */
    PopoverComponent.prototype.popoverCls;
    /** @type {?} */
    PopoverComponent.prototype.options;
}
