// ng-zorro-antd-mobile/esm5/popover/popover.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PopoverDirective } from './popover.directive';
import { PopoverComponent } from './popover.component';
import { PopoverOptions } from './popover-options.provider';
/**
 * @param {?} userOptions
 * @return {?}
 */
export function PopoverOptionsFactory(userOptions) {
    /** @type {?} */
    var options = new PopoverOptions();
    Object.assign(options, userOptions);
    return options;
}
var PopoverModule = /** @class */ (function () {
    function PopoverModule() {
    }
    PopoverModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [PopoverDirective, PopoverComponent],
                    imports: [CommonModule],
                    exports: [PopoverDirective, PopoverComponent],
                    entryComponents: [PopoverComponent],
                },] }
    ];
    return PopoverModule;
}());
export { PopoverModule };
