// ng-zorro-antd-mobile/esm5/action-sheet/action-sheet-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var ActionSheetOptions = /** @class */ (function () {
    function ActionSheetOptions() {
        this.prefixCls = 'am-action-sheet';
        this.maskClosable = true;
        this.transitionName = 'am-slide-up';
        this.maskTransitionName = 'am-fade';
    }
    ActionSheetOptions.decorators = [
        { type: Injectable }
    ];
    return ActionSheetOptions;
}());
export { ActionSheetOptions };
if (false) {
    /** @type {?} */
    ActionSheetOptions.prototype.prefixCls;
    /** @type {?} */
    ActionSheetOptions.prototype.maskClosable;
    /** @type {?} */
    ActionSheetOptions.prototype.cancelButtonIndex;
    /** @type {?} */
    ActionSheetOptions.prototype.destructiveButtonIndex;
    /** @type {?} */
    ActionSheetOptions.prototype.title;
    /** @type {?} */
    ActionSheetOptions.prototype.message;
    /** @type {?} */
    ActionSheetOptions.prototype.className;
    /** @type {?} */
    ActionSheetOptions.prototype.transitionName;
    /** @type {?} */
    ActionSheetOptions.prototype.maskTransitionName;
}
var ShareOption = /** @class */ (function () {
    function ShareOption() {
    }
    ShareOption.decorators = [
        { type: Injectable }
    ];
    return ShareOption;
}());
export { ShareOption };
if (false) {
    /** @type {?} */
    ShareOption.prototype.icon;
    /** @type {?} */
    ShareOption.prototype.title;
}
var ShareActionSheetWithOptions = /** @class */ (function (_super) {
    tslib_1.__extends(ShareActionSheetWithOptions, _super);
    function ShareActionSheetWithOptions() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.cancelButtonText = 'Cancel';
        return _this;
    }
    ShareActionSheetWithOptions.decorators = [
        { type: Injectable }
    ];
    return ShareActionSheetWithOptions;
}(ActionSheetOptions));
export { ShareActionSheetWithOptions };
if (false) {
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.options;
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.cancelButtonText;
    /** @type {?} */
    ShareActionSheetWithOptions.prototype.locale;
}
var ActionSheetWithOptions = /** @class */ (function (_super) {
    tslib_1.__extends(ActionSheetWithOptions, _super);
    function ActionSheetWithOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ActionSheetWithOptions.decorators = [
        { type: Injectable }
    ];
    return ActionSheetWithOptions;
}(ActionSheetOptions));
export { ActionSheetWithOptions };
if (false) {
    /** @type {?} */
    ActionSheetWithOptions.prototype.options;
}
