// ng-zorro-antd-mobile/esm5/action-sheet/action-sheet.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ActionSheetComponent } from './action-sheet.component';
import { ActionSheetOptions, ShareActionSheetWithOptions } from './action-sheet-options.provider';
import { PopupService } from '../core/services/popup.service';
import * as i0 from "@angular/core";
import * as i1 from "./public-api";
import * as i2 from "@angular/cdk/overlay";
/** @type {?} */
var NORMAL = 'NORMAL';
/** @type {?} */
var SHARE = 'SHARE';
/**
 * @return {?}
 */
function noop() { }
var ActionSheetService = /** @class */ (function (_super) {
    tslib_1.__extends(ActionSheetService, _super);
    function ActionSheetService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} config
     * @param {?=} options
     * @return {?}
     */
    ActionSheetService._initConfig = /**
     * @param {?} config
     * @param {?=} options
     * @return {?}
     */
    function (config, options) {
        if (options === void 0) { options = {}; }
        /** @type {?} */
        var props = new ActionSheetOptions();
        /** @type {?} */
        var optionalParams = [
            'prefixCls',
            'maskClosable',
            'cancelButtonText',
            'cancelButtonIndex',
            'destructiveButtonIndex',
            'title',
            'message',
            'className',
            'transitionName',
            'maskTransitionName',
            'options',
            'locale'
        ];
        config = Object.assign(options, config, {
            close: (/**
             * @return {?}
             */
            function () {
                if (config.maskClosable) {
                    ActionSheetService.closeWithAnimation(config.transitionName, config.maskTransitionName);
                }
            })
        });
        optionalParams.forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            if (config[key] !== undefined) {
                props[key] = config[key];
            }
        }));
        return props;
    };
    /**
     * @param {?} props
     * @return {?}
     */
    ActionSheetService._open = /**
     * @param {?} props
     * @return {?}
     */
    function (props) {
        setTimeout((/**
         * @return {?}
         */
        function () {
            ActionSheetService.comRef = ActionSheetService.showPopup('ActionSheetComponent', ActionSheetComponent);
            ActionSheetService.comRef.instance.option = props;
        }), 0);
    };
    /**
     * @param {?} flag
     * @param {?} config
     * @param {?} callback
     * @return {?}
     */
    ActionSheetService.createActionSheet = /**
     * @param {?} flag
     * @param {?} config
     * @param {?} callback
     * @return {?}
     */
    function (flag, config, callback) {
        /** @type {?} */
        var options = flag === NORMAL ? new ActionSheetOptions() : new ShareActionSheetWithOptions();
        /** @type {?} */
        var transitionName = config.transitionName ? config.transitionName : options.transitionName;
        options.transitionName = transitionName + "-enter " + transitionName + "-enter-active";
        /** @type {?} */
        var maskTransitionName = config.maskTransitionName ? config.maskTransitionName : options.maskTransitionName;
        options.maskTransitionName = maskTransitionName + "-enter " + maskTransitionName + "-enter-active";
        /** @type {?} */
        var props = ActionSheetService._initConfig(config, options);
        Object.assign(props, { onPress: cb }, { flag: flag }, { maskClose: props.maskClosable ? cb : (/**
             * @return {?}
             */
            function () { }) });
        /**
         * @param {?} index
         * @param {?=} rowIndex
         * @param {?=} event
         * @return {?}
         */
        function cb(index, rowIndex, event) {
            if (rowIndex === void 0) { rowIndex = 0; }
            event.stopPropagation();
            /** @type {?} */
            var res = callback(index, rowIndex);
            if (res && res.then) {
                res.then((/**
                 * @return {?}
                 */
                function () {
                    ActionSheetService.closeWithAnimation(transitionName, maskTransitionName);
                }));
            }
            else {
                ActionSheetService.closeWithAnimation(transitionName, maskTransitionName);
            }
        }
        return ActionSheetService._open(props);
    };
    /**
     * @param {?} transitionName
     * @param {?} maskTransitionName
     * @return {?}
     */
    ActionSheetService.closeWithAnimation = /**
     * @param {?} transitionName
     * @param {?} maskTransitionName
     * @return {?}
     */
    function (transitionName, maskTransitionName) {
        ActionSheetService.comRef.instance.option.transitionName = transitionName + "-leave " + transitionName + "-leave-active";
        ActionSheetService.comRef.instance.option.maskTransitionName = maskTransitionName + "-leave " + maskTransitionName + "-leave-active";
        setTimeout((/**
         * @return {?}
         */
        function () {
            ActionSheetService.close();
        }), 200);
    };
    /**
     * @param {?} config
     * @param {?=} callback
     * @return {?}
     */
    ActionSheetService.showActionSheetWithOptions = /**
     * @param {?} config
     * @param {?=} callback
     * @return {?}
     */
    function (config, callback) {
        if (callback === void 0) { callback = noop; }
        ActionSheetService.createActionSheet(NORMAL, config, callback);
    };
    /**
     * @param {?} config
     * @param {?=} callback
     * @return {?}
     */
    ActionSheetService.showShareActionSheetWithOptions = /**
     * @param {?} config
     * @param {?=} callback
     * @return {?}
     */
    function (config, callback) {
        if (callback === void 0) { callback = noop; }
        ActionSheetService.createActionSheet(SHARE, config, callback);
    };
    /**
     * @return {?}
     */
    ActionSheetService.close = /**
     * @return {?}
     */
    function () {
        ActionSheetService.hidePopup('ActionSheetComponent');
    };
    ActionSheetService.compRef = null;
    ActionSheetService._actionSheetCompFactory = null;
    ActionSheetService.appRef = null;
    ActionSheetService.comRef = null;
    ActionSheetService.instance = null;
    ActionSheetService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ActionSheetService.ngInjectableDef = i0.defineInjectable({ factory: function ActionSheetService_Factory() { return new i1.ActionSheet(i0.inject(i2.Overlay)); }, token: i1.ActionSheet, providedIn: "root" });
    return ActionSheetService;
}(PopupService));
export { ActionSheetService };
if (false) {
    /** @type {?} */
    ActionSheetService.compRef;
    /** @type {?} */
    ActionSheetService._actionSheetCompFactory;
    /** @type {?} */
    ActionSheetService.appRef;
    /** @type {?} */
    ActionSheetService.comRef;
    /** @type {?} */
    ActionSheetService.instance;
}
