// ng-zorro-antd-mobile/esm5/input-item/input-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, HostBinding, Renderer2, ElementRef, forwardRef, TemplateRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var InputItemComponent = /** @class */ (function () {
    function InputItemComponent(element, render) {
        this.element = element;
        this.render = render;
        this.prefixCls = 'am-input';
        this.setFocus = {};
        this.pattern = '';
        this.autoFocus = false;
        this.inputType = 'text';
        this.ngTemplate = false;
        this._type = 'text';
        this._defaultValue = '';
        this._placeholder = '';
        this._editable = true;
        this._disabled = false;
        this._clear = false;
        this._error = false;
        this._extra = '';
        this._labelNumber = 5;
        this._updatePlaceholder = false;
        this._prefixListCls = 'am-list';
        this._moneyKeyboardAlign = 'right';
        this._focus = false;
        this._isClear = false;
        this._inputLock = false;
        this.onChange = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onErrorClick = new EventEmitter();
        this.onExtraClick = new EventEmitter();
        this.clsItem = true;
        this.clsDisabled = this._disabled;
        this.clsError = this._error;
        this.clsFocus = this._focus;
        this.clsAndroid = this._focus;
        this._onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        this._el = element.nativeElement;
    }
    Object.defineProperty(InputItemComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this._type;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value.length > 0) {
                this.inputType = value;
                if (value === 'bankCard' || value === 'phone') {
                    this._type = 'tel';
                }
                else if (value === 'password') {
                    this._type = 'password';
                }
                else if (value === 'digit') {
                    this._type = 'number';
                }
                else if (value !== 'text' && value !== 'number') {
                    this._type = value;
                }
                if (value === 'number') {
                    this.pattern = '[0-9]*';
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (typeof v === 'undefined' || v === null) {
                this._value = '';
            }
            else {
                this._value = v;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "defaultValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._defaultValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._defaultValue = value;
            if (!this._value) {
                this._value = this._defaultValue;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "placeholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this._placeholder;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._placeholder = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "editable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._editable;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._editable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
            this.clsDisabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "clear", {
        get: /**
         * @return {?}
         */
        function () {
            return this._clear;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._clear = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "maxLength", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maxLength;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._maxLength = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "error", {
        get: /**
         * @return {?}
         */
        function () {
            return this._error;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._error = value;
            this.clsError = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "extra", {
        get: /**
         * @return {?}
         */
        function () {
            return this._extra;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this.ngTemplate = true;
            }
            else {
                this.ngTemplate = false;
            }
            this._extra = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "labelNumber", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._labelNumber = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "updatePlaceholder", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._updatePlaceholder = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "prefixListCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._prefixListCls;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._prefixListCls = value;
            this.render.addClass(this._el, value + '-item');
            this.render.addClass(this._el, value + '-item-middle');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "name", {
        get: /**
         * @return {?}
         */
        function () {
            return this._name;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._name = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "moneyKeyboardAlign", {
        get: /**
         * @return {?}
         */
        function () {
            return this._moneyKeyboardAlign;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._moneyKeyboardAlign = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "locale", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._locale = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "fontColor", {
        get: /**
         * @return {?}
         */
        function () {
            return this._fontColor;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._fontColor = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputItemComponent.prototype, "focus", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value.focus) {
                this.autoFocus = value.focus;
                if (this._type === 'money') {
                    this.setFocus = value;
                }
                else if (this.inputElementRef) {
                    this._focus = true;
                    this.inputElementRef.nativeElement.focus();
                    this.inputFocus('');
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    InputItemComponent.prototype.setCls = /**
     * @return {?}
     */
    function () {
        var _a, _b;
        if (this.lableRef.nativeElement.children.length > 0 ||
            (this.lableRef.nativeElement && this.lableRef.nativeElement.innerText !== '')) {
            this.labelCls = (_a = {},
                _a[this.prefixCls + "-label"] = true,
                _a[this.prefixCls + "-label-2"] = this._labelNumber === 2,
                _a[this.prefixCls + "-label-3"] = this._labelNumber === 3,
                _a[this.prefixCls + "-label-4"] = this._labelNumber === 4,
                _a[this.prefixCls + "-label-5"] = this._labelNumber === 5,
                _a[this.prefixCls + "-label-6"] = this._labelNumber === 6,
                _a[this.prefixCls + "-label-7"] = this._labelNumber === 7,
                _a);
        }
        this.controlCls = (_b = {}, _b[this.prefixCls + "-control"] = true, _b);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    InputItemComponent.prototype.inputChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this._inputLock && _this.inputType === 'text') {
                return;
            }
            /** @type {?} */
            var value = e;
            switch (_this.inputType) {
                case 'text':
                    break;
                case 'bankCard':
                    value = value.replace(/\D/g, '').replace(/(....)(?=.)/g, '$1 ');
                    break;
                case 'phone':
                    value = value.replace(/\D/g, '').substring(0, 11);
                    /** @type {?} */
                    var valueLen = value.length;
                    if (valueLen > 3 && valueLen < 8) {
                        value = value.substr(0, 3) + " " + value.substr(3);
                    }
                    else if (valueLen >= 8) {
                        value = value.substr(0, 3) + " " + value.substr(3, 4) + " " + value.substr(7);
                    }
                    break;
                case 'number':
                    value = value.replace(/\D/g, '');
                    break;
                case 'password':
                    break;
                default:
                    _this._value = value;
                    break;
            }
            _this._value = value;
            if (_this._type !== 'money') {
                _this.inputElementRef.nativeElement.value = _this._value;
            }
            _this._onChange(_this._value);
            _this.onChange.emit(_this._value);
        }), 0);
    };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.compositionStart = /**
     * @return {?}
     */
    function () {
        this._inputLock = true;
    };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.compositionEnd = /**
     * @return {?}
     */
    function () {
        this._inputLock = false;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    InputItemComponent.prototype.inputFocus = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this._focus = true;
            _this.clsFocus = true;
            _this.clsAndroid = true;
        }), 100);
        this.onFocus.emit(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    InputItemComponent.prototype.inputBlur = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (!_this._isClear) {
                _this._focus = false;
                _this.clsFocus = false;
                _this.clsAndroid = false;
                _this.onBlur.emit(value);
            }
            _this._isClear = false;
        }), 100);
    };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.clearInput = /**
     * @return {?}
     */
    function () {
        if (this._type !== 'password' && this._updatePlaceholder) {
            this._placeholder = this._value;
        }
        this._value = '';
        this.onChange.emit(this._value);
        this._onChange(this._value);
        this._isClear = true;
        if (this._type !== 'money') {
            this.inputElementRef.nativeElement.value = this._value;
        }
        this.inputFocus(this._value);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    InputItemComponent.prototype.errorClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.onErrorClick) {
            this.onErrorClick.emit(e);
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    InputItemComponent.prototype.extraClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.onExtraClick) {
            this.onExtraClick.emit(e);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    InputItemComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (typeof value === undefined || value === null) {
            this._value = '';
        }
        else {
            this._value = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    InputItemComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    InputItemComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) { };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (this.inputElementRef && this._type !== 'money' && this._value !== undefined) {
            this.inputElementRef.nativeElement.value = this._value;
        }
    };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setCls();
        this.render.addClass(this._el, this._prefixListCls + '-item');
        this.render.addClass(this._el, this._prefixListCls + '-item-middle');
    };
    /**
     * @return {?}
     */
    InputItemComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.setCls();
        }), 0);
    };
    InputItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'InputItem, nzm-input-item',
                    template: "<div class=\"{{prefixListCls}}-line\">\n  <div #lableContent [ngClass]=\"labelCls\">\n    <ng-content></ng-content>\n  </div>\n  <div [ngClass]=\"controlCls\">\n    <CustomInput *ngIf=\"type==='money'\"\n                 [value]=\"value\"\n                 [defaultValue]=\"defaultValue\"\n                 [placeholder]=\"placeholder\"\n                 [disabled]=\"disabled\"\n                 [editable]=\"editable\"\n                 [fontColor]=\"fontColor\"\n                 [moneyKeyboardAlign]=\"moneyKeyboardAlign\"\n                 [setFocus]=\"setFocus\"\n                 [maxLength]=\"maxLength\"\n                 (onChange)=\"inputChange($event)\"\n                 (onBlur)=\"inputBlur(value)\"\n                 (onFocus)=\"inputFocus(value)\">\n    </CustomInput>\n    <div *ngIf=\"type!=='money'\">\n      <input #inputElement\n             [type]=\"type\"\n             [name]=\"name\"\n             [ngModel]=\"value\"\n             [defaultValue]=\"defaultValue\"\n             [placeholder]=\"placeholder\"\n             [disabled]=\"disabled\"\n             [readOnly]=\"!editable\"\n             [autofocus]=\"autoFocus\"\n             [maxlength]=\"maxLength\"\n             [pattern]=\"pattern\"\n             [style.color]='fontColor'\n             (ngModelChange)=\"inputChange($event)\"\n             (compositionstart)=\"compositionStart()\"\n             (compositionend)=\"compositionEnd()\"\n             (blur)=\"inputBlur(value)\"\n             (focus)=\"inputFocus(value)\"\n             style=\"outline:none\" />\n    </div>\n  </div>\n  <div  *ngIf=\"clear && editable && !disabled && (value && value.length >0 )\"\n        class=\"{{prefixCls}}-clear\"\n        (click)=\"clearInput()\"\n  ></div>\n  <div *ngIf=\"error\"\n       class=\"{{prefixCls}}-error-extra\"\n       (click)=\"errorClick($event)\"\n  ></div>\n  <div *ngIf=\"extra!==''\"\n       class=\"{{prefixCls}}-extra\"\n       (click)=\"extraClick($event)\">\n    <ng-container *ngIf=\"!ngTemplate\">{{extra}}</ng-container>\n    <ng-template *ngIf=\"ngTemplate\" [ngTemplateOutlet]=\"extra\"></ng-template>\n  </div>\n</div>",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return InputItemComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    InputItemComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    InputItemComponent.propDecorators = {
        lableRef: [{ type: ViewChild, args: ['lableContent',] }],
        inputElementRef: [{ type: ViewChild, args: ['inputElement',] }],
        type: [{ type: Input }],
        value: [{ type: Input }],
        defaultValue: [{ type: Input }],
        placeholder: [{ type: Input }],
        editable: [{ type: Input }],
        disabled: [{ type: Input }],
        clear: [{ type: Input }],
        maxLength: [{ type: Input }],
        error: [{ type: Input }],
        extra: [{ type: Input }],
        labelNumber: [{ type: Input }],
        updatePlaceholder: [{ type: Input }],
        prefixListCls: [{ type: Input }],
        name: [{ type: Input }],
        moneyKeyboardAlign: [{ type: Input }],
        locale: [{ type: Input }],
        fontColor: [{ type: Input }],
        focus: [{ type: Input }],
        onChange: [{ type: Output }],
        onBlur: [{ type: Output }],
        onFocus: [{ type: Output }],
        onErrorClick: [{ type: Output }],
        onExtraClick: [{ type: Output }],
        clsItem: [{ type: HostBinding, args: ['class.am-input-item',] }],
        clsDisabled: [{ type: HostBinding, args: ['class.am-input-disabled',] }],
        clsError: [{ type: HostBinding, args: ['class.am-input-error',] }],
        clsFocus: [{ type: HostBinding, args: ['class.am-input-focus',] }],
        clsAndroid: [{ type: HostBinding, args: ['class.am-input-android,',] }]
    };
    return InputItemComponent;
}());
export { InputItemComponent };
if (false) {
    /** @type {?} */
    InputItemComponent.prototype.prefixCls;
    /** @type {?} */
    InputItemComponent.prototype.wrapCls;
    /** @type {?} */
    InputItemComponent.prototype.labelCls;
    /** @type {?} */
    InputItemComponent.prototype.controlCls;
    /** @type {?} */
    InputItemComponent.prototype.setFocus;
    /** @type {?} */
    InputItemComponent.prototype.pattern;
    /** @type {?} */
    InputItemComponent.prototype.autoFocus;
    /** @type {?} */
    InputItemComponent.prototype.inputType;
    /** @type {?} */
    InputItemComponent.prototype.ngTemplate;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._el;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._type;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._defaultValue;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._placeholder;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._editable;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._clear;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._maxLength;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._error;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._extra;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._labelNumber;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._updatePlaceholder;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._prefixListCls;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._name;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._moneyKeyboardAlign;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._locale;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._focus;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._isClear;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._fontColor;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype._inputLock;
    /** @type {?} */
    InputItemComponent.prototype.lableRef;
    /** @type {?} */
    InputItemComponent.prototype.inputElementRef;
    /** @type {?} */
    InputItemComponent.prototype.onChange;
    /** @type {?} */
    InputItemComponent.prototype.onBlur;
    /** @type {?} */
    InputItemComponent.prototype.onFocus;
    /** @type {?} */
    InputItemComponent.prototype.onErrorClick;
    /** @type {?} */
    InputItemComponent.prototype.onExtraClick;
    /** @type {?} */
    InputItemComponent.prototype.clsItem;
    /** @type {?} */
    InputItemComponent.prototype.clsDisabled;
    /** @type {?} */
    InputItemComponent.prototype.clsError;
    /** @type {?} */
    InputItemComponent.prototype.clsFocus;
    /** @type {?} */
    InputItemComponent.prototype.clsAndroid;
    /** @type {?} */
    InputItemComponent.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype.element;
    /**
     * @type {?}
     * @private
     */
    InputItemComponent.prototype.render;
}
